// @midscene/shared/dist/lib/img.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/img/index.ts
var img_exports = {};
__export(img_exports, {
  bufferFromBase64: () => bufferFromBase64,
  compositeElementInfoImg: () => compositeElementInfoImg,
  cropByRect: () => cropByRect,
  drawBoxOnImage: () => drawBoxOnImage,
  httpImg2Base64: () => httpImg2Base64,
  imageInfo: () => imageInfo,
  imageInfoOfBase64: () => imageInfoOfBase64,
  isValidPNGImageBuffer: () => isValidPNGImageBuffer,
  jimpFromBase64: () => jimpFromBase64,
  jimpToBase64: () => jimpToBase64,
  localImg2Base64: () => localImg2Base64,
  paddingToMatchBlock: () => paddingToMatchBlock,
  paddingToMatchBlockByBase64: () => paddingToMatchBlockByBase64,
  preProcessImageUrl: () => preProcessImageUrl,
  processImageElementInfo: () => processImageElementInfo,
  resizeImg: () => resizeImg,
  resizeImgBase64: () => resizeImgBase64,
  saveBase64Image: () => saveBase64Image,
  savePositionImg: () => savePositionImg,
  zoomForGPT4o: () => zoomForGPT4o
});
module.exports = __toCommonJS(img_exports);

// src/img/info.ts
var import_node_assert = __toESM(require("assert"));
var import_node_buffer = require("buffer");

// src/img/get-jimp.ts
var import_jimp = __toESM(require("jimp"));

// src/utils.ts
var import_js_sha256 = require("js-sha256");
var ifInBrowser = typeof window !== "undefined";
var ifInWorker = typeof WorkerGlobalScope !== "undefined";
var ifInNode = typeof process !== "undefined" && process.versions?.node;

// src/img/get-jimp.ts
async function getJimp() {
  if (ifInBrowser || ifInWorker) {
    await import("jimp/browser/lib/jimp.js");
    return (typeof window !== "undefined" ? window : self).Jimp;
  }
  return import_jimp.default;
}

// src/img/info.ts
async function imageInfo(image) {
  const Jimp2 = await getJimp();
  let jimpImage;
  if (typeof image === "string") {
    jimpImage = await Jimp2.read(image);
  } else if (import_node_buffer.Buffer.isBuffer(image)) {
    jimpImage = await Jimp2.read(image);
  } else if (image instanceof Jimp2) {
    jimpImage = image;
  } else {
    throw new Error("Invalid image input: must be a string path or a Buffer");
  }
  const { width, height } = jimpImage.bitmap;
  (0, import_node_assert.default)(
    width && height,
    `Invalid image: ${typeof image === "string" ? image : "Buffer"}`
  );
  return { width, height, jimpImage };
}
async function imageInfoOfBase64(imageBase64) {
  const buffer = await bufferFromBase64(imageBase64);
  return imageInfo(buffer);
}
async function bufferFromBase64(imageBase64) {
  const base64Data = imageBase64.replace(/^data:image\/\w+;base64,/, "");
  return import_node_buffer.Buffer.from(base64Data, "base64");
}
function isValidPNGImageBuffer(buffer) {
  if (!buffer || buffer.length < 8) {
    return false;
  }
  const isPNG = buffer[0] === 137 && buffer[1] === 80 && buffer[2] === 78 && buffer[3] === 71;
  return isPNG;
}

// src/img/transform.ts
var import_node_assert2 = __toESM(require("assert"));
var import_node_buffer2 = require("buffer");
var import_node_fs3 = require("fs");
var import_node_path3 = __toESM(require("path"));

// src/logger.ts
var import_node_fs2 = __toESM(require("fs"));
var import_node_path2 = __toESM(require("path"));
var import_node_util = __toESM(require("util"));
var import_debug = __toESM(require("debug"));

// src/common.ts
var import_node_fs = require("fs");
var import_node_os = require("os");
var import_node_path = __toESM(require("path"));

// src/env.ts
var MIDSCENE_OPENAI_INIT_CONFIG_JSON = "MIDSCENE_OPENAI_INIT_CONFIG_JSON";
var MIDSCENE_MODEL_NAME = "MIDSCENE_MODEL_NAME";
var MIDSCENE_LANGSMITH_DEBUG = "MIDSCENE_LANGSMITH_DEBUG";
var MIDSCENE_DEBUG_AI_PROFILE = "MIDSCENE_DEBUG_AI_PROFILE";
var MIDSCENE_DEBUG_AI_RESPONSE = "MIDSCENE_DEBUG_AI_RESPONSE";
var MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG = "MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG";
var MIDSCENE_DEBUG_MODE = "MIDSCENE_DEBUG_MODE";
var MIDSCENE_MCP_USE_PUPPETEER_MODE = "MIDSCENE_MCP_USE_PUPPETEER_MODE";
var MIDSCENE_MCP_ANDROID_MODE = "MIDSCENE_MCP_ANDROID_MODE";
var MIDSCENE_FORCE_DEEP_THINK = "MIDSCENE_FORCE_DEEP_THINK";
var MIDSCENE_OPENAI_SOCKS_PROXY = "MIDSCENE_OPENAI_SOCKS_PROXY";
var MIDSCENE_OPENAI_HTTP_PROXY = "MIDSCENE_OPENAI_HTTP_PROXY";
var OPENAI_API_KEY = "OPENAI_API_KEY";
var OPENAI_BASE_URL = "OPENAI_BASE_URL";
var OPENAI_MAX_TOKENS = "OPENAI_MAX_TOKENS";
var MIDSCENE_ADB_PATH = "MIDSCENE_ADB_PATH";
var MIDSCENE_ADB_REMOTE_HOST = "MIDSCENE_ADB_REMOTE_HOST";
var MIDSCENE_ADB_REMOTE_PORT = "MIDSCENE_ADB_REMOTE_PORT";
var MIDSCENE_ANDROID_IME_STRATEGY = "MIDSCENE_ANDROID_IME_STRATEGY";
var MIDSCENE_CACHE = "MIDSCENE_CACHE";
var MIDSCENE_USE_VLM_UI_TARS = "MIDSCENE_USE_VLM_UI_TARS";
var MIDSCENE_USE_QWEN_VL = "MIDSCENE_USE_QWEN_VL";
var MIDSCENE_USE_DOUBAO_VISION = "MIDSCENE_USE_DOUBAO_VISION";
var MIDSCENE_USE_GEMINI = "MIDSCENE_USE_GEMINI";
var MIDSCENE_USE_VL_MODEL = "MIDSCENE_USE_VL_MODEL";
var MATCH_BY_POSITION = "MATCH_BY_POSITION";
var MIDSCENE_REPORT_TAG_NAME = "MIDSCENE_REPORT_TAG_NAME";
var MIDSCENE_REPLANNING_CYCLE_LIMIT = "MIDSCENE_REPLANNING_CYCLE_LIMIT";
var MIDSCENE_PREFERRED_LANGUAGE = "MIDSCENE_PREFERRED_LANGUAGE";
var MIDSCENE_USE_AZURE_OPENAI = "MIDSCENE_USE_AZURE_OPENAI";
var MIDSCENE_AZURE_OPENAI_SCOPE = "MIDSCENE_AZURE_OPENAI_SCOPE";
var MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON = "MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON";
var MIDSCENE_CACHE_MAX_FILENAME_LENGTH = "MIDSCENE_CACHE_MAX_FILENAME_LENGTH";
var AZURE_OPENAI_ENDPOINT = "AZURE_OPENAI_ENDPOINT";
var AZURE_OPENAI_KEY = "AZURE_OPENAI_KEY";
var AZURE_OPENAI_API_VERSION = "AZURE_OPENAI_API_VERSION";
var AZURE_OPENAI_DEPLOYMENT = "AZURE_OPENAI_DEPLOYMENT";
var MIDSCENE_USE_ANTHROPIC_SDK = "MIDSCENE_USE_ANTHROPIC_SDK";
var ANTHROPIC_API_KEY = "ANTHROPIC_API_KEY";
var MIDSCENE_RUN_DIR = "MIDSCENE_RUN_DIR";
var OPENAI_USE_AZURE = "OPENAI_USE_AZURE";
var allConfigFromEnv = () => {
  return {
    [MIDSCENE_MCP_ANDROID_MODE]: process.env[MIDSCENE_MCP_ANDROID_MODE] || void 0,
    [MIDSCENE_OPENAI_INIT_CONFIG_JSON]: process.env[MIDSCENE_OPENAI_INIT_CONFIG_JSON] || void 0,
    [MIDSCENE_MODEL_NAME]: process.env[MIDSCENE_MODEL_NAME] || void 0,
    [MIDSCENE_DEBUG_MODE]: process.env[MIDSCENE_DEBUG_MODE] || void 0,
    [MIDSCENE_FORCE_DEEP_THINK]: process.env[MIDSCENE_FORCE_DEEP_THINK] || void 0,
    [MIDSCENE_LANGSMITH_DEBUG]: process.env[MIDSCENE_LANGSMITH_DEBUG] || void 0,
    [MIDSCENE_DEBUG_AI_PROFILE]: process.env[MIDSCENE_DEBUG_AI_PROFILE] || void 0,
    [MIDSCENE_DEBUG_AI_RESPONSE]: process.env[MIDSCENE_DEBUG_AI_RESPONSE] || void 0,
    [MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG]: process.env[MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG] || void 0,
    [OPENAI_API_KEY]: process.env[OPENAI_API_KEY] || void 0,
    [OPENAI_BASE_URL]: process.env[OPENAI_BASE_URL] || void 0,
    [OPENAI_MAX_TOKENS]: process.env[OPENAI_MAX_TOKENS] || void 0,
    [OPENAI_USE_AZURE]: process.env[OPENAI_USE_AZURE] || void 0,
    [MIDSCENE_ADB_PATH]: process.env[MIDSCENE_ADB_PATH] || void 0,
    [MIDSCENE_ADB_REMOTE_HOST]: process.env[MIDSCENE_ADB_REMOTE_HOST] || void 0,
    [MIDSCENE_ADB_REMOTE_PORT]: process.env[MIDSCENE_ADB_REMOTE_PORT] || void 0,
    [MIDSCENE_ANDROID_IME_STRATEGY]: process.env[MIDSCENE_ANDROID_IME_STRATEGY] || void 0,
    [MIDSCENE_CACHE]: process.env[MIDSCENE_CACHE] || void 0,
    [MATCH_BY_POSITION]: process.env[MATCH_BY_POSITION] || void 0,
    [MIDSCENE_REPORT_TAG_NAME]: process.env[MIDSCENE_REPORT_TAG_NAME] || void 0,
    [MIDSCENE_OPENAI_SOCKS_PROXY]: process.env[MIDSCENE_OPENAI_SOCKS_PROXY] || void 0,
    [MIDSCENE_OPENAI_HTTP_PROXY]: process.env[MIDSCENE_OPENAI_HTTP_PROXY] || void 0,
    [MIDSCENE_USE_AZURE_OPENAI]: process.env[MIDSCENE_USE_AZURE_OPENAI] || void 0,
    [MIDSCENE_AZURE_OPENAI_SCOPE]: process.env[MIDSCENE_AZURE_OPENAI_SCOPE] || void 0,
    [MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON]: process.env[MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON] || void 0,
    [MIDSCENE_USE_ANTHROPIC_SDK]: process.env[MIDSCENE_USE_ANTHROPIC_SDK] || void 0,
    [MIDSCENE_USE_VLM_UI_TARS]: process.env[MIDSCENE_USE_VLM_UI_TARS] || void 0,
    [MIDSCENE_USE_QWEN_VL]: process.env[MIDSCENE_USE_QWEN_VL] || void 0,
    [MIDSCENE_USE_DOUBAO_VISION]: process.env[MIDSCENE_USE_DOUBAO_VISION] || void 0,
    [MIDSCENE_USE_GEMINI]: process.env[MIDSCENE_USE_GEMINI] || void 0,
    [MIDSCENE_USE_VL_MODEL]: process.env[MIDSCENE_USE_VL_MODEL] || void 0,
    [ANTHROPIC_API_KEY]: process.env[ANTHROPIC_API_KEY] || void 0,
    [AZURE_OPENAI_ENDPOINT]: process.env[AZURE_OPENAI_ENDPOINT] || void 0,
    [AZURE_OPENAI_KEY]: process.env[AZURE_OPENAI_KEY] || void 0,
    [AZURE_OPENAI_API_VERSION]: process.env[AZURE_OPENAI_API_VERSION] || void 0,
    [AZURE_OPENAI_DEPLOYMENT]: process.env[AZURE_OPENAI_DEPLOYMENT] || void 0,
    [MIDSCENE_MCP_USE_PUPPETEER_MODE]: process.env[MIDSCENE_MCP_USE_PUPPETEER_MODE] || void 0,
    [MIDSCENE_RUN_DIR]: process.env[MIDSCENE_RUN_DIR] || void 0,
    [MIDSCENE_PREFERRED_LANGUAGE]: process.env[MIDSCENE_PREFERRED_LANGUAGE] || void 0,
    [MIDSCENE_REPLANNING_CYCLE_LIMIT]: process.env[MIDSCENE_REPLANNING_CYCLE_LIMIT] || void 0,
    [MIDSCENE_CACHE_MAX_FILENAME_LENGTH]: process.env[MIDSCENE_CACHE_MAX_FILENAME_LENGTH] || void 0
  };
};
var getGlobalConfig = () => {
  if (!globalThis.midsceneGlobalConfig) {
    globalThis.midsceneGlobalConfig = allConfigFromEnv();
  }
  return globalThis.midsceneGlobalConfig;
};
var getAIConfig = (configKey) => {
  if (configKey === MATCH_BY_POSITION) {
    throw new Error(
      "MATCH_BY_POSITION is deprecated, use MIDSCENE_USE_VL_MODEL instead"
    );
  }
  const value = getGlobalConfig()[configKey];
  if (typeof value === "string") {
    return value.trim();
  }
  return value;
};

// src/common.ts
var defaultRunDirName = "midscene_run";
var getMidsceneRunDir = () => {
  if (!ifInNode) {
    return "";
  }
  return getAIConfig(MIDSCENE_RUN_DIR) || defaultRunDirName;
};
var getMidsceneRunBaseDir = () => {
  if (!ifInNode) {
    return "";
  }
  let basePath = import_node_path.default.resolve(process.cwd(), getMidsceneRunDir());
  if (!(0, import_node_fs.existsSync)(basePath)) {
    try {
      (0, import_node_fs.mkdirSync)(basePath, { recursive: true });
    } catch (error) {
      basePath = import_node_path.default.join((0, import_node_os.tmpdir)(), defaultRunDirName);
      (0, import_node_fs.mkdirSync)(basePath, { recursive: true });
    }
  }
  return basePath;
};
var getMidsceneRunSubDir = (subdir) => {
  if (!ifInNode) {
    return "";
  }
  const basePath = getMidsceneRunBaseDir();
  const logPath = import_node_path.default.join(basePath, subdir);
  if (!(0, import_node_fs.existsSync)(logPath)) {
    (0, import_node_fs.mkdirSync)(logPath, { recursive: true });
  }
  return logPath;
};

// src/logger.ts
var topicPrefix = "midscene";
var logStreams = /* @__PURE__ */ new Map();
var debugInstances = /* @__PURE__ */ new Map();
function getLogStream(topic) {
  const topicFileName = topic.replace(/:/g, "-");
  if (!logStreams.has(topicFileName)) {
    const logFile = import_node_path2.default.join(
      getMidsceneRunSubDir("log"),
      `${topicFileName}.log`
    );
    const stream = import_node_fs2.default.createWriteStream(logFile, { flags: "a" });
    logStreams.set(topicFileName, stream);
  }
  return logStreams.get(topicFileName);
}
function writeLogToFile(topic, message) {
  if (!ifInNode)
    return;
  const stream = getLogStream(topic);
  const now = /* @__PURE__ */ new Date();
  const isoDate = now.toLocaleDateString("sv-SE");
  const isoTime = now.toLocaleTimeString("sv-SE");
  const milliseconds = now.getMilliseconds().toString().padStart(3, "0");
  const timezoneOffsetMinutes = now.getTimezoneOffset();
  const sign = timezoneOffsetMinutes <= 0 ? "+" : "-";
  const hours = Math.floor(Math.abs(timezoneOffsetMinutes) / 60).toString().padStart(2, "0");
  const minutes = (Math.abs(timezoneOffsetMinutes) % 60).toString().padStart(2, "0");
  const timezoneString = `${sign}${hours}:${minutes}`;
  const localISOTime = `${isoDate}T${isoTime}.${milliseconds}${timezoneString}`;
  stream.write(`[${localISOTime}] ${message}
`);
}
function getDebug(topic) {
  const fullTopic = `${topicPrefix}:${topic}`;
  if (!debugInstances.has(fullTopic)) {
    const debugFn = (0, import_debug.default)(fullTopic);
    const wrapper = (...args) => {
      if (ifInNode) {
        const message = import_node_util.default.format(...args);
        writeLogToFile(topic, message);
      }
      debugFn(...args);
    };
    debugInstances.set(fullTopic, wrapper);
  }
  return debugInstances.get(fullTopic);
}

// src/img/get-photon.ts
var photonModule = null;
var isInitialized = false;
async function getPhoton() {
  if (photonModule && isInitialized) {
    return photonModule;
  }
  try {
    if (ifInBrowser || ifInWorker) {
      const photon = await import("@silvia-odwyer/photon");
      if (typeof photon.default === "function") {
        await photon.default();
      }
      photonModule = photon;
    } else if (ifInNode) {
      photonModule = await import("@silvia-odwyer/photon-node");
    }
    if (!photonModule.PhotonImage || !photonModule.PhotonImage.new_from_byteslice) {
      throw new Error("PhotonImage.new_from_byteslice is not available");
    }
    isInitialized = true;
    return photonModule;
  } catch (error) {
    throw new Error(
      `Failed to load photon module: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}

// src/img/get-sharp.ts
var sharpModule = null;
async function getSharp() {
  if (sharpModule) {
    return sharpModule;
  }
  if (!ifInNode) {
    throw new Error("Sharp is only available in Node.js environment");
  }
  try {
    const sharp = await import("sharp");
    sharpModule = sharp.default;
    return sharpModule;
  } catch (error) {
    throw new Error(
      `Failed to load sharp module: ${error instanceof Error ? error.message : String(error)}`
    );
  }
}

// src/img/transform.ts
var imgDebug = getDebug("img");
async function saveBase64Image(options) {
  const { base64Data, outputPath } = options;
  const base64Image = base64Data.split(";base64,").pop() || base64Data;
  const imageBuffer = import_node_buffer2.Buffer.from(base64Image, "base64");
  const Jimp2 = await getJimp();
  const image = await Jimp2.read(imageBuffer);
  await image.writeAsync(outputPath);
}
async function resizeImg(inputData, newSize) {
  if (typeof inputData === "string")
    throw Error("inputData is base64, use resizeImgBase64 instead");
  (0, import_node_assert2.default)(
    newSize && newSize.width > 0 && newSize.height > 0,
    "newSize must be positive"
  );
  const resizeStartTime = Date.now();
  imgDebug(`resizeImg start, target size: ${newSize.width}x${newSize.height}`);
  if (ifInNode) {
    try {
      const Sharp = await getSharp();
      const metadata = await Sharp(inputData).metadata();
      const { width: originalWidth2, height: originalHeight2 } = metadata;
      if (!originalWidth2 || !originalHeight2) {
        throw Error("Undefined width or height from the input image.");
      }
      if (newSize.width === originalWidth2 && newSize.height === originalHeight2) {
        return inputData;
      }
      const resizedBuffer2 = await Sharp(inputData).resize(newSize.width, newSize.height).jpeg({ quality: 90 }).toBuffer();
      const resizeEndTime2 = Date.now();
      imgDebug(
        `resizeImg done (Sharp), target size: ${newSize.width}x${newSize.height}, cost: ${resizeEndTime2 - resizeStartTime}ms`
      );
      return resizedBuffer2;
    } catch (error) {
      imgDebug("Sharp failed, falling back to Photon:", error);
    }
  }
  const { PhotonImage, SamplingFilter, resize } = await getPhoton();
  const inputBytes = new Uint8Array(inputData);
  const inputImage = PhotonImage.new_from_byteslice(inputBytes);
  const originalWidth = inputImage.get_width();
  const originalHeight = inputImage.get_height();
  if (!originalWidth || !originalHeight) {
    inputImage.free();
    throw Error("Undefined width or height from the input image.");
  }
  if (newSize.width === originalWidth && newSize.height === originalHeight) {
    inputImage.free();
    return inputData;
  }
  const outputImage = resize(
    inputImage,
    newSize.width,
    newSize.height,
    SamplingFilter.CatmullRom
  );
  const outputBytes = outputImage.get_bytes_jpeg(90);
  const resizedBuffer = import_node_buffer2.Buffer.from(outputBytes);
  inputImage.free();
  outputImage.free();
  const resizeEndTime = Date.now();
  imgDebug(
    `resizeImg done (Photon), target size: ${newSize.width}x${newSize.height}, cost: ${resizeEndTime - resizeStartTime}ms`
  );
  return resizedBuffer;
}
async function bufferFromBase642(base64) {
  const splitFlag = ";base64,";
  const dataSplitted = base64.split(splitFlag);
  if (dataSplitted.length !== 2) {
    throw Error("Invalid base64 data");
  }
  const res = import_node_buffer2.Buffer.from(dataSplitted[1], "base64");
  return res;
}
async function resizeImgBase64(inputBase64, newSize) {
  const splitFlag = ";base64,";
  const dataSplitted = inputBase64.split(splitFlag);
  if (dataSplitted.length !== 2) {
    throw Error("Invalid base64 data");
  }
  const imageBuffer = import_node_buffer2.Buffer.from(dataSplitted[1], "base64");
  const buffer = await resizeImg(imageBuffer, newSize);
  const content = buffer.toString("base64");
  const res = `${dataSplitted[0]}${splitFlag}${content}`;
  return res;
}
function zoomForGPT4o(originalWidth, originalHeight) {
  const maxWidth = 2048;
  const maxHeight = 768;
  let newWidth = originalWidth;
  let newHeight = originalHeight;
  const aspectRatio = originalWidth / originalHeight;
  if (originalWidth > maxWidth) {
    newWidth = maxWidth;
    newHeight = newWidth / aspectRatio;
  }
  if (newHeight > maxHeight) {
    newHeight = maxHeight;
    newWidth = newHeight * aspectRatio;
  }
  return {
    width: Math.round(newWidth),
    height: Math.round(newHeight)
  };
}
async function jimpFromBase64(base64) {
  const Jimp2 = await getJimp();
  const imageBuffer = await bufferFromBase642(base64);
  return Jimp2.read(imageBuffer);
}
async function paddingToMatchBlock(image, blockSize = 28) {
  const { width, height } = image.bitmap;
  const targetWidth = Math.ceil(width / blockSize) * blockSize;
  const targetHeight = Math.ceil(height / blockSize) * blockSize;
  if (targetWidth === width && targetHeight === height) {
    return image;
  }
  const Jimp2 = await getJimp();
  const paddedImage = new Jimp2(targetWidth, targetHeight, 4294967295);
  paddedImage.composite(image, 0, 0);
  return paddedImage;
}
async function paddingToMatchBlockByBase64(imageBase64, blockSize = 28) {
  const jimpImage = await jimpFromBase64(imageBase64);
  const paddedImage = await paddingToMatchBlock(jimpImage, blockSize);
  return jimpToBase64(paddedImage);
}
async function cropByRect(imageBase64, rect, paddingImage) {
  const jimpImage = await jimpFromBase64(imageBase64);
  const { left, top, width, height } = rect;
  jimpImage.crop(left, top, width, height);
  if (paddingImage) {
    const paddedImage = await paddingToMatchBlock(jimpImage);
    return jimpToBase64(paddedImage);
  }
  return jimpToBase64(jimpImage);
}
async function jimpToBase64(image) {
  const Jimp2 = await getJimp();
  return image.getBase64Async(Jimp2.MIME_JPEG);
}
var httpImg2Base64 = async (url) => {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Failed to fetch image: ${url}`);
  }
  const contentType = response.headers.get("content-type");
  if (!contentType) {
    throw new Error(`Failed to fetch image: ${url}`);
  }
  (0, import_node_assert2.default)(
    contentType.startsWith("image/"),
    `The url ${url} is not a image, because of content-type in header is ${contentType}.`
  );
  const buffer = import_node_buffer2.Buffer.from(await response.arrayBuffer());
  return `data:${contentType};base64,${buffer.toString("base64")}`;
};
var localImg2Base64 = (imgPath, withoutHeader = false) => {
  const body = (0, import_node_fs3.readFileSync)(imgPath).toString("base64");
  if (withoutHeader) {
    return body;
  }
  const type = import_node_path3.default.extname(imgPath).slice(1);
  const finalType = type === "svg" ? "svg+xml" : type || "jpg";
  return `data:image/${finalType};base64,${body}`;
};
var preProcessImageUrl = async (url, convertHttpImage2Base64) => {
  if (url.startsWith("data:")) {
    return url;
  } else if (url.startsWith("http://") || url.startsWith("https://")) {
    if (!convertHttpImage2Base64) {
      return url;
    }
    return await httpImg2Base64(url);
  } else {
    return await localImg2Base64(url);
  }
};

// src/img/box-select.ts
var import_node_assert3 = __toESM(require("assert"));
var cachedFont = null;
var loadFonts = async () => {
  const Jimp2 = await getJimp();
  try {
    const fonts = await Jimp2.loadFont(Jimp2.FONT_SANS_16_WHITE);
    return fonts;
  } catch (error) {
    console.warn("Error loading font, will try to load online fonts", error);
    const onlineFonts = "https://cdn.jsdelivr.net/npm/jimp-compact@0.16.1-2/fonts/open-sans/open-sans-16-white/open-sans-16-white.fnt";
    const fonts = await Jimp2.loadFont(onlineFonts);
    return fonts;
  }
};
var createSvgOverlay = async (elements, imageWidth, imageHeight, boxPadding = 5, borderThickness = 2, prompt) => {
  const Jimp2 = await getJimp();
  const image = new Jimp2(imageWidth, imageHeight, 0);
  const colors = [
    { rect: 3324182783, text: 4294967295 },
    // red, white
    { rect: 65535, text: 4294967295 },
    // blue, white
    { rect: 2336560127, text: 4294967295 },
    // brown, white
    { rect: 1048258559, text: 4294967295 },
    // green, white
    { rect: 1342206975, text: 4294967295 }
    // purple, white
  ];
  if (prompt) {
    try {
      cachedFont = cachedFont || await loadFonts();
      const promptPadding = 10;
      const promptMargin = 20;
      const promptHeight = 30;
      const promptY = imageHeight - promptHeight - promptMargin;
      image.scan(
        0,
        promptY,
        imageWidth,
        promptHeight,
        (x, y, idx) => {
          image.bitmap.data[idx + 0] = 0;
          image.bitmap.data[idx + 1] = 0;
          image.bitmap.data[idx + 2] = 0;
          image.bitmap.data[idx + 3] = 204;
        }
      );
      image.print(
        cachedFont,
        promptPadding,
        promptY,
        {
          text: prompt,
          alignmentX: Jimp2.HORIZONTAL_ALIGN_LEFT,
          alignmentY: Jimp2.VERTICAL_ALIGN_MIDDLE
        },
        imageWidth - promptPadding * 2,
        promptHeight
      );
    } catch (error) {
      console.error("Error drawing prompt text", error);
    }
  }
  for (let index = 0; index < elements.length; index++) {
    const element = elements[index];
    const color = colors[index % colors.length];
    const paddedLeft = Math.max(0, element.rect.left - boxPadding);
    const paddedTop = Math.max(0, element.rect.top - boxPadding);
    const paddedWidth = Math.min(
      imageWidth - paddedLeft,
      element.rect.width + boxPadding * 2
    );
    const paddedHeight = Math.min(
      imageHeight - paddedTop,
      element.rect.height + boxPadding * 2
    );
    const paddedRect = {
      left: paddedLeft,
      top: paddedTop,
      width: paddedWidth,
      height: paddedHeight
    };
    image.scan(
      paddedRect.left,
      paddedRect.top,
      paddedRect.width,
      paddedRect.height,
      (x, y, idx) => {
        if (x >= paddedRect.left && x < paddedRect.left + borderThickness || // Left border
        x <= paddedRect.left + paddedRect.width - 1 && x > paddedRect.left + paddedRect.width - borderThickness || // Right border
        y >= paddedRect.top && y < paddedRect.top + borderThickness || // Top border
        y <= paddedRect.top + paddedRect.height - 1 && y > paddedRect.top + paddedRect.height - borderThickness) {
          image.bitmap.data[idx + 0] = color.rect >> 24 & 255;
          image.bitmap.data[idx + 1] = color.rect >> 16 & 255;
          image.bitmap.data[idx + 2] = color.rect >> 8 & 255;
          image.bitmap.data[idx + 3] = color.rect & 255;
        }
      }
    );
    const indexId = element.indexId;
    if (typeof indexId !== "number") {
      continue;
    }
    const textWidth = indexId.toString().length * 8;
    const textHeight = 12;
    const rectWidth = textWidth + 5;
    const rectHeight = textHeight + 4;
    let rectX = paddedRect.left - rectWidth;
    let rectY = paddedRect.top + paddedRect.height / 2 - textHeight / 2 - 2;
    const checkOverlap = (x, y) => {
      return elements.slice(0, index).some((otherElement) => {
        return x < otherElement.rect.left + otherElement.rect.width && x + rectWidth > otherElement.rect.left && y < otherElement.rect.top + otherElement.rect.height && y + rectHeight > otherElement.rect.top;
      });
    };
    const isWithinBounds = (x, y) => {
      return x >= 0 && x + rectWidth <= imageWidth && y >= 0 && y + rectHeight <= imageHeight;
    };
    if (checkOverlap(rectX, rectY) || !isWithinBounds(rectX, rectY)) {
      if (!checkOverlap(paddedRect.left, paddedRect.top - rectHeight - 2) && isWithinBounds(paddedRect.left, paddedRect.top - rectHeight - 2)) {
        rectX = paddedRect.left;
        rectY = paddedRect.top - rectHeight - 2;
      } else if (!checkOverlap(
        paddedRect.left,
        paddedRect.top + paddedRect.height + 2
      ) && isWithinBounds(paddedRect.left, paddedRect.top + paddedRect.height + 2)) {
        rectX = paddedRect.left;
        rectY = paddedRect.top + paddedRect.height + 2;
      } else if (!checkOverlap(paddedRect.left + paddedRect.width + 2, paddedRect.top) && isWithinBounds(paddedRect.left + paddedRect.width + 2, paddedRect.top)) {
        rectX = paddedRect.left + paddedRect.width + 2;
        rectY = paddedRect.top;
      } else {
        rectX = paddedRect.left;
        rectY = paddedRect.top + 2;
      }
    }
    image.scan(
      rectX,
      rectY,
      rectWidth,
      rectHeight,
      (x, y, idx) => {
        image.bitmap.data[idx + 0] = color.rect >> 24 & 255;
        image.bitmap.data[idx + 1] = color.rect >> 16 & 255;
        image.bitmap.data[idx + 2] = color.rect >> 8 & 255;
        image.bitmap.data[idx + 3] = color.rect & 255;
      }
    );
    try {
      cachedFont = cachedFont || await loadFonts();
    } catch (error) {
      console.error("Error loading font", error);
    }
    image.print(
      cachedFont,
      rectX,
      rectY,
      {
        text: indexId.toString(),
        alignmentX: Jimp2.HORIZONTAL_ALIGN_CENTER,
        alignmentY: Jimp2.VERTICAL_ALIGN_MIDDLE
      },
      rectWidth,
      rectHeight
    );
  }
  return image;
};
var compositeElementInfoImg = async (options) => {
  (0, import_node_assert3.default)(options.inputImgBase64, "inputImgBase64 is required");
  let width = 0;
  let height = 0;
  let jimpImage;
  const Jimp2 = await getJimp();
  if (options.size) {
    width = options.size.width;
    height = options.size.height;
  }
  if (!width || !height) {
    const info = await imageInfoOfBase64(options.inputImgBase64);
    width = info.width;
    height = info.height;
    jimpImage = info.jimpImage;
  } else {
    const imageBuffer = await bufferFromBase64(options.inputImgBase64);
    jimpImage = await Jimp2.read(imageBuffer);
    const imageBitmap = jimpImage.bitmap;
    if (imageBitmap.width !== width || imageBitmap.height !== height) {
      jimpImage.resize(width, height, Jimp2.RESIZE_NEAREST_NEIGHBOR);
    }
  }
  if (!width || !height) {
    throw Error("Image processing failed because width or height is undefined");
  }
  const { elementsPositionInfo, prompt } = options;
  const result = await Promise.resolve(jimpImage).then(async (image) => {
    const svgOverlay = await createSvgOverlay(
      elementsPositionInfo,
      width,
      height,
      options.annotationPadding,
      options.borderThickness,
      prompt
    );
    const svgImage = await Jimp2.read(svgOverlay);
    const compositeImage = await image.composite(svgImage, 0, 0, {
      mode: Jimp2.BLEND_SOURCE_OVER,
      opacitySource: 1,
      opacityDest: 1
    });
    return compositeImage;
  }).then(async (compositeImage) => {
    compositeImage.quality(90);
    const base64 = await compositeImage.getBase64Async(Jimp2.MIME_JPEG);
    return base64;
  }).catch((error) => {
    throw error;
  });
  return result;
};
var processImageElementInfo = async (options) => {
  const base64Image = options.inputImgBase64.split(";base64,").pop();
  (0, import_node_assert3.default)(base64Image, "base64Image is undefined");
  const [
    compositeElementInfoImgBase64,
    compositeElementInfoImgWithoutTextBase64
  ] = await Promise.all([
    compositeElementInfoImg({
      inputImgBase64: options.inputImgBase64,
      elementsPositionInfo: options.elementsPositionInfo
    }),
    compositeElementInfoImg({
      inputImgBase64: options.inputImgBase64,
      elementsPositionInfo: options.elementsPositionInfoWithoutText
    })
  ]);
  return {
    compositeElementInfoImgBase64,
    compositeElementInfoImgWithoutTextBase64
  };
};

// src/img/draw-box.ts
async function drawBoxOnImage(options) {
  const { inputImgBase64, rect } = options;
  const color = { r: 255, g: 0, b: 0, a: 255 };
  const Jimp2 = await getJimp();
  const imageBuffer = await bufferFromBase64(inputImgBase64);
  const image = await Jimp2.read(imageBuffer);
  const centerX = rect.x;
  const centerY = rect.y;
  const radius = 5;
  image.scan(
    Math.floor(centerX - radius),
    Math.floor(centerY - radius),
    radius * 2,
    radius * 2,
    (x, y, idx) => {
      const distance = Math.sqrt((x - centerX) ** 2 + (y - centerY) ** 2);
      if (distance <= radius) {
        image.bitmap.data[idx + 0] = color.r;
        image.bitmap.data[idx + 1] = color.g;
        image.bitmap.data[idx + 2] = color.b;
        image.bitmap.data[idx + 3] = color.a;
      }
    }
  );
  image.quality(90);
  const resultBase64 = await image.getBase64Async(Jimp2.MIME_JPEG);
  return resultBase64;
}
async function savePositionImg(options) {
  const { inputImgBase64, rect, outputPath } = options;
  const imgBase64 = await drawBoxOnImage({ inputImgBase64, rect });
  await saveBase64Image({
    base64Data: imgBase64,
    outputPath
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  bufferFromBase64,
  compositeElementInfoImg,
  cropByRect,
  drawBoxOnImage,
  httpImg2Base64,
  imageInfo,
  imageInfoOfBase64,
  isValidPNGImageBuffer,
  jimpFromBase64,
  jimpToBase64,
  localImg2Base64,
  paddingToMatchBlock,
  paddingToMatchBlockByBase64,
  preProcessImageUrl,
  processImageElementInfo,
  resizeImg,
  resizeImgBase64,
  saveBase64Image,
  savePositionImg,
  zoomForGPT4o
});
