// @midscene/core/dist/es/chunk-I5LBWOQA.js
// src/ai-model/service-caller/index.ts
import { Anthropic } from "@anthropic-ai/sdk";
import {
  DefaultAzureCredential,
  getBearerTokenProvider
} from "@azure/identity";
import {
  ANTHROPIC_API_KEY,
  AZURE_OPENAI_API_VERSION,
  AZURE_OPENAI_DEPLOYMENT,
  AZURE_OPENAI_ENDPOINT,
  AZURE_OPENAI_KEY,
  MIDSCENE_API_TYPE,
  MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON,
  MIDSCENE_AZURE_OPENAI_SCOPE,
  MIDSCENE_DEBUG_AI_PROFILE,
  MIDSCENE_DEBUG_AI_RESPONSE,
  MIDSCENE_LANGSMITH_DEBUG,
  MIDSCENE_MODEL_NAME,
  MIDSCENE_OPENAI_HTTP_PROXY,
  MIDSCENE_OPENAI_INIT_CONFIG_JSON,
  MIDSCENE_OPENAI_SOCKS_PROXY,
  MIDSCENE_USE_ANTHROPIC_SDK,
  MIDSCENE_USE_AZURE_OPENAI,
  OPENAI_API_KEY,
  OPENAI_BASE_URL,
  OPENAI_MAX_TOKENS,
  OPENAI_USE_AZURE,
  getAIConfig,
  getAIConfigInBoolean,
  getAIConfigInJson,
  uiTarsModelVersion,
  vlLocateMode as vlLocateMode3
} from "@midscene/shared/env";
import { enableDebug, getDebug as getDebug2 } from "@midscene/shared/logger";
import { assert as assert3 } from "@midscene/shared/utils";
import { ifInBrowser } from "@midscene/shared/utils";
import { HttpsProxyAgent } from "https-proxy-agent";
import { jsonrepair } from "jsonrepair";
import OpenAI, { AzureOpenAI } from "openai";
import { SocksProxyAgent } from "socks-proxy-agent";

// src/ai-model/common.ts
import { assert } from "@midscene/shared/utils";
import { NodeType } from "@midscene/shared/constants";
import { vlLocateMode } from "@midscene/shared/env";
import { treeToList } from "@midscene/shared/extractor";
import { compositeElementInfoImg } from "@midscene/shared/img";
import { getDebug } from "@midscene/shared/logger";
var AIActionType = /* @__PURE__ */ ((AIActionType2) => {
  AIActionType2[AIActionType2["ASSERT"] = 0] = "ASSERT";
  AIActionType2[AIActionType2["INSPECT_ELEMENT"] = 1] = "INSPECT_ELEMENT";
  AIActionType2[AIActionType2["EXTRACT_DATA"] = 2] = "EXTRACT_DATA";
  AIActionType2[AIActionType2["PLAN"] = 3] = "PLAN";
  AIActionType2[AIActionType2["DESCRIBE_ELEMENT"] = 4] = "DESCRIBE_ELEMENT";
  return AIActionType2;
})(AIActionType || {});
async function callAiFn(msgs, AIActionTypeValue) {
  const { content, usage } = await callToGetJSONObject(
    msgs,
    AIActionTypeValue
  );
  return { content, usage };
}
var defaultBboxSize = 20;
var debugInspectUtils = getDebug("ai:common");
function fillBboxParam(locate, width, height) {
  if (locate.bbox_2d && !locate?.bbox) {
    locate.bbox = locate.bbox_2d;
    delete locate.bbox_2d;
  }
  if (locate?.bbox) {
    locate.bbox = adaptBbox(locate.bbox, width, height);
  }
  return locate;
}
function adaptQwenBbox(bbox) {
  if (bbox.length < 2) {
    const msg = `invalid bbox data for qwen-vl mode: ${JSON.stringify(bbox)} `;
    throw new Error(msg);
  }
  const result = [
    Math.round(bbox[0]),
    Math.round(bbox[1]),
    typeof bbox[2] === "number" ? Math.round(bbox[2]) : Math.round(bbox[0] + defaultBboxSize),
    typeof bbox[3] === "number" ? Math.round(bbox[3]) : Math.round(bbox[1] + defaultBboxSize)
  ];
  return result;
}
function adaptDoubaoBbox(bbox, width, height) {
  assert(
    width > 0 && height > 0,
    "width and height must be greater than 0 in doubao mode"
  );
  if (typeof bbox === "string") {
    assert(
      /^(\d+)\s(\d+)\s(\d+)\s(\d+)$/.test(bbox.trim()),
      `invalid bbox data string for doubao-vision mode: ${bbox}`
    );
    const splitted = bbox.split(" ");
    if (splitted.length === 4) {
      return [
        Math.round(Number(splitted[0]) * width / 1e3),
        Math.round(Number(splitted[1]) * height / 1e3),
        Math.round(Number(splitted[2]) * width / 1e3),
        Math.round(Number(splitted[3]) * height / 1e3)
      ];
    }
    throw new Error(`invalid bbox data string for doubao-vision mode: ${bbox}`);
  }
  if (Array.isArray(bbox) && Array.isArray(bbox[0])) {
    bbox = bbox[0];
  }
  let bboxList = [];
  if (Array.isArray(bbox) && typeof bbox[0] === "string") {
    bbox.forEach((item) => {
      if (typeof item === "string" && item.includes(",")) {
        const [x, y] = item.split(",");
        bboxList.push(Number(x.trim()), Number(y.trim()));
      } else if (typeof item === "string" && item.includes(" ")) {
        const [x, y] = item.split(" ");
        bboxList.push(Number(x.trim()), Number(y.trim()));
      } else {
        bboxList.push(Number(item));
      }
    });
  } else {
    bboxList = bbox;
  }
  if (bboxList.length === 4 || bboxList.length === 5) {
    return [
      Math.round(bboxList[0] * width / 1e3),
      Math.round(bboxList[1] * height / 1e3),
      Math.round(bboxList[2] * width / 1e3),
      Math.round(bboxList[3] * height / 1e3)
    ];
  }
  if (bboxList.length === 6 || bboxList.length === 2 || bboxList.length === 3 || bboxList.length === 7) {
    return [
      Math.max(
        0,
        Math.round(bboxList[0] * width / 1e3) - defaultBboxSize / 2
      ),
      Math.max(
        0,
        Math.round(bboxList[1] * height / 1e3) - defaultBboxSize / 2
      ),
      Math.min(
        width,
        Math.round(bboxList[0] * width / 1e3) + defaultBboxSize / 2
      ),
      Math.min(
        height,
        Math.round(bboxList[1] * height / 1e3) + defaultBboxSize / 2
      )
    ];
  }
  if (bbox.length === 8) {
    return [
      Math.round(bboxList[0] * width / 1e3),
      Math.round(bboxList[1] * height / 1e3),
      Math.round(bboxList[4] * width / 1e3),
      Math.round(bboxList[5] * height / 1e3)
    ];
  }
  const msg = `invalid bbox data for doubao-vision mode: ${JSON.stringify(bbox)} `;
  throw new Error(msg);
}
function adaptBbox(bbox, width, height) {
  if (vlLocateMode() === "doubao-vision" || vlLocateMode() === "vlm-ui-tars") {
    return adaptDoubaoBbox(bbox, width, height);
  }
  if (vlLocateMode() === "gemini") {
    return adaptGeminiBbox(bbox, width, height);
  }
  return adaptQwenBbox(bbox);
}
function adaptGeminiBbox(bbox, width, height) {
  const left = Math.round(bbox[1] * width / 1e3);
  const top = Math.round(bbox[0] * height / 1e3);
  const right = Math.round(bbox[3] * width / 1e3);
  const bottom = Math.round(bbox[2] * height / 1e3);
  return [left, top, right, bottom];
}
function adaptBboxToRect(bbox, width, height, offsetX = 0, offsetY = 0) {
  debugInspectUtils("adaptBboxToRect", bbox, width, height, offsetX, offsetY);
  const [left, top, right, bottom] = adaptBbox(bbox, width, height);
  const rect = {
    left: left + offsetX,
    top: top + offsetY,
    width: right - left,
    height: bottom - top
  };
  debugInspectUtils("adaptBboxToRect, result=", rect);
  return rect;
}
var warned = false;
function warnGPT4oSizeLimit(size) {
  if (warned)
    return;
  if (getModelName()?.toLowerCase().includes("gpt-4o")) {
    const warningMsg = `GPT-4o has a maximum image input size of 2000x768 or 768x2000, but got ${size.width}x${size.height}. Please set your page to a smaller resolution. Otherwise, the result may be inaccurate.`;
    if (Math.max(size.width, size.height) > 2e3 || Math.min(size.width, size.height) > 768) {
      console.warn(warningMsg);
      warned = true;
    }
  } else if (size.width > 1800 || size.height > 1800) {
    console.warn(
      `The image size seems too large (${size.width}x${size.height}). It may lead to more token usage, slower response, and inaccurate result.`
    );
    warned = true;
  }
}
function mergeRects(rects) {
  const minLeft = Math.min(...rects.map((r) => r.left));
  const minTop = Math.min(...rects.map((r) => r.top));
  const maxRight = Math.max(...rects.map((r) => r.left + r.width));
  const maxBottom = Math.max(...rects.map((r) => r.top + r.height));
  return {
    left: minLeft,
    top: minTop,
    width: maxRight - minLeft,
    height: maxBottom - minTop
  };
}
function expandSearchArea(rect, screenSize) {
  const minEdgeSize = vlLocateMode() === "doubao-vision" ? 500 : 300;
  const defaultPadding = 160;
  const paddingSizeHorizontal = rect.width < minEdgeSize ? Math.ceil((minEdgeSize - rect.width) / 2) : defaultPadding;
  const paddingSizeVertical = rect.height < minEdgeSize ? Math.ceil((minEdgeSize - rect.height) / 2) : defaultPadding;
  rect.left = Math.max(0, rect.left - paddingSizeHorizontal);
  rect.width = Math.min(
    rect.width + paddingSizeHorizontal * 2,
    screenSize.width - rect.left
  );
  rect.top = Math.max(0, rect.top - paddingSizeVertical);
  rect.height = Math.min(
    rect.height + paddingSizeVertical * 2,
    screenSize.height - rect.top
  );
  return rect;
}
async function markupImageForLLM(screenshotBase64, tree, size) {
  const elementsInfo = treeToList(tree);
  const elementsPositionInfoWithoutText = elementsInfo.filter(
    (elementInfo) => {
      if (elementInfo.attributes.nodeType === NodeType.TEXT) {
        return false;
      }
      return true;
    }
  );
  const imagePayload = await compositeElementInfoImg({
    inputImgBase64: screenshotBase64,
    elementsPositionInfo: elementsPositionInfoWithoutText,
    size
  });
  return imagePayload;
}
function buildYamlFlowFromPlans(plans, sleep) {
  const flow = [];
  for (const plan2 of plans) {
    const type = plan2.type;
    const locate = plan2.locate?.prompt;
    if (type === "Tap") {
      flow.push({
        aiTap: locate
      });
    } else if (type === "Hover") {
      flow.push({
        aiHover: locate
      });
    } else if (type === "Input") {
      const param = plan2.param;
      flow.push({
        aiInput: param.value,
        locate
      });
    } else if (type === "KeyboardPress") {
      const param = plan2.param;
      flow.push({
        aiKeyboardPress: param.value,
        locate
      });
    } else if (type === "Scroll") {
      const param = plan2.param;
      flow.push({
        aiScroll: null,
        locate,
        direction: param.direction,
        scrollType: param.scrollType,
        distance: param.distance
      });
    } else if (type === "Sleep") {
      const param = plan2.param;
      flow.push({
        sleep: param.timeMs
      });
    } else if (type === "AndroidBackButton" || type === "AndroidHomeButton" || type === "AndroidRecentAppsButton" || type === "AndroidLongPress" || type === "AndroidPull") {
    } else if (type === "Error" || type === "ExpectedFalsyCondition" || type === "Assert" || type === "AssertWithoutThrow" || type === "Finished") {
    } else {
      console.warn(
        `Cannot convert action ${type} to yaml flow. This should be a bug of Midscene.`
      );
    }
  }
  if (sleep) {
    flow.push({
      sleep
    });
  }
  return flow;
}

// src/ai-model/prompt/assertion.ts
import { getPreferredLanguage } from "@midscene/shared/env";
var defaultAssertionPrompt = "You are a senior testing engineer. User will give an assertion and a screenshot of a page. By carefully viewing the screenshot, please tell whether the assertion is truthy.";
var defaultAssertionResponseJsonFormat = `Return in the following JSON format:
{
  pass: boolean, // whether the assertion is truthy
  thought: string | null, // string, if the result is falsy, give the reason why it is falsy. Otherwise, put null.
}`;
var getUiTarsAssertionResponseJsonFormat = () => `## Output Json String Format
\`\`\`
"{
  "pass": <<is a boolean value from the enum [true, false], true means the assertion is truthy>>, 
  "thought": "<<is a string, give the reason why the assertion is falsy or truthy. Otherwise.>>"
}"
\`\`\`

## Rules **MUST** follow
- Make sure to return **only** the JSON, with **no additional** text or explanations.
- Use ${getPreferredLanguage()} in \`thought\` part.
- You **MUST** strictly follow up the **Output Json String Format**.`;
function systemPromptToAssert(model) {
  return `${defaultAssertionPrompt}

${model.isUITars ? getUiTarsAssertionResponseJsonFormat() : defaultAssertionResponseJsonFormat}`;
}
var assertSchema = {
  type: "json_schema",
  json_schema: {
    name: "assert",
    strict: true,
    schema: {
      type: "object",
      properties: {
        pass: {
          type: "boolean",
          description: "Whether the assertion passed or failed"
        },
        thought: {
          type: ["string", "null"],
          description: "The thought process behind the assertion"
        }
      },
      required: ["pass", "thought"],
      additionalProperties: false
    }
  }
};

// src/ai-model/prompt/llm-locator.ts
import { PromptTemplate } from "@langchain/core/prompts";

// src/ai-model/prompt/common.ts
function bboxDescription(vlMode) {
  if (vlMode === "gemini") {
    return "2d bounding box as [ymin, xmin, ymax, xmax]";
  }
  return "2d bounding box as [xmin, ymin, xmax, ymax]";
}

// src/ai-model/prompt/llm-locator.ts
function systemPromptToLocateElement(vlMode) {
  if (vlMode) {
    const bboxComment = bboxDescription(vlMode);
    return `
## Role:
You are an expert in software testing.

## Objective:
- Identify elements in screenshots and text that match the user's description.
- Give the coordinates of the element that matches the user's description best in the screenshot.
- Determine whether the user's description is order-sensitive (e.g., contains phrases like 'the third item in the list', 'the last button', etc.).

## Output Format:
\`\`\`json
{
  "bbox": [number, number, number, number],  // ${bboxComment}
  "errors"?: string[],
  "isOrderSensitive": boolean // Whether the targetElementDescription is order-sensitive (true/false)
}
\`\`\`

Fields:
* \`bbox\` is the bounding box of the element that matches the user's description best in the screenshot
* \`isOrderSensitive\` is a boolean indicating whether the user's description is order-sensitive (true/false)
* \`errors\` is an optional array of error messages (if any)

Order-sensitive means the description contains phrases like:
- "the third item in the list"
- "the last button"
- "the first input box"
- "the second row"

Not order-sensitive means the description is like:
- "confirm button"
- "search box"
- "password input"

For example, when an element is found and the description is order-sensitive:
\`\`\`json
{
  "bbox": [100, 100, 200, 200],
  "isOrderSensitive": true,
  "errors": []
}
\`\`\`

When no element is found and the description is not order-sensitive:
\`\`\`json
{
  "bbox": [],
  "isOrderSensitive": false,
  "errors": ["I can see ..., but {some element} is not found"]
}
\`\`\`
`;
  }
  return `
## Role:
You are an expert in software page image (2D) and page element text analysis.

## Objective:
- Identify elements in screenshots and text that match the user's description.
- Return JSON data containing the selection reason and element ID.
- Determine whether the user's description is order-sensitive (e.g., contains phrases like 'the third item in the list', 'the last button', etc.).

## Skills:
- Image analysis and recognition
- Multilingual text understanding
- Software UI design and testing

## Workflow:
1. Receive the user's element description, screenshot, and element description information. Note that the text may contain non-English characters (e.g., Chinese), indicating that the application may be non-English.
2. Based on the user's description, locate the target element ID in the list of element descriptions and the screenshot.
3. Found the required number of elements
4. Return JSON data containing the selection reason and element ID.
5. Judge whether the user's description is order-sensitive (see below for definition and examples).

## Constraints:
- Strictly adhere to the specified location when describing the required element; do not select elements from other locations.
- Elements in the image with NodeType other than "TEXT Node" have been highlighted to identify the element among multiple non-text elements.
- Accurately identify element information based on the user's description and return the corresponding element ID from the element description information, not extracted from the image.
- If no elements are found, the "elements" array should be empty.
- The returned data must conform to the specified JSON format.
- The returned value id information must use the id from element info (important: **use id not indexId, id is hash content**)

## Order-Sensitive Definition:
- If the description contains phrases like "the third item in the list", "the last button", "the first input box", "the second row", etc., it is order-sensitive (isOrderSensitive = true).
- If the description is like "confirm button", "search box", "password input", etc., it is not order-sensitive (isOrderSensitive = false).

## Output Format:

Please return the result in JSON format as follows:

\`\`\`json
{
  "elements": [
    // If no matching elements are found, return an empty array []
    {
      "reason": "PLACEHOLDER", // The thought process for finding the element, replace PLACEHOLDER with your thought process
      "text": "PLACEHOLDER", // Replace PLACEHOLDER with the text of elementInfo, if none, leave empty
      "id": "PLACEHOLDER" // Replace PLACEHOLDER with the ID (important: **use id not indexId, id is hash content**) of elementInfo
    }
    // More elements...
  ],
  "isOrderSensitive": true, // or false, depending on the user's description
  "errors": [] // Array of strings containing any error messages
}
\`\`\`

## Example:
Example 1:
Input Example:
\`\`\`json
// Description: "Shopping cart icon in the upper right corner"
{
  "description": "PLACEHOLDER", // Description of the target element
  "screenshot": "path/screenshot.png",
  "text": '{
      "pageSize": {
        "width": 400, // Width of the page
        "height": 905 // Height of the page
      },
      "elementInfos": [
        {
          "id": "1231", // ID of the element
          "indexId": "0", // Index of the element，The image is labeled to the left of the element
          "attributes": { // Attributes of the element
            "nodeType": "IMG Node", // Type of element, types include: TEXT Node, IMG Node, BUTTON Node, INPUT Node
            "src": "https://ap-southeast-3.m",
            "class": ".img"
          },
          "content": "", // Text content of the element
          "rect": {
            "left": 280, // Distance from the left side of the page
            "top": 8, // Distance from the top of the page
            "width": 44, // Width of the element
            "height": 44 // Height of the element
          }
        },
        {
          "id": "66551", // ID of the element
          "indexId": "1", // Index of the element,The image is labeled to the left of the element
          "attributes": { // Attributes of the element
            "nodeType": "IMG Node", // Type of element, types include: TEXT Node, IMG Node, BUTTON Node, INPUT Node
            "src": "data:image/png;base64,iVBORw0KGgoAAAANSU...",
            "class": ".icon"
          },
          "content": "", // Text content of the element
          "rect": {
            "left": 350, // Distance from the left side of the page
            "top": 16, // Distance from the top of the page
            "width": 25, // Width of the element
            "height": 25 // Height of the element
          }
        },
        ...
        {
          "id": "12344",
          "indexId": "2", // Index of the element，The image is labeled to the left of the element
          "attributes": {
            "nodeType": "TEXT Node",
            "class": ".product-name"
          },
          "center": [
            288,
            834
          ],
          "content": "Mango Drink",
          "rect": {
            "left": 188,
            "top": 827,
            "width": 199,
            "height": 13
          }
        },
        ...
      ]
    }
  '
}
\`\`\`
Output Example:
\`\`\`json
{
  "elements": [
    {
      // Describe the reason for finding this element, replace with actual value in practice
      "reason": "Reason for finding element 4: It is located in the upper right corner, is an image type, and according to the screenshot, it is a shopping cart icon button",
      "text": "",
      // ID(**use id not indexId**) of this element, replace with actual value in practice, **use id not indexId**
      "id": "1231"
    }
  ],
  "isOrderSensitive": true,
  "errors": []
}
\`\`\`
  
  `;
}
var locatorSchema = {
  type: "json_schema",
  json_schema: {
    name: "find_elements",
    strict: true,
    schema: {
      type: "object",
      properties: {
        elements: {
          type: "array",
          items: {
            type: "object",
            properties: {
              reason: {
                type: "string",
                description: "Reason for finding this element"
              },
              text: {
                type: "string",
                description: "Text content of the element"
              },
              id: {
                type: "string",
                description: "ID of this element"
              }
            },
            required: ["reason", "text", "id"],
            additionalProperties: false
          },
          description: "List of found elements"
        },
        isOrderSensitive: {
          type: "boolean",
          description: "Whether the targetElementDescription is order-sensitive (true/false)"
        },
        errors: {
          type: "array",
          items: {
            type: "string"
          },
          description: "List of error messages, if any"
        }
      },
      required: ["elements", "isOrderSensitive", "errors"],
      additionalProperties: false
    }
  }
};
var findElementPrompt = new PromptTemplate({
  template: `
Here is the item user want to find:
=====================================
{targetElementDescription}
=====================================

{pageDescription}
  `,
  inputVariables: ["pageDescription", "targetElementDescription"]
});

// src/ai-model/prompt/llm-planning.ts
import { PromptTemplate as PromptTemplate2 } from "@langchain/core/prompts";

// src/image/index.ts
import {
  imageInfo,
  imageInfoOfBase64,
  localImg2Base64,
  httpImg2Base64,
  resizeImg,
  saveBase64Image,
  zoomForGPT4o
} from "@midscene/shared/img";

// src/ai-model/prompt/util.ts
import { NodeType as NodeType2 } from "@midscene/shared/constants";
import { vlLocateMode as vlLocateMode2 } from "@midscene/shared/env";
import {
  descriptionOfTree,
  generateElementByPosition,
  treeToList as treeToList2
} from "@midscene/shared/extractor";
import { assert as assert2 } from "@midscene/shared/utils";
function describeSize(size) {
  return `${size.width} x ${size.height}`;
}
var distanceThreshold = 16;
function elementByPositionWithElementInfo(treeRoot, position, options) {
  const requireStrictDistance = options?.requireStrictDistance ?? true;
  const filterPositionElements = options?.filterPositionElements ?? false;
  assert2(typeof position !== "undefined", "position is required for query");
  const matchingElements = [];
  function dfs(node) {
    if (node?.node) {
      const item = node.node;
      if (item.rect.left <= position.x && position.x <= item.rect.left + item.rect.width && item.rect.top <= position.y && position.y <= item.rect.top + item.rect.height) {
        if (!(filterPositionElements && item.attributes?.nodeType === NodeType2.POSITION) && item.isVisible) {
          matchingElements.push(item);
        }
      }
    }
    for (const child of node.children) {
      dfs(child);
    }
  }
  dfs(treeRoot);
  if (matchingElements.length === 0) {
    return void 0;
  }
  const element = matchingElements.reduce((smallest, current) => {
    const smallestArea = smallest.rect.width * smallest.rect.height;
    const currentArea = current.rect.width * current.rect.height;
    return currentArea < smallestArea ? current : smallest;
  });
  const distanceToCenter = distance(
    { x: element.center[0], y: element.center[1] },
    position
  );
  if (requireStrictDistance) {
    return distanceToCenter <= distanceThreshold ? element : void 0;
  }
  return element;
}
function distance(point1, point2) {
  return Math.sqrt((point1.x - point2.x) ** 2 + (point1.y - point2.y) ** 2);
}
var samplePageDescription = `
And the page is described as follows:
====================
The size of the page: 1280 x 720
Some of the elements are marked with a rectangle in the screenshot corresponding to the markerId, some are not.

Description of all the elements in screenshot:
<div id="969f1637" markerId="1" left="100" top="100" width="100" height="100"> // The markerId indicated by the rectangle label in the screenshot
  <h4 id="b211ecb2" markerId="5" left="150" top="150" width="90" height="60">
    The username is accepted
  </h4>
  ...many more
</div>
====================
`;
async function describeUserPage(context, opt) {
  const { screenshotBase64 } = context;
  let width;
  let height;
  if (context.size) {
    ({ width, height } = context.size);
  } else {
    const imgSize = await imageInfoOfBase64(screenshotBase64);
    ({ width, height } = imgSize);
  }
  const treeRoot = context.tree;
  const idElementMap = {};
  const flatElements = treeToList2(treeRoot);
  if (opt?.domIncluded === true && flatElements.length >= 5e3) {
    console.warn(
      'The number of elements is too large, it may cause the prompt to be too long, please use domIncluded: "visible-only" to reduce the number of elements'
    );
  }
  flatElements.forEach((element) => {
    idElementMap[element.id] = element;
    if (typeof element.indexId !== "undefined") {
      idElementMap[`${element.indexId}`] = element;
    }
  });
  let pageDescription = "";
  const visibleOnly = opt?.visibleOnly ?? opt?.domIncluded === "visible-only";
  if (opt?.domIncluded || !vlLocateMode2()) {
    const contentTree = await descriptionOfTree(
      treeRoot,
      opt?.truncateTextLength,
      opt?.filterNonTextContent,
      visibleOnly
    );
    const sizeDescription = describeSize({ width, height });
    pageDescription = `The size of the page: ${sizeDescription} 
 The page elements tree:
${contentTree}`;
  }
  return {
    description: pageDescription,
    elementById(idOrIndexId) {
      assert2(typeof idOrIndexId !== "undefined", "id is required for query");
      const item = idElementMap[`${idOrIndexId}`];
      return item;
    },
    elementByPosition(position, size) {
      return elementByPositionWithElementInfo(treeRoot, position);
    },
    insertElementByPosition(position) {
      const element = generateElementByPosition(position);
      treeRoot.children.push({
        node: element,
        children: []
      });
      flatElements.push(element);
      idElementMap[element.id] = element;
      return element;
    },
    size: { width, height }
  };
}

// src/ai-model/prompt/llm-planning.ts
var vlCoTLog = `"what_the_user_wants_to_do_next_by_instruction": string, // What the user wants to do according to the instruction and previous logs. `;
var vlCurrentLog = `"log": string, // Log what the next one action (ONLY ONE!) you can do according to the screenshot and the instruction. The typical log looks like "Now i want to use action '{{ action-type }}' to do .. first". If no action should be done, log the reason. ". Use the same language as the user's instruction.`;
var llmCurrentLog = `"log": string, // Log what the next actions you can do according to the screenshot and the instruction. The typical log looks like "Now i want to use action '{{ action-type }}' to do ..". If no action should be done, log the reason. ". Use the same language as the user's instruction.`;
var commonOutputFields = `"error"?: string, // Error messages about unexpected situations, if any. Only think it is an error when the situation is not expected according to the instruction. Use the same language as the user's instruction.
  "more_actions_needed_by_instruction": boolean, // Consider if there is still more action(s) to do after the action in "Log" is done, according to the instruction. If so, set this field to true. Otherwise, set it to false.`;
var vlLocateParam = "locate: {bbox: [number, number, number, number], prompt: string }";
var systemTemplateOfVLPlanning = ({
  pageType,
  vlMode
}) => `
Target: User will give you a screenshot, an instruction and some previous logs indicating what have been done. Please tell what the next one action is (or null if no action should be done) to do the tasks the instruction requires. 

Restriction:
- Don't give extra actions or plans beyond the instruction. ONLY plan for what the instruction requires. For example, don't try to submit the form if the instruction is only to fill something.
- Always give ONLY ONE action in \`log\` field (or null if no action should be done), instead of multiple actions. Supported actions are Tap, Hover, Input, KeyboardPress, Scroll${pageType === "android" ? ", AndroidBackButton, AndroidHomeButton, AndroidRecentAppsButton, AndroidLongPress, AndroidPull." : "."}
- Don't repeat actions in the previous logs.
- Bbox is the bounding box of the element to be located. It's an array of 4 numbers, representing ${bboxDescription(vlMode)}.

Supporting actions:
- Tap: { type: "Tap", ${vlLocateParam} }
- RightClick: { type: "RightClick", ${vlLocateParam} }
- Hover: { type: "Hover", ${vlLocateParam} }
- Input: { type: "Input", ${vlLocateParam}, param: { value: string } } // Replace the input field with a new value. \`value\` is the final that should be filled in the input box. No matter what modifications are required, just provide the final value to replace the existing input value. Giving a blank string means clear the input field.
- KeyboardPress: { type: "KeyboardPress", param: { value: string } }
- Scroll: { type: "Scroll", ${vlLocateParam} | null, param: { direction: 'down'(default) | 'up' | 'right' | 'left', scrollType: 'once' (default) | 'untilBottom' | 'untilTop' | 'untilRight' | 'untilLeft', distance: null | number }} // locate is the element to scroll. If it's a page scroll, put \`null\` in the \`locate\` field.
${pageType === "android" ? `- AndroidBackButton: { type: "AndroidBackButton", param: {} }
- AndroidHomeButton: { type: "AndroidHomeButton", param: {} }
- AndroidRecentAppsButton: { type: "AndroidRecentAppsButton", param: {} }
- AndroidLongPress: { type: "AndroidLongPress", param: { x: number, y: number, duration?: number } }
- AndroidPull: { type: "AndroidPull", param: { direction: 'up' | 'down', startPoint?: { x: number, y: number }, distance?: number, duration?: number } } // Pull down to refresh (direction: 'down') or pull up to load more (direction: 'up')` : ""}

Field description:
* The \`prompt\` field inside the \`locate\` field is a short description that could be used to locate the element.

Return in JSON format:
{
  ${vlCoTLog}
  ${vlCurrentLog}
  ${commonOutputFields}
  "action": 
    {
      // one of the supporting actions
    } | null,
  ,
  "sleep"?: number, // The sleep time after the action, in milliseconds.
}

For example, when the instruction is "click 'Confirm' button, and click 'Yes' in popup" and the log is "I will use action Tap to click 'Confirm' button", by viewing the screenshot and previous logs, you should consider: We have already clicked the 'Confirm' button, so next we should find and click 'Yes' in popup.

this and output the JSON:

{
  "what_the_user_wants_to_do_next_by_instruction": "We have already clicked the 'Confirm' button, so next we should find and click 'Yes' in popup",
  "log": "I will use action Tap to click 'Yes' in popup",
  "more_actions_needed_by_instruction": false,
  "action": {
    "type": "Tap",
    "locate": {
      "bbox": [100, 100, 200, 200],
      "prompt": "The 'Yes' button in popup"
    }
  }
}
`;
var llmLocateParam = `locate: {{"id": string, "prompt": string}} | null`;
var systemTemplateOfLLM = ({ pageType }) => `
## Role

You are a versatile professional in software UI automation. Your outstanding contributions will impact the user experience of billions of users.

## Objective

- Decompose the instruction user asked into a series of actions
- Locate the target element if possible
- If the instruction cannot be accomplished, give a further plan.

## Workflow

1. Receive the screenshot, element description of screenshot(if any), user's instruction and previous logs.
2. Decompose the user's task into a sequence of actions, and place it in the \`actions\` field. There are different types of actions (Tap / Hover / Input / KeyboardPress / Scroll / FalsyConditionStatement / Sleep ${pageType === "android" ? "/ AndroidBackButton / AndroidHomeButton / AndroidRecentAppsButton / AndroidLongPress / AndroidPull" : ""}). The "About the action" section below will give you more details.
3. Precisely locate the target element if it's already shown in the screenshot, put the location info in the \`locate\` field of the action.
4. If some target elements is not shown in the screenshot, consider the user's instruction is not feasible on this page. Follow the next steps.
5. Consider whether the user's instruction will be accomplished after all the actions
 - If yes, set \`taskWillBeAccomplished\` to true
 - If no, don't plan more actions by closing the array. Get ready to reevaluate the task. Some talent people like you will handle this. Give him a clear description of what have been done and what to do next. Put your new plan in the \`furtherPlan\` field. The "How to compose the \`taskWillBeAccomplished\` and \`furtherPlan\` fields" section will give you more details.

## Constraints

- All the actions you composed MUST be based on the page context information you get.
- Trust the "What have been done" field about the task (if any), don't repeat actions in it.
- Respond only with valid JSON. Do not write an introduction or summary or markdown prefix like \`\`\`json\`\`\`.
- If the screenshot and the instruction are totally irrelevant, set reason in the \`error\` field.

## About the \`actions\` field

The \`locate\` param is commonly used in the \`param\` field of the action, means to locate the target element to perform the action, it conforms to the following scheme:

type LocateParam = {{
  "id": string, // the id of the element found. It should either be the id marked with a rectangle in the screenshot or the id described in the description.
  "prompt"?: string // the description of the element to find. It can only be omitted when locate is null.
}} | null // If it's not on the page, the LocateParam should be null

## Supported actions

Each action has a \`type\` and corresponding \`param\`. To be detailed:
- type: 'Tap'
  * {{ ${llmLocateParam} }}
- type: 'RightClick'
  * {{ ${llmLocateParam} }}
- type: 'Hover'
  * {{ ${llmLocateParam} }}
- type: 'Input', replace the value in the input field
  * {{ ${llmLocateParam}, param: {{ value: string }} }}
  * \`value\` is the final value that should be filled in the input field. No matter what modifications are required, just provide the final value user should see after the action is done. 
- type: 'KeyboardPress', press a key
  * {{ param: {{ value: string }} }}
- type: 'Scroll', scroll up or down.
  * {{ 
      ${llmLocateParam}, 
      param: {{ 
        direction: 'down'(default) | 'up' | 'right' | 'left', 
        scrollType: 'once' (default) | 'untilBottom' | 'untilTop' | 'untilRight' | 'untilLeft', 
        distance: null | number 
      }} 
    }}
    * To scroll some specific element, put the element at the center of the region in the \`locate\` field. If it's a page scroll, put \`null\` in the \`locate\` field. 
    * \`param\` is required in this action. If some fields are not specified, use direction \`down\`, \`once\` scroll type, and \`null\` distance.
  * {{ param: {{ button: 'Back' | 'Home' | 'RecentApp' }} }}
- type: 'ExpectedFalsyCondition'
  * {{ param: {{ reason: string }} }}
  * use this action when the conditional statement talked about in the instruction is falsy.
- type: 'Sleep'
  * {{ param: {{ timeMs: number }} }}
${pageType === "android" ? `- type: 'AndroidBackButton', trigger the system "back" operation on Android devices
  * {{ param: {{}} }}
- type: 'AndroidHomeButton', trigger the system "home" operation on Android devices
  * {{ param: {{}} }}
- type: 'AndroidRecentAppsButton', trigger the system "recent apps" operation on Android devices
  * {{ param: {{}} }}
- type: 'AndroidLongPress', trigger a long press on the screen at specified coordinates on Android devices
  * {{ param: {{ x: number, y: number, duration?: number }} }}
- type: 'AndroidPull', trigger pull down to refresh or pull up actions on Android devices
  * {{ param: {{ direction: 'up' | 'down', startPoint?: {{ x: number, y: number }}, distance?: number, duration?: number }} }}` : ""}
`;
var outputTemplate = `
## Output JSON Format:

The JSON format is as follows:

{{
  "actions": [
    // ... some actions
  ],
  ${llmCurrentLog}
  ${commonOutputFields}
}}

## Examples

### Example: Decompose a task

When the instruction is 'Click the language switch button, wait 1s, click "English"', and not log is provided

By viewing the page screenshot and description, you should consider this and output the JSON:

* The main steps should be: tap the switch button, sleep, and tap the 'English' option
* The language switch button is shown in the screenshot, but it's not marked with a rectangle. So we have to use the page description to find the element. By carefully checking the context information (coordinates, attributes, content, etc.), you can find the element.
* The "English" option button is not shown in the screenshot now, it means it may only show after the previous actions are finished. So don't plan any action to do this.
* Log what these action do: Click the language switch button to open the language options. Wait for 1 second.
* The task cannot be accomplished (because we cannot see the "English" option now), so the \`more_actions_needed_by_instruction\` field is true.

{{
  "actions":[
    {{
      "type": "Tap", 
      "thought": "Click the language switch button to open the language options.",
      "param": null,
      "locate": {{ id: "c81c4e9a33", prompt: "The language switch button" }},
    }},
    {{
      "type": "Sleep",
      "thought": "Wait for 1 second to ensure the language options are displayed.",
      "param": {{ "timeMs": 1000 }},
    }}
  ],
  "error": null,
  "more_actions_needed_by_instruction": true,
  "log": "Click the language switch button to open the language options. Wait for 1 second",
}}

### Example: What NOT to do
Wrong output:
{{
  "actions":[
    {{
      "type": "Tap",
      "thought": "Click the language switch button to open the language options.",
      "param": null,
      "locate": {{
        {{ "id": "c81c4e9a33" }}, // WRONG: prompt is missing
      }}
    }},
    {{
      "type": "Tap", 
      "thought": "Click the English option",
      "param": null,
      "locate": null, // This means the 'English' option is not shown in the screenshot, the task cannot be accomplished
    }}
  ],
  "more_actions_needed_by_instruction": false, // WRONG: should be true
  "log": "Click the language switch button to open the language options",
}}

Reason:
* The \`prompt\` is missing in the first 'Locate' action
* Since the option button is not shown in the screenshot, there are still more actions to be done, so the \`more_actions_needed_by_instruction\` field should be true
`;
async function systemPromptToTaskPlanning({
  pageType,
  vlMode
}) {
  if (vlMode) {
    return systemTemplateOfVLPlanning({ pageType, vlMode });
  }
  const promptTemplate = new PromptTemplate2({
    template: `${systemTemplateOfLLM({ pageType })}

${outputTemplate}`,
    inputVariables: ["pageDescription"]
  });
  return await promptTemplate.format({
    pageDescription: samplePageDescription
  });
}
var planSchema = {
  type: "json_schema",
  json_schema: {
    name: "action_items",
    strict: true,
    schema: {
      type: "object",
      strict: true,
      properties: {
        actions: {
          //  TODO
          type: "array",
          items: {
            type: "object",
            strict: true,
            properties: {
              thought: {
                type: "string",
                description: "Reasons for generating this task, and why this task is feasible on this page"
              },
              type: {
                type: "string",
                description: 'Type of action, one of "Tap", "RightClick", "Hover" , "Input", "KeyboardPress", "Scroll", "ExpectedFalsyCondition", "Sleep", "AndroidBackButton", "AndroidHomeButton", "AndroidRecentAppsButton", "AndroidLongPress"'
              },
              param: {
                anyOf: [
                  { type: "null" },
                  {
                    type: "object",
                    properties: { value: { type: ["string", "number"] } },
                    required: ["value"],
                    additionalProperties: false
                  },
                  {
                    type: "object",
                    properties: { timeMs: { type: ["number", "string"] } },
                    required: ["timeMs"],
                    additionalProperties: false
                  },
                  {
                    type: "object",
                    properties: {
                      direction: { type: "string" },
                      scrollType: { type: "string" },
                      distance: { type: ["number", "string", "null"] }
                    },
                    required: ["direction", "scrollType", "distance"],
                    additionalProperties: false
                  },
                  {
                    type: "object",
                    properties: { reason: { type: "string" } },
                    required: ["reason"],
                    additionalProperties: false
                  },
                  {
                    type: "object",
                    properties: { button: { type: "string" } },
                    required: ["button"],
                    additionalProperties: false
                  }
                ],
                description: "Parameter of the action, can be null ONLY when the type field is Tap or Hover"
              },
              locate: {
                type: ["object", "null"],
                properties: {
                  id: { type: "string" },
                  prompt: { type: "string" }
                },
                required: ["id", "prompt"],
                additionalProperties: false,
                description: "Location information for the target element"
              }
            },
            required: ["thought", "type", "param", "locate"],
            additionalProperties: false
          },
          description: "List of actions to be performed"
        },
        more_actions_needed_by_instruction: {
          type: "boolean",
          description: "If all the actions described in the instruction have been covered by this action and logs, set this field to false."
        },
        log: {
          type: "string",
          description: "Log what these planned actions do. Do not include further actions that have not been planned."
        },
        error: {
          type: ["string", "null"],
          description: "Error messages about unexpected situations"
        }
      },
      required: [
        "actions",
        "more_actions_needed_by_instruction",
        "log",
        "error"
      ],
      additionalProperties: false
    }
  }
};
var generateTaskBackgroundContext = (userInstruction, log, userActionContext) => {
  if (log) {
    return `
Here is the user's instruction:

<instruction>
  <high_priority_knowledge>
    ${userActionContext}
  </high_priority_knowledge>

  ${userInstruction}
</instruction>

These are the logs from previous executions, which indicate what was done in the previous actions.
Do NOT repeat these actions.
<previous_logs>
${log}
</previous_logs>
`;
  }
  return `
Here is the user's instruction:
<instruction>
  <high_priority_knowledge>
    ${userActionContext}
  </high_priority_knowledge>

  ${userInstruction}
</instruction>
`;
};
var automationUserPrompt = (vlMode) => {
  if (vlMode) {
    return new PromptTemplate2({
      template: "{taskBackgroundContext}",
      inputVariables: ["taskBackgroundContext"]
    });
  }
  return new PromptTemplate2({
    template: `
pageDescription:
=====================================
{pageDescription}
=====================================

{taskBackgroundContext}`,
    inputVariables: ["pageDescription", "taskBackgroundContext"]
  });
};

// src/ai-model/service-caller/index.ts
function checkAIConfig() {
  const openaiKey = getAIConfig(OPENAI_API_KEY);
  const azureConfig = getAIConfig(MIDSCENE_USE_AZURE_OPENAI);
  const anthropicKey = getAIConfig(ANTHROPIC_API_KEY);
  const initConfigJson = getAIConfig(MIDSCENE_OPENAI_INIT_CONFIG_JSON);
  if (openaiKey)
    return true;
  if (azureConfig)
    return true;
  if (anthropicKey)
    return true;
  return Boolean(initConfigJson);
}
var debugConfigInitialized = false;
function initDebugConfig() {
  if (debugConfigInitialized)
    return;
  const shouldPrintTiming = getAIConfigInBoolean(MIDSCENE_DEBUG_AI_PROFILE);
  let debugConfig = "";
  if (shouldPrintTiming) {
    console.warn(
      "MIDSCENE_DEBUG_AI_PROFILE is deprecated, use DEBUG=midscene:ai:profile instead"
    );
    debugConfig = "ai:profile";
  }
  const shouldPrintAIResponse = getAIConfigInBoolean(
    MIDSCENE_DEBUG_AI_RESPONSE
  );
  if (shouldPrintAIResponse) {
    console.warn(
      "MIDSCENE_DEBUG_AI_RESPONSE is deprecated, use DEBUG=midscene:ai:response instead"
    );
    if (debugConfig) {
      debugConfig = "ai:*";
    } else {
      debugConfig = "ai:call";
    }
  }
  if (debugConfig) {
    enableDebug(debugConfig);
  }
  debugConfigInitialized = true;
}
var defaultModel = "gpt-4o";
function getModelName() {
  let modelName = defaultModel;
  const nameInConfig = getAIConfig(MIDSCENE_MODEL_NAME);
  if (nameInConfig) {
    modelName = nameInConfig;
  }
  return modelName;
}
async function createChatClient({
  AIActionTypeValue
}) {
  initDebugConfig();
  let openai;
  const extraConfig = getAIConfigInJson(MIDSCENE_OPENAI_INIT_CONFIG_JSON);
  const socksProxy = getAIConfig(MIDSCENE_OPENAI_SOCKS_PROXY);
  const httpProxy = getAIConfig(MIDSCENE_OPENAI_HTTP_PROXY);
  let proxyAgent = void 0;
  const debugProxy = getDebug2("ai:call:proxy");
  if (httpProxy) {
    debugProxy("using http proxy", httpProxy);
    proxyAgent = new HttpsProxyAgent(httpProxy);
  } else if (socksProxy) {
    debugProxy("using socks proxy", socksProxy);
    proxyAgent = new SocksProxyAgent(socksProxy);
  }
  if (getAIConfig(OPENAI_USE_AZURE)) {
    openai = new AzureOpenAI({
      baseURL: getAIConfig(OPENAI_BASE_URL),
      apiKey: getAIConfig(OPENAI_API_KEY),
      httpAgent: proxyAgent,
      ...extraConfig,
      dangerouslyAllowBrowser: true
    });
  } else if (getAIConfig(MIDSCENE_USE_AZURE_OPENAI)) {
    const extraAzureConfig = getAIConfigInJson(
      MIDSCENE_AZURE_OPENAI_INIT_CONFIG_JSON
    );
    const scope = getAIConfig(MIDSCENE_AZURE_OPENAI_SCOPE);
    let tokenProvider = void 0;
    if (scope) {
      assert3(
        !ifInBrowser,
        "Azure OpenAI is not supported in browser with Midscene."
      );
      const credential = new DefaultAzureCredential();
      assert3(scope, "MIDSCENE_AZURE_OPENAI_SCOPE is required");
      tokenProvider = getBearerTokenProvider(credential, scope);
      openai = new AzureOpenAI({
        azureADTokenProvider: tokenProvider,
        endpoint: getAIConfig(AZURE_OPENAI_ENDPOINT),
        apiVersion: getAIConfig(AZURE_OPENAI_API_VERSION),
        deployment: getAIConfig(AZURE_OPENAI_DEPLOYMENT),
        ...extraConfig,
        ...extraAzureConfig
      });
    } else {
      openai = new AzureOpenAI({
        apiKey: getAIConfig(AZURE_OPENAI_KEY),
        endpoint: getAIConfig(AZURE_OPENAI_ENDPOINT),
        apiVersion: getAIConfig(AZURE_OPENAI_API_VERSION),
        deployment: getAIConfig(AZURE_OPENAI_DEPLOYMENT),
        dangerouslyAllowBrowser: true,
        ...extraConfig,
        ...extraAzureConfig
      });
    }
  } else if (!getAIConfig(MIDSCENE_USE_ANTHROPIC_SDK)) {
    const baseURL = getAIConfig(OPENAI_BASE_URL);
    if (typeof baseURL === "string") {
      if (!/^https?:\/\//.test(baseURL)) {
        throw new Error(
          `OPENAI_BASE_URL must be a valid URL starting with http:// or https://, but got: ${baseURL}
Please check your config.`
        );
      }
    }
    openai = new OpenAI({
      baseURL: getAIConfig(OPENAI_BASE_URL),
      apiKey: getAIConfig(OPENAI_API_KEY),
      httpAgent: proxyAgent,
      ...extraConfig,
      defaultHeaders: {
        ...extraConfig?.defaultHeaders || {},
        [MIDSCENE_API_TYPE]: AIActionTypeValue.toString()
      },
      dangerouslyAllowBrowser: true
    });
  }
  if (openai && getAIConfigInBoolean(MIDSCENE_LANGSMITH_DEBUG)) {
    if (ifInBrowser) {
      throw new Error("langsmith is not supported in browser");
    }
    console.log("DEBUGGING MODE: langsmith wrapper enabled");
    const { wrapOpenAI } = await import("langsmith/wrappers");
    openai = wrapOpenAI(openai);
  }
  if (typeof openai !== "undefined") {
    return {
      completion: openai.chat.completions,
      style: "openai"
    };
  }
  if (getAIConfig(MIDSCENE_USE_ANTHROPIC_SDK)) {
    const apiKey = getAIConfig(ANTHROPIC_API_KEY);
    assert3(apiKey, "ANTHROPIC_API_KEY is required");
    openai = new Anthropic({
      apiKey,
      httpAgent: proxyAgent,
      dangerouslyAllowBrowser: true
    });
  }
  if (typeof openai !== "undefined" && openai.messages) {
    return {
      completion: openai.messages,
      style: "anthropic"
    };
  }
  throw new Error("Openai SDK or Anthropic SDK is not initialized");
}
async function call2(messages, AIActionTypeValue, responseFormat, options) {
  assert3(
    checkAIConfig(),
    "Cannot find config for AI model service. If you are using a self-hosted model without validating the API key, please set `OPENAI_API_KEY` to any non-null value. https://midscenejs.com/model-provider.html"
  );
  const { completion, style } = await createChatClient({
    AIActionTypeValue
  });
  const maxTokens = getAIConfig(OPENAI_MAX_TOKENS);
  const debugCall = getDebug2("ai:call");
  const debugProfileStats = getDebug2("ai:profile:stats");
  const debugProfileDetail = getDebug2("ai:profile:detail");
  const startTime = Date.now();
  const model = getModelName();
  const isStreaming = options?.stream && options?.onChunk;
  let content;
  let accumulated = "";
  let usage;
  let timeCost;
  const commonConfig = {
    temperature: vlLocateMode3() === "vlm-ui-tars" ? 0 : 0.1,
    stream: !!isStreaming,
    max_tokens: typeof maxTokens === "number" ? maxTokens : Number.parseInt(maxTokens || "2048", 10),
    ...vlLocateMode3() === "qwen-vl" ? {
      vl_high_resolution_images: true
    } : {}
  };
  try {
    if (style === "openai") {
      debugCall(
        `sending ${isStreaming ? "streaming " : ""}request to ${model}`
      );
      if (isStreaming) {
        const stream = await completion.create(
          {
            model,
            messages,
            response_format: responseFormat,
            ...commonConfig
          },
          {
            stream: true
          }
        );
        for await (const chunk of stream) {
          const content2 = chunk.choices?.[0]?.delta?.content || "";
          const reasoning_content = chunk.choices?.[0]?.delta?.reasoning_content || "";
          if (chunk.usage) {
            usage = chunk.usage;
          }
          if (content2 || reasoning_content) {
            accumulated += content2;
            const chunkData = {
              content: content2,
              reasoning_content,
              accumulated,
              isComplete: false,
              usage: void 0
            };
            options.onChunk(chunkData);
          }
          if (chunk.choices?.[0]?.finish_reason) {
            timeCost = Date.now() - startTime;
            if (!usage) {
              const estimatedTokens = Math.max(
                1,
                Math.floor(accumulated.length / 4)
              );
              usage = {
                prompt_tokens: estimatedTokens,
                completion_tokens: estimatedTokens,
                total_tokens: estimatedTokens * 2
              };
            }
            const finalChunk = {
              content: "",
              accumulated,
              reasoning_content: "",
              isComplete: true,
              usage: {
                prompt_tokens: usage.prompt_tokens ?? 0,
                completion_tokens: usage.completion_tokens ?? 0,
                total_tokens: usage.total_tokens ?? 0,
                time_cost: timeCost ?? 0
              }
            };
            options.onChunk(finalChunk);
            break;
          }
        }
        content = accumulated;
        debugProfileStats(
          `streaming model, ${model}, mode, ${vlLocateMode3() || "default"}, cost-ms, ${timeCost}`
        );
      } else {
        const result = await completion.create({
          model,
          messages,
          response_format: responseFormat,
          ...commonConfig
        });
        timeCost = Date.now() - startTime;
        debugProfileStats(
          `model, ${model}, mode, ${vlLocateMode3() || "default"}, ui-tars-version, ${uiTarsModelVersion()}, prompt-tokens, ${result.usage?.prompt_tokens || ""}, completion-tokens, ${result.usage?.completion_tokens || ""}, total-tokens, ${result.usage?.total_tokens || ""}, cost-ms, ${timeCost}, requestId, ${result._request_id || ""}`
        );
        debugProfileDetail(
          `model usage detail: ${JSON.stringify(result.usage)}`
        );
        assert3(
          result.choices,
          `invalid response from LLM service: ${JSON.stringify(result)}`
        );
        content = result.choices[0].message.content;
        usage = result.usage;
      }
      debugCall(`response: ${content}`);
      assert3(content, "empty content");
    } else if (style === "anthropic") {
      const convertImageContent = (content2) => {
        if (content2.type === "image_url") {
          const imgBase64 = content2.image_url.url;
          assert3(imgBase64, "image_url is required");
          return {
            source: {
              type: "base64",
              media_type: imgBase64.includes("data:image/png;base64,") ? "image/png" : "image/jpeg",
              data: imgBase64.split(",")[1]
            },
            type: "image"
          };
        }
        return content2;
      };
      if (isStreaming) {
        const stream = await completion.create({
          model,
          system: "You are a versatile professional in software UI automation",
          messages: messages.map((m) => ({
            role: "user",
            content: Array.isArray(m.content) ? m.content.map(convertImageContent) : m.content
          })),
          response_format: responseFormat,
          ...commonConfig
        });
        for await (const chunk of stream) {
          const content2 = chunk.delta?.text || "";
          if (content2) {
            accumulated += content2;
            const chunkData = {
              content: content2,
              accumulated,
              reasoning_content: "",
              isComplete: false,
              usage: void 0
            };
            options.onChunk(chunkData);
          }
          if (chunk.type === "message_stop") {
            timeCost = Date.now() - startTime;
            const anthropicUsage = chunk.usage;
            const finalChunk = {
              content: "",
              accumulated,
              reasoning_content: "",
              isComplete: true,
              usage: anthropicUsage ? {
                prompt_tokens: anthropicUsage.input_tokens ?? 0,
                completion_tokens: anthropicUsage.output_tokens ?? 0,
                total_tokens: (anthropicUsage.input_tokens ?? 0) + (anthropicUsage.output_tokens ?? 0),
                time_cost: timeCost ?? 0
              } : void 0
            };
            options.onChunk(finalChunk);
            break;
          }
        }
        content = accumulated;
      } else {
        const result = await completion.create({
          model,
          system: "You are a versatile professional in software UI automation",
          messages: messages.map((m) => ({
            role: "user",
            content: Array.isArray(m.content) ? m.content.map(convertImageContent) : m.content
          })),
          response_format: responseFormat,
          ...commonConfig
        });
        timeCost = Date.now() - startTime;
        content = result.content[0].text;
        usage = result.usage;
      }
      assert3(content, "empty content");
    }
    if (isStreaming && !usage) {
      const estimatedTokens = Math.max(
        1,
        Math.floor((content || "").length / 4)
      );
      usage = {
        prompt_tokens: estimatedTokens,
        completion_tokens: estimatedTokens,
        total_tokens: estimatedTokens * 2
      };
    }
    return {
      content: content || "",
      usage: usage ? {
        prompt_tokens: usage.prompt_tokens ?? 0,
        completion_tokens: usage.completion_tokens ?? 0,
        total_tokens: usage.total_tokens ?? 0,
        time_cost: timeCost ?? 0
      } : void 0,
      isStreamed: !!isStreaming
    };
  } catch (e) {
    console.error(" call AI error", e);
    const newError = new Error(
      `failed to call ${isStreaming ? "streaming " : ""}AI model service: ${e.message}. Trouble shooting: https://midscenejs.com/model-provider.html`,
      {
        cause: e
      }
    );
    throw newError;
  }
}
async function callToGetJSONObject(messages, AIActionTypeValue) {
  let responseFormat;
  const model = getModelName();
  if (model.includes("gpt-4")) {
    switch (AIActionTypeValue) {
      case 0 /* ASSERT */:
        responseFormat = assertSchema;
        break;
      case 1 /* INSPECT_ELEMENT */:
        responseFormat = locatorSchema;
        break;
      case 3 /* PLAN */:
        responseFormat = planSchema;
        break;
      case 2 /* EXTRACT_DATA */:
      case 4 /* DESCRIBE_ELEMENT */:
        responseFormat = { type: "json_object" /* JSON */ };
        break;
    }
  }
  if (model === "gpt-4o-2024-05-13") {
    responseFormat = { type: "json_object" /* JSON */ };
  }
  const response = await call2(messages, AIActionTypeValue, responseFormat);
  assert3(response, "empty response");
  const jsonContent = safeParseJson(response.content);
  return { content: jsonContent, usage: response.usage };
}
async function callAiFnWithStringResponse(msgs, AIActionTypeValue) {
  const { content, usage } = await call2(msgs, AIActionTypeValue);
  return { content, usage };
}
function extractJSONFromCodeBlock(response) {
  try {
    const jsonMatch = response.match(/^\s*(\{[\s\S]*\})\s*$/);
    if (jsonMatch) {
      return jsonMatch[1];
    }
    const codeBlockMatch = response.match(
      /```(?:json)?\s*(\{[\s\S]*?\})\s*```/
    );
    if (codeBlockMatch) {
      return codeBlockMatch[1];
    }
    const jsonLikeMatch = response.match(/\{[\s\S]*\}/);
    if (jsonLikeMatch) {
      return jsonLikeMatch[0];
    }
  } catch {
  }
  return response;
}
function preprocessDoubaoBboxJson(input) {
  if (input.includes("bbox")) {
    while (/\d+\s+\d+/.test(input)) {
      input = input.replace(/(\d+)\s+(\d+)/g, "$1,$2");
    }
  }
  return input;
}
function safeParseJson(input) {
  const cleanJsonString = extractJSONFromCodeBlock(input);
  if (cleanJsonString?.match(/\((\d+),(\d+)\)/)) {
    return cleanJsonString.match(/\((\d+),(\d+)\)/)?.slice(1).map(Number);
  }
  try {
    return JSON.parse(cleanJsonString);
  } catch {
  }
  try {
    return JSON.parse(jsonrepair(cleanJsonString));
  } catch (e) {
  }
  if (vlLocateMode3() === "doubao-vision" || vlLocateMode3() === "vlm-ui-tars") {
    const jsonString = preprocessDoubaoBboxJson(cleanJsonString);
    return JSON.parse(jsonrepair(jsonString));
  }
  throw Error(`failed to parse json response: ${input}`);
}

// src/ai-model/prompt/playwright-generator.ts
import { PLAYWRIGHT_EXAMPLE_CODE } from "@midscene/shared/constants";

// src/ai-model/prompt/yaml-generator.ts
import { YAML_EXAMPLE_CODE } from "@midscene/shared/constants";
var getScreenshotsForLLM = (events, maxScreenshots = 1) => {
  const eventsWithScreenshots = events.filter(
    (event) => event.screenshotBefore || event.screenshotAfter || event.screenshotWithBox
  );
  const sortedEvents = [...eventsWithScreenshots].sort((a, b) => {
    if (a.type === "navigation" && b.type !== "navigation")
      return -1;
    if (a.type !== "navigation" && b.type === "navigation")
      return 1;
    if (a.type === "click" && b.type !== "click")
      return -1;
    if (a.type !== "click" && b.type === "click")
      return 1;
    return 0;
  });
  const screenshots = [];
  for (const event of sortedEvents) {
    const screenshot = event.screenshotWithBox || event.screenshotAfter || event.screenshotBefore;
    if (screenshot && !screenshots.includes(screenshot)) {
      screenshots.push(screenshot);
      if (screenshots.length >= maxScreenshots)
        break;
    }
  }
  return screenshots;
};
var filterEventsByType = (events) => {
  return {
    navigationEvents: events.filter((event) => event.type === "navigation"),
    clickEvents: events.filter((event) => event.type === "click"),
    inputEvents: events.filter((event) => event.type === "input"),
    scrollEvents: events.filter((event) => event.type === "scroll")
  };
};
var createEventCounts = (filteredEvents, totalEvents) => {
  return {
    navigation: filteredEvents.navigationEvents.length,
    click: filteredEvents.clickEvents.length,
    input: filteredEvents.inputEvents.length,
    scroll: filteredEvents.scrollEvents.length,
    total: totalEvents
  };
};
var extractInputDescriptions = (inputEvents) => {
  return inputEvents.map((event) => ({
    description: event.elementDescription || "",
    value: event.value || ""
  })).filter((item) => item.description && item.value);
};
var processEventsForLLM = (events) => {
  return events.map((event) => ({
    type: event.type,
    timestamp: event.timestamp,
    url: event.url,
    title: event.title,
    elementDescription: event.elementDescription,
    value: event.value,
    pageInfo: event.pageInfo,
    elementRect: event.elementRect
  }));
};
var prepareEventSummary = (events, options = {}) => {
  const filteredEvents = filterEventsByType(events);
  const eventCounts = createEventCounts(filteredEvents, events.length);
  const startUrl = filteredEvents.navigationEvents.length > 0 ? filteredEvents.navigationEvents[0].url || "" : "";
  const clickDescriptions = filteredEvents.clickEvents.map((event) => event.elementDescription).filter((desc) => Boolean(desc)).slice(0, 10);
  const inputDescriptions = extractInputDescriptions(
    filteredEvents.inputEvents
  ).slice(0, 10);
  const urls = filteredEvents.navigationEvents.map((e) => e.url).filter((url) => Boolean(url)).slice(0, 5);
  const processedEvents = processEventsForLLM(events);
  return {
    testName: options.testName || "Automated test from recorded events",
    startUrl,
    eventCounts,
    urls,
    clickDescriptions,
    inputDescriptions,
    events: processedEvents
  };
};
var createMessageContent = (promptText, screenshots = [], includeScreenshots = true) => {
  const messageContent = [
    {
      type: "text",
      text: promptText
    }
  ];
  if (includeScreenshots && screenshots.length > 0) {
    messageContent.unshift({
      type: "text",
      text: "Here are screenshots from the recording session to help you understand the context:"
    });
    screenshots.forEach((screenshot) => {
      messageContent.push({
        type: "image_url",
        image_url: {
          url: screenshot
        }
      });
    });
  }
  return messageContent;
};
var validateEvents = (events) => {
  if (!events.length) {
    throw new Error("No events provided for test generation");
  }
};
var generateYamlTest = async (events, options = {}) => {
  try {
    validateEvents(events);
    const summary = prepareEventSummary(events, {
      testName: options.testName,
      maxScreenshots: options.maxScreenshots || 3
    });
    const yamlSummary = {
      ...summary,
      includeTimestamps: options.includeTimestamps || false
    };
    const screenshots = getScreenshotsForLLM(
      events,
      options.maxScreenshots || 3
    );
    const prompt = [
      {
        role: "system",
        content: `You are an expert in Midscene.js YAML test generation. Generate clean, accurate YAML following these rules: ${YAML_EXAMPLE_CODE}`
      },
      {
        role: "user",
        content: `Generate YAML test for Midscene.js automation from recorded browser events.

Event Summary:
${JSON.stringify(yamlSummary, null, 2)}

Convert events:
- navigation → target.url
- click → aiTap with element description
- input → aiInput with value and locate
- scroll → aiScroll with appropriate direction
- Add aiAssert for important state changes

Respond with YAML only, no explanations.`
      }
    ];
    if (screenshots.length > 0) {
      prompt.push({
        role: "user",
        content: "Here are screenshots from the recording session to help you understand the context:"
      });
      prompt.push({
        role: "user",
        content: screenshots.map((screenshot) => ({
          type: "image_url",
          image_url: {
            url: screenshot
          }
        }))
      });
    }
    const response = await call2(prompt, 2 /* EXTRACT_DATA */);
    if (response?.content && typeof response.content === "string") {
      return response.content;
    }
    throw new Error("Failed to generate YAML test configuration");
  } catch (error) {
    throw new Error(`Failed to generate YAML test: ${error}`);
  }
};
var generateYamlTestStream = async (events, options = {}) => {
  try {
    validateEvents(events);
    const summary = prepareEventSummary(events, {
      testName: options.testName,
      maxScreenshots: options.maxScreenshots || 3
    });
    const yamlSummary = {
      ...summary,
      includeTimestamps: options.includeTimestamps || false
    };
    const screenshots = getScreenshotsForLLM(
      events,
      options.maxScreenshots || 3
    );
    const prompt = [
      {
        role: "system",
        content: `You are an expert in Midscene.js YAML test generation. Generate clean, accurate YAML following these rules: ${YAML_EXAMPLE_CODE}`
      },
      {
        role: "user",
        content: `Generate YAML test for Midscene.js automation from recorded browser events.

Event Summary:
${JSON.stringify(yamlSummary, null, 2)}

Convert events:
- navigation → target.url
- click → aiTap with element description
- input → aiInput with value and locate
- scroll → aiScroll with appropriate direction
- Add aiAssert for important state changes

Respond with YAML only, no explanations.`
      }
    ];
    if (screenshots.length > 0) {
      prompt.push({
        role: "user",
        content: "Here are screenshots from the recording session to help you understand the context:"
      });
      prompt.push({
        role: "user",
        content: screenshots.map((screenshot) => ({
          type: "image_url",
          image_url: {
            url: screenshot
          }
        }))
      });
    }
    if (options.stream && options.onChunk) {
      return await call2(prompt, 2 /* EXTRACT_DATA */, void 0, {
        stream: true,
        onChunk: options.onChunk
      });
    } else {
      const response = await call2(prompt, 2 /* EXTRACT_DATA */);
      if (response?.content && typeof response.content === "string") {
        return {
          content: response.content,
          usage: response.usage,
          isStreamed: false
        };
      }
      throw new Error("Failed to generate YAML test configuration");
    }
  } catch (error) {
    throw new Error(`Failed to generate YAML test: ${error}`);
  }
};

// src/ai-model/prompt/playwright-generator.ts
var generatePlaywrightTest = async (events, options = {}) => {
  validateEvents(events);
  const summary = prepareEventSummary(events, {
    testName: options.testName,
    maxScreenshots: options.maxScreenshots || 3
  });
  const playwrightSummary = {
    ...summary,
    waitForNetworkIdle: options.waitForNetworkIdle !== false,
    waitForNetworkIdleTimeout: options.waitForNetworkIdleTimeout || 2e3,
    viewportSize: options.viewportSize || { width: 1280, height: 800 }
  };
  const screenshots = getScreenshotsForLLM(events, options.maxScreenshots || 3);
  const promptText = `Generate a Playwright test using @midscene/web/playwright that reproduces this recorded browser session. The test should be based on the following events and follow the structure of the example provided. Make the test descriptive with appropriate assertions and validations.

Event Summary:
${JSON.stringify(playwrightSummary, null, 2)}

Generated code should:
1. Import required dependencies
2. Set up the test with proper configuration
3. Include a beforeEach hook to navigate to the starting URL
4. Implement a test that uses Midscene AI methods (aiTap, aiInput, aiAssert, etc.)
5. Include appropriate assertions and validations
6. Follow best practices for Playwright tests
7. Be ready to execute without further modification

Respond ONLY with the complete Playwright test code, no explanations.`;
  const messageContent = createMessageContent(
    promptText,
    screenshots,
    options.includeScreenshots !== false
  );
  const systemPrompt = `You are an expert test automation engineer specializing in Playwright and Midscene. 
Your task is to generate a complete, executable Playwright test using @midscene/web/playwright that reproduces a recorded browser session.

${PLAYWRIGHT_EXAMPLE_CODE}`;
  const prompt = [
    {
      role: "system",
      content: systemPrompt
    },
    {
      role: "user",
      content: messageContent
    }
  ];
  const response = await call2(prompt, 2 /* EXTRACT_DATA */);
  if (response?.content && typeof response.content === "string") {
    return response.content;
  }
  throw new Error("Failed to generate Playwright test code");
};
var generatePlaywrightTestStream = async (events, options = {}) => {
  validateEvents(events);
  const summary = prepareEventSummary(events, {
    testName: options.testName,
    maxScreenshots: options.maxScreenshots || 3
  });
  const playwrightSummary = {
    ...summary,
    waitForNetworkIdle: options.waitForNetworkIdle !== false,
    waitForNetworkIdleTimeout: options.waitForNetworkIdleTimeout || 2e3,
    viewportSize: options.viewportSize || { width: 1280, height: 800 }
  };
  const screenshots = getScreenshotsForLLM(events, options.maxScreenshots || 3);
  const promptText = `Generate a Playwright test using @midscene/web/playwright that reproduces this recorded browser session. The test should be based on the following events and follow the structure of the example provided. Make the test descriptive with appropriate assertions and validations.

Event Summary:
${JSON.stringify(playwrightSummary, null, 2)}

Generated code should:
1. Import required dependencies
2. Set up the test with proper configuration
3. Include a beforeEach hook to navigate to the starting URL
4. Implement a test that uses Midscene AI methods (aiTap, aiInput, aiAssert, etc.)
5. Include appropriate assertions and validations
6. Follow best practices for Playwright tests
7. Be ready to execute without further modification
8. can't wrap this test code in markdown code block

Respond ONLY with the complete Playwright test code, no explanations.`;
  const messageContent = createMessageContent(
    promptText,
    screenshots,
    options.includeScreenshots !== false
  );
  const systemPrompt = `You are an expert test automation engineer specializing in Playwright and Midscene. 
Your task is to generate a complete, executable Playwright test using @midscene/web/playwright that reproduces a recorded browser session.

${PLAYWRIGHT_EXAMPLE_CODE}`;
  const prompt = [
    {
      role: "system",
      content: systemPrompt
    },
    {
      role: "user",
      content: messageContent
    }
  ];
  if (options.stream && options.onChunk) {
    return await call2(prompt, 2 /* EXTRACT_DATA */, void 0, {
      stream: true,
      onChunk: options.onChunk
    });
  } else {
    const response = await call2(prompt, 2 /* EXTRACT_DATA */);
    if (response?.content && typeof response.content === "string") {
      return {
        content: response.content,
        usage: response.usage,
        isStreamed: false
      };
    }
    throw new Error("Failed to generate Playwright test code");
  }
};

// src/ai-model/inspect.ts
import {
  MIDSCENE_USE_QWEN_VL,
  MIDSCENE_USE_VLM_UI_TARS,
  getAIConfigInBoolean as getAIConfigInBoolean2,
  vlLocateMode as vlLocateMode4
} from "@midscene/shared/env";
import {
  cropByRect,
  paddingToMatchBlockByBase64,
  preProcessImageUrl
} from "@midscene/shared/img";
import { getDebug as getDebug3 } from "@midscene/shared/logger";
import { assert as assert4 } from "@midscene/shared/utils";

// src/ai-model/prompt/extraction.ts
import { PromptTemplate as PromptTemplate3 } from "@langchain/core/prompts";
function systemPromptToExtract() {
  return `
You are a versatile professional in software UI design and testing. Your outstanding contributions will impact the user experience of billions of users.

The user will give you a screenshot, the contents of it (optional), and some data requirements in <DATA_DEMAND>. You need to extract the data according to the <DATA_DEMAND>.

If a key specifies a JSON data type (such as Number, String, Boolean, Object, Array), ensure the returned value strictly matches that data type.

If the user provides multiple reference images, please carefully review the reference images with the screenshot and provide the correct answer for <DATA_DEMAND>.

If the user requests reasons to be provided, please provide the thought field in response, less then 100 words.

Return in the following JSON format:
{
  thought: string, // the thought process of the extraction, less then 100 words, not required by default.
  data: any, // the extracted data. Make sure both the value and scheme meet the DATA_DEMAND. If you want to write some description in this field, use the same language as the DATA_DEMAND.
  errors: [], // string[], error message if any
}

# Example 1
For example, if the DATA_DEMAND is:

<DATA_DEMAND>
{
  "name": "name shows on the left panel, string",
  "age": "age shows on the right panel, number",
  "isAdmin": "if the user is admin, boolean"
}
</DATA_DEMAND>

By viewing the screenshot and page contents, you can extract the following data:

{
  data: {
    name: "John",
    age: 30,
    isAdmin: true
  },
}

# Example 2
If the DATA_DEMAND is:

<DATA_DEMAND>
the todo items list, string[]
</DATA_DEMAND>

By viewing the screenshot and page contents, you can extract the following data:

{
  data: ["todo 1", "todo 2", "todo 3"],
}

# Example 3
If the DATA_DEMAND is:

<DATA_DEMAND>
the page title, string
</DATA_DEMAND>

By viewing the screenshot and page contents, you can extract the following data:

{
  data: "todo list",
}

# Example 4
If the DATA_DEMAND is:

<DATA_DEMAND>
{
  "result": "Boolean, is it currently the SMS page?"
}
</DATA_DEMAND>

By viewing the screenshot and page contents, you can extract the following data:

{
  data: { result: true },
}
`;
}
var extractDataQueryPrompt = async (pageDescription, dataQuery) => {
  let dataQueryText = "";
  if (typeof dataQuery === "string") {
    dataQueryText = dataQuery;
  } else {
    dataQueryText = JSON.stringify(dataQuery, null, 2);
  }
  const extractDataPrompt = new PromptTemplate3({
    template: `
<PageDescription>
{pageDescription}
</PageDescription>

<DATA_DEMAND>
{dataQuery}
</DATA_DEMAND>
  `,
    inputVariables: ["pageDescription", "dataQuery"]
  });
  return await extractDataPrompt.format({
    pageDescription,
    dataQuery: dataQueryText
  });
};

// src/ai-model/prompt/llm-section-locator.ts
import { PromptTemplate as PromptTemplate4 } from "@langchain/core/prompts";
function systemPromptToLocateSection(vlMode) {
  return `
You goal is to find out one section containing the target element in the screenshot, put it in the \`bbox\` field. If the user describe the target element with some reference elements, you should also find the section containing the reference elements, put it in the \`references_bbox\` field.

Usually, it should be approximately an area not more than 300x300px. Changes of the size are allowed if there are many elements to cover.

return in this JSON format:
\`\`\`json
{
  "bbox": [number, number, number, number],
  "references_bbox"?: [
    [number, number, number, number],
    [number, number, number, number],
    ...
  ],
  "error"?: string
}
\`\`\`

In which, all the numbers in the \`bbox\` and \`references_bbox\` represent ${bboxDescription(vlMode)}.

For example, if the user describe the target element as "the delete button on the second row with title 'Peter'", you should put the bounding box of the delete button in the \`bbox\` field, and the bounding box of the second row in the \`references_bbox\` field.

the return value should be like this:
\`\`\`json
{
  "bbox": [100, 100, 200, 200],
  "references_bbox": [[100, 100, 200, 200]]
}
\`\`\`
`;
}
var sectionLocatorInstruction = new PromptTemplate4({
  template: `Here is the target element user interested in:
<targetDescription>
{sectionDescription}
</targetDescription>
  `,
  inputVariables: ["sectionDescription"]
});

// src/ai-model/inspect.ts
var debugInspect = getDebug3("ai:inspect");
var debugSection = getDebug3("ai:section");
var extraTextFromUserPrompt = (prompt) => {
  if (typeof prompt === "string") {
    return prompt;
  } else {
    return prompt.prompt;
  }
};
var promptsToChatParam = async (multimodalPrompt) => {
  const msgs = [];
  if (multimodalPrompt?.images?.length) {
    msgs.push({
      role: "user",
      content: [
        {
          type: "text",
          text: "Next, I will provide all the reference images."
        }
      ]
    });
    for (const item of multimodalPrompt.images) {
      const base64 = await preProcessImageUrl(
        item.url,
        !!multimodalPrompt.convertHttpImage2Base64
      );
      msgs.push({
        role: "user",
        content: [
          {
            type: "text",
            text: `reference image ${item.name}:`
          }
        ]
      });
      msgs.push({
        role: "user",
        content: [
          {
            type: "image_url",
            image_url: {
              url: base64,
              detail: "high"
            }
          }
        ]
      });
    }
  }
  return msgs;
};
async function AiLocateElement(options) {
  const { context, targetElementDescription, callAI } = options;
  const { screenshotBase64 } = context;
  const { description, elementById, insertElementByPosition } = await describeUserPage(context);
  assert4(
    targetElementDescription,
    "cannot find the target element description"
  );
  const userInstructionPrompt = await findElementPrompt.format({
    pageDescription: description,
    targetElementDescription: extraTextFromUserPrompt(targetElementDescription)
  });
  const systemPrompt = systemPromptToLocateElement(vlLocateMode4());
  let imagePayload = screenshotBase64;
  if (options.searchConfig) {
    assert4(
      options.searchConfig.rect,
      "searchArea is provided but its rect cannot be found. Failed to locate element"
    );
    assert4(
      options.searchConfig.imageBase64,
      "searchArea is provided but its imageBase64 cannot be found. Failed to locate element"
    );
    imagePayload = options.searchConfig.imageBase64;
  } else if (vlLocateMode4() === "qwen-vl") {
    imagePayload = await paddingToMatchBlockByBase64(imagePayload);
  } else if (!vlLocateMode4()) {
    imagePayload = await markupImageForLLM(
      screenshotBase64,
      context.tree,
      context.size
    );
  }
  const msgs = [
    { role: "system", content: systemPrompt },
    {
      role: "user",
      content: [
        {
          type: "image_url",
          image_url: {
            url: imagePayload,
            detail: "high"
          }
        },
        {
          type: "text",
          text: userInstructionPrompt
        }
      ]
    }
  ];
  if (typeof targetElementDescription !== "string") {
    const addOns = await promptsToChatParam({
      images: targetElementDescription.images,
      convertHttpImage2Base64: targetElementDescription.convertHttpImage2Base64
    });
    msgs.push(...addOns);
  }
  const callAIFn = callAI || callToGetJSONObject;
  const res = await callAIFn(msgs, 1 /* INSPECT_ELEMENT */);
  const rawResponse = JSON.stringify(res.content);
  let resRect;
  let matchedElements = "elements" in res.content ? res.content.elements : [];
  let errors = "errors" in res.content ? res.content.errors : [];
  try {
    if ("bbox" in res.content && Array.isArray(res.content.bbox)) {
      resRect = adaptBboxToRect(
        res.content.bbox,
        options.searchConfig?.rect?.width || context.size.width,
        options.searchConfig?.rect?.height || context.size.height,
        options.searchConfig?.rect?.left,
        options.searchConfig?.rect?.top
      );
      debugInspect("resRect", resRect);
      const rectCenter = {
        x: resRect.left + resRect.width / 2,
        y: resRect.top + resRect.height / 2
      };
      let element = elementByPositionWithElementInfo(context.tree, rectCenter);
      const distanceToCenter = element ? distance({ x: element.center[0], y: element.center[1] }, rectCenter) : 0;
      if (!element || distanceToCenter > distanceThreshold) {
        element = insertElementByPosition(rectCenter);
      }
      if (element) {
        matchedElements = [element];
        errors = [];
      }
    }
  } catch (e) {
    const msg = e instanceof Error ? `Failed to parse bbox: ${e.message}` : "unknown error in locate";
    if (!errors || errors?.length === 0) {
      errors = [msg];
    } else {
      errors.push(`(${msg})`);
    }
  }
  return {
    rect: resRect,
    parseResult: {
      elements: matchedElements,
      errors
    },
    rawResponse,
    elementById,
    usage: res.usage,
    isOrderSensitive: typeof res.content === "object" && res.content !== null && "isOrderSensitive" in res.content ? res.content.isOrderSensitive : void 0
  };
}
async function AiLocateSection(options) {
  const { context, sectionDescription } = options;
  const { screenshotBase64 } = context;
  const systemPrompt = systemPromptToLocateSection(vlLocateMode4());
  const sectionLocatorInstructionText = await sectionLocatorInstruction.format({
    sectionDescription: extraTextFromUserPrompt(sectionDescription)
  });
  const msgs = [
    { role: "system", content: systemPrompt },
    {
      role: "user",
      content: [
        {
          type: "image_url",
          image_url: {
            url: screenshotBase64,
            detail: "high"
          }
        },
        {
          type: "text",
          text: sectionLocatorInstructionText
        }
      ]
    }
  ];
  if (typeof sectionDescription !== "string") {
    const addOns = await promptsToChatParam({
      images: sectionDescription.images,
      convertHttpImage2Base64: sectionDescription.convertHttpImage2Base64
    });
    msgs.push(...addOns);
  }
  const result = await callAiFn(
    msgs,
    2 /* EXTRACT_DATA */
  );
  let sectionRect;
  const sectionBbox = result.content.bbox;
  if (sectionBbox) {
    const targetRect = adaptBboxToRect(
      sectionBbox,
      context.size.width,
      context.size.height
    );
    debugSection("original targetRect %j", targetRect);
    const referenceBboxList = result.content.references_bbox || [];
    debugSection("referenceBboxList %j", referenceBboxList);
    const referenceRects = referenceBboxList.filter((bbox) => Array.isArray(bbox)).map((bbox) => {
      return adaptBboxToRect(bbox, context.size.width, context.size.height);
    });
    debugSection("referenceRects %j", referenceRects);
    const mergedRect = mergeRects([targetRect, ...referenceRects]);
    debugSection("mergedRect %j", mergedRect);
    sectionRect = expandSearchArea(mergedRect, context.size);
    debugSection("expanded sectionRect %j", sectionRect);
  }
  let imageBase64 = screenshotBase64;
  if (sectionRect) {
    imageBase64 = await cropByRect(
      screenshotBase64,
      sectionRect,
      getAIConfigInBoolean2(MIDSCENE_USE_QWEN_VL)
    );
  }
  return {
    rect: sectionRect,
    imageBase64,
    error: result.content.error,
    rawResponse: JSON.stringify(result.content),
    usage: result.usage
  };
}
async function AiExtractElementInfo(options) {
  const { dataQuery, context, extractOption, multimodalPrompt } = options;
  const systemPrompt = systemPromptToExtract();
  const { screenshotBase64 } = context;
  const { description, elementById } = await describeUserPage(context, {
    truncateTextLength: 200,
    filterNonTextContent: false,
    visibleOnly: false,
    domIncluded: extractOption?.domIncluded
  });
  const extractDataPromptText = await extractDataQueryPrompt(
    description,
    dataQuery
  );
  const userContent = [];
  if (extractOption?.screenshotIncluded !== false) {
    userContent.push({
      type: "image_url",
      image_url: {
        url: screenshotBase64,
        detail: "high"
      }
    });
  }
  userContent.push({
    type: "text",
    text: extractDataPromptText
  });
  const msgs = [
    { role: "system", content: systemPrompt },
    {
      role: "user",
      content: userContent
    }
  ];
  if (options.extractOption?.returnThought) {
    msgs.push({
      role: "user",
      content: "Please provide reasons."
    });
  }
  if (multimodalPrompt) {
    const addOns = await promptsToChatParam({
      images: multimodalPrompt.images,
      convertHttpImage2Base64: multimodalPrompt.convertHttpImage2Base64
    });
    msgs.push(...addOns);
  }
  const result = await callAiFn(
    msgs,
    2 /* EXTRACT_DATA */
  );
  return {
    parseResult: result.content,
    elementById,
    usage: result.usage
  };
}
async function AiAssert(options) {
  const { assertion, context } = options;
  assert4(assertion, "assertion should not be empty");
  const { screenshotBase64 } = context;
  const systemPrompt = systemPromptToAssert({
    isUITars: getAIConfigInBoolean2(MIDSCENE_USE_VLM_UI_TARS)
  });
  const assertionText = extraTextFromUserPrompt(assertion);
  const msgs = [
    { role: "system", content: systemPrompt },
    {
      role: "user",
      content: [
        {
          type: "image_url",
          image_url: {
            url: screenshotBase64,
            detail: "high"
          }
        },
        {
          type: "text",
          text: `
Here is the assertion. Please tell whether it is truthy according to the screenshot.
=====================================
${assertionText}
=====================================
  `
        }
      ]
    }
  ];
  if (typeof assertion !== "string") {
    const addOns = await promptsToChatParam({
      images: assertion.images,
      convertHttpImage2Base64: assertion.convertHttpImage2Base64
    });
    msgs.push(...addOns);
  }
  const { content: assertResult, usage } = await callAiFn(
    msgs,
    0 /* ASSERT */
  );
  return {
    content: assertResult,
    usage
  };
}

// src/ai-model/llm-planning.ts
import { vlLocateMode as vlLocateMode5 } from "@midscene/shared/env";
import { paddingToMatchBlockByBase64 as paddingToMatchBlockByBase642 } from "@midscene/shared/img";
import { assert as assert5 } from "@midscene/shared/utils";
async function plan(userInstruction, opts) {
  const { callAI, context } = opts || {};
  const { screenshotBase64, size } = context;
  const { description: pageDescription, elementById } = await describeUserPage(context);
  const systemPrompt = await systemPromptToTaskPlanning({
    pageType: opts.pageType,
    vlMode: vlLocateMode5()
  });
  const taskBackgroundContextText = generateTaskBackgroundContext(
    userInstruction,
    opts.log,
    opts.actionContext
  );
  const userInstructionPrompt = await automationUserPrompt(
    vlLocateMode5()
  ).format({
    pageDescription,
    taskBackgroundContext: taskBackgroundContextText
  });
  let imagePayload = screenshotBase64;
  if (vlLocateMode5() === "qwen-vl") {
    imagePayload = await paddingToMatchBlockByBase642(imagePayload);
  } else if (!vlLocateMode5()) {
    imagePayload = await markupImageForLLM(
      screenshotBase64,
      context.tree,
      context.size
    );
  }
  warnGPT4oSizeLimit(size);
  const msgs = [
    { role: "system", content: systemPrompt },
    {
      role: "user",
      content: [
        {
          type: "image_url",
          image_url: {
            url: imagePayload,
            detail: "high"
          }
        },
        {
          type: "text",
          text: userInstructionPrompt
        }
      ]
    }
  ];
  const call3 = callAI || callAiFn;
  const { content, usage } = await call3(msgs, 3 /* PLAN */);
  const rawResponse = JSON.stringify(content, void 0, 2);
  const planFromAI = content;
  const actions = (planFromAI.action?.type ? [planFromAI.action] : planFromAI.actions) || [];
  const returnValue = {
    ...planFromAI,
    actions,
    rawResponse,
    usage,
    yamlFlow: buildYamlFlowFromPlans(actions, planFromAI.sleep)
  };
  assert5(planFromAI, "can't get plans from AI");
  if (vlLocateMode5()) {
    actions.forEach((action) => {
      if (action.locate) {
        try {
          action.locate = fillBboxParam(action.locate, size.width, size.height);
        } catch (e) {
          throw new Error(
            `Failed to fill locate param: ${planFromAI.error} (${e instanceof Error ? e.message : "unknown error"})`,
            {
              cause: e
            }
          );
        }
      }
    });
    assert5(!planFromAI.error, `Failed to plan actions: ${planFromAI.error}`);
  } else {
    actions.forEach((action) => {
      if (action.locate?.id) {
        const element = elementById(action.locate.id);
        if (element) {
          action.locate.id = element.id;
        }
      }
    });
  }
  if (actions.length === 0 && returnValue.more_actions_needed_by_instruction && !returnValue.sleep) {
    console.warn(
      "No actions planned for the prompt, but model said more actions are needed:",
      userInstruction
    );
  }
  return returnValue;
}

// src/ai-model/ui-tars-planning.ts
import {
  UITarsModelVersion,
  uiTarsModelVersion as uiTarsModelVersion2,
  vlLocateMode as vlLocateMode6
} from "@midscene/shared/env";
import { resizeImgBase64 } from "@midscene/shared/img";
import { transformHotkeyInput } from "@midscene/shared/keyboard-layout";
import { getDebug as getDebug4 } from "@midscene/shared/logger";
import { assert as assert6 } from "@midscene/shared/utils";
import { actionParser } from "@ui-tars/action-parser";

// src/ai-model/prompt/ui-tars-planning.ts
import { getPreferredLanguage as getPreferredLanguage2 } from "@midscene/shared/env";
function getUiTarsPlanningPrompt() {
  const preferredLanguage = getPreferredLanguage2();
  return `
You are a GUI agent. You are given a task and your action history, with screenshots. You need to perform the next action to complete the task. 

## Output Format
\`\`\`
Thought: ...
Action: ...
\`\`\`

## Action Space

click(start_box='[x1, y1, x2, y2]')
left_double(start_box='[x1, y1, x2, y2]')
right_single(start_box='[x1, y1, x2, y2]')
drag(start_box='[x1, y1, x2, y2]', end_box='[x3, y3, x4, y4]')
hotkey(key='')
type(content='xxx') # Use escape characters \\', \\", and \\n in content part to ensure we can parse the content in normal python string format. If you want to submit your input, use \\n at the end of content. 
scroll(start_box='[x1, y1, x2, y2]', direction='down or up or right or left')
wait() #Sleep for 5s and take a screenshot to check for any changes.
finished(content='xxx') # Use escape characters \\', \\", and \\n in content part to ensure we can parse the content in normal python string format.


## Note
- Use ${preferredLanguage} in \`Thought\` part.
- Write a small plan and finally summarize your next action (with its target element) in one sentence in \`Thought\` part.

## User Instruction
`;
}
var getSummary = (prediction) => prediction.replace(/Reflection:[\s\S]*?(?=Action_Summary:|Action:|$)/g, "").trim();

// src/ai-model/ui-tars-planning.ts
var debug = getDebug4("ui-tars-planning");
var bboxSize = 10;
var pointToBbox = (point, width, height) => {
  return [
    Math.round(Math.max(point.x - bboxSize / 2, 0)),
    Math.round(Math.max(point.y - bboxSize / 2, 0)),
    Math.round(Math.min(point.x + bboxSize / 2, width)),
    Math.round(Math.min(point.y + bboxSize / 2, height))
  ];
};
async function vlmPlanning(options) {
  const { conversationHistory, userInstruction, size } = options;
  const systemPrompt = getUiTarsPlanningPrompt() + userInstruction;
  const res = await call2(
    [
      {
        role: "user",
        content: systemPrompt
      },
      ...conversationHistory
    ],
    1 /* INSPECT_ELEMENT */
  );
  const convertedText = convertBboxToCoordinates(res.content);
  const modelVer = uiTarsModelVersion2();
  const { parsed } = actionParser({
    prediction: convertedText,
    factor: [1e3, 1e3],
    screenContext: {
      width: size.width,
      height: size.height
    },
    modelVer: modelVer || void 0
  });
  debug("modelVer", modelVer, "parsed", JSON.stringify(parsed));
  const transformActions = [];
  parsed.forEach((action) => {
    if (action.action_type === "click") {
      assert6(action.action_inputs.start_box, "start_box is required");
      const point = getPoint(action.action_inputs.start_box, size);
      transformActions.push({
        type: "Locate",
        param: {},
        locate: {
          prompt: action.thought || "",
          bbox: pointToBbox(
            { x: point[0], y: point[1] },
            size.width,
            size.height
          )
        }
      });
      transformActions.push({
        type: "Tap",
        locate: {
          prompt: action.thought || "",
          bbox: pointToBbox(
            { x: point[0], y: point[1] },
            size.width,
            size.height
          )
        },
        param: action.thought || ""
      });
    } else if (action.action_type === "drag") {
      assert6(action.action_inputs.start_box, "start_box is required");
      assert6(action.action_inputs.end_box, "end_box is required");
      const startPoint = getPoint(action.action_inputs.start_box, size);
      const endPoint = getPoint(action.action_inputs.end_box, size);
      transformActions.push({
        type: "Drag",
        param: {
          start_box: { x: startPoint[0], y: startPoint[1] },
          end_box: { x: endPoint[0], y: endPoint[1] }
        },
        locate: null,
        thought: action.thought || ""
      });
    } else if (action.action_type === "type") {
      transformActions.push({
        type: "Input",
        param: {
          value: action.action_inputs.content
        },
        locate: null,
        thought: action.thought || ""
      });
    } else if (action.action_type === "scroll") {
      transformActions.push({
        type: "Scroll",
        param: {
          direction: action.action_inputs.direction
        },
        locate: null,
        thought: action.thought || ""
      });
    } else if (action.action_type === "finished") {
      transformActions.push({
        type: "Finished",
        param: {},
        locate: null,
        thought: action.thought || ""
      });
    } else if (action.action_type === "hotkey") {
      if (!action.action_inputs.key) {
        console.warn(
          "No key found in action: hotkey. Will not perform action."
        );
      } else {
        const keys = transformHotkeyInput(action.action_inputs.key);
        transformActions.push({
          type: "KeyboardPress",
          param: {
            value: keys
          },
          locate: null,
          thought: action.thought || ""
        });
      }
    } else if (action.action_type === "wait") {
      transformActions.push({
        type: "Sleep",
        param: {
          timeMs: 1e3
        },
        locate: null,
        thought: action.thought || ""
      });
    } else if (action.action_type === "androidBackButton") {
      transformActions.push({
        type: "AndroidBackButton",
        param: {},
        locate: null,
        thought: action.thought || ""
      });
    } else if (action.action_type === "androidHomeButton") {
      transformActions.push({
        type: "AndroidHomeButton",
        param: {},
        locate: null,
        thought: action.thought || ""
      });
    } else if (action.action_type === "androidRecentAppsButton") {
      transformActions.push({
        type: "AndroidRecentAppsButton",
        param: {}
      });
    } else if (action.action_type === "androidLongPress") {
      assert6(
        action.action_inputs.start_coords,
        "start_coords is required for androidLongPress"
      );
      const point = action.action_inputs.start_coords;
      transformActions.push({
        type: "AndroidLongPress",
        param: {
          x: point[0],
          y: point[1],
          duration: 1e3
        },
        locate: null,
        thought: action.thought || ""
      });
    } else if (action.action_type === "androidPull") {
      const pullDirection = action.action_inputs.direction || "down";
      const startPoint = action.action_inputs.start_coords ? {
        x: action.action_inputs.start_coords[0],
        y: action.action_inputs.start_coords[1]
      } : void 0;
      transformActions.push({
        type: "AndroidPull",
        param: {
          direction: pullDirection,
          startPoint,
          distance: action.action_inputs.distance,
          duration: action.action_inputs.duration || 500
        },
        locate: null,
        thought: action.thought || ""
      });
    }
  });
  if (transformActions.length === 0) {
    throw new Error(`No actions found, response: ${res.content}`, {
      cause: {
        prediction: res.content,
        parsed
      }
    });
  }
  return {
    actions: transformActions,
    actionsFromModel: parsed,
    action_summary: getSummary(res.content),
    usage: res.usage,
    rawResponse: JSON.stringify(res.content, void 0, 2)
  };
}
function convertBboxToCoordinates(text) {
  const pattern = /<bbox>(\d+)\s+(\d+)\s+(\d+)\s+(\d+)<\/bbox>/g;
  function replaceMatch(match, x1, y1, x2, y2) {
    const x1Num = Number.parseInt(x1, 10);
    const y1Num = Number.parseInt(y1, 10);
    const x2Num = Number.parseInt(x2, 10);
    const y2Num = Number.parseInt(y2, 10);
    const x = Math.floor((x1Num + x2Num) / 2);
    const y = Math.floor((y1Num + y2Num) / 2);
    return `(${x},${y})`;
  }
  const cleanedText = text.replace(/\[EOS\]/g, "");
  return cleanedText.replace(pattern, replaceMatch).trim();
}
function getPoint(startBox, size) {
  const [x, y] = JSON.parse(startBox);
  return [x * size.width, y * size.height];
}
async function resizeImageForUiTars(imageBase64, size) {
  if (vlLocateMode6() === "vlm-ui-tars" && uiTarsModelVersion2() === UITarsModelVersion.V1_5) {
    debug("ui-tars-v1.5, will check image size", size);
    const currentPixels = size.width * size.height;
    const maxPixels = 16384 * 28 * 28;
    if (currentPixels > maxPixels) {
      const resizeFactor = Math.sqrt(maxPixels / currentPixels);
      const newWidth = Math.floor(size.width * resizeFactor);
      const newHeight = Math.floor(size.height * resizeFactor);
      debug(
        "resize image for ui-tars, new width: %s, new height: %s",
        newWidth,
        newHeight
      );
      const resizedImage = await resizeImgBase64(imageBase64, {
        width: newWidth,
        height: newHeight
      });
      return resizedImage;
    }
  }
  return imageBase64;
}

export {
  systemPromptToLocateElement,
  elementByPositionWithElementInfo,
  describeUserPage,
  call2 as call,
  callToGetJSONObject,
  callAiFnWithStringResponse,
  AIActionType,
  callAiFn,
  adaptBboxToRect,
  expandSearchArea,
  generateYamlTest,
  generateYamlTestStream,
  generatePlaywrightTest,
  generatePlaywrightTestStream,
  AiLocateElement,
  AiLocateSection,
  AiExtractElementInfo,
  AiAssert,
  plan,
  vlmPlanning,
  resizeImageForUiTars
};
