// @midscene/web/dist/lib/playwright-report.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/playwright/reporter/index.ts
var reporter_exports = {};
__export(reporter_exports, {
  default: () => reporter_default
});
module.exports = __toCommonJS(reporter_exports);

// src/common/utils.ts
var import_ai_model = require("@midscene/core/ai-model");
var import_utils4 = require("@midscene/core/utils");
var import_env2 = require("@midscene/shared/env");
var import_extractor = require("@midscene/shared/extractor");
var import_img = require("@midscene/shared/img");
var import_logger2 = require("@midscene/shared/logger");
var import_utils5 = require("@midscene/shared/utils");
var import_dayjs = __toESM(require("dayjs"));

// src/common/task-cache.ts
var import_node_assert = __toESM(require("assert"));
var import_node_fs = require("fs");
var import_node_path = require("path");
var import_node_util = require("util");
var import_common = require("@midscene/shared/common");
var import_env = require("@midscene/shared/env");
var import_logger = require("@midscene/shared/logger");
var import_utils = require("@midscene/shared/utils");
var import_utils2 = require("@midscene/shared/utils");
var import_js_yaml = __toESM(require("js-yaml"));
var import_semver = __toESM(require("semver"));
var debug = (0, import_logger.getDebug)("cache");

// src/common/utils.ts
var debug2 = (0, import_logger2.getDebug)("tool:profile");
function getReportFileName(tag = "web") {
  const reportTagName = (0, import_env2.getAIConfig)(import_env2.MIDSCENE_REPORT_TAG_NAME);
  const dateTimeInFileName = (0, import_dayjs.default)().format("YYYY-MM-DD_HH-mm-ss");
  const uniqueId = (0, import_utils5.uuid)().substring(0, 8);
  return `${reportTagName || tag}-${dateTimeInFileName}-${uniqueId}`;
}
function printReportMsg(filepath) {
  (0, import_utils5.logMsg)(`Midscene - report file updated: ${filepath}`);
}
function replaceIllegalPathCharsAndSpace(str) {
  return str.replace(/[:*?"<>| ]/g, "-");
}

// src/playwright/reporter/index.ts
var import_utils7 = require("@midscene/core/utils");
function logger(...message) {
  if (process.env.DEBUG === "true") {
    console.log("Midscene e2e report:", ...message);
  }
}
var MidsceneReporter = class _MidsceneReporter {
  constructor(options = {}) {
    this.testTitleToFilename = /* @__PURE__ */ new Map();
    this.mode = _MidsceneReporter.getMode(options.type ?? "merged");
  }
  static getMode(reporterType) {
    if (!reporterType) {
      return "merged";
    }
    if (reporterType !== "merged" && reporterType !== "separate") {
      throw new Error(
        `Unknown reporter type in playwright config: ${reporterType}, only support 'merged' or 'separate'`
      );
    }
    return reporterType;
  }
  getSeparatedFilename(testTitle) {
    if (!this.testTitleToFilename.has(testTitle)) {
      const baseTag = `playwright-${replaceIllegalPathCharsAndSpace(testTitle)}`;
      const generatedFilename = getReportFileName(baseTag);
      this.testTitleToFilename.set(testTitle, generatedFilename);
    }
    return this.testTitleToFilename.get(testTitle);
  }
  getReportFilename(testTitle) {
    if (this.mode === "merged") {
      if (!this.mergedFilename) {
        this.mergedFilename = getReportFileName("playwright-merged");
      }
      return this.mergedFilename;
    } else if (this.mode === "separate") {
      if (!testTitle)
        throw new Error("testTitle is required in separate mode");
      return this.getSeparatedFilename(testTitle);
    }
    throw new Error(`Unknown mode: ${this.mode}`);
  }
  updateReport(testData) {
    if (!testData || !this.mode)
      return;
    const fileName = this.getReportFilename(
      testData.attributes?.playwright_test_title
    );
    const reportPath = (0, import_utils7.writeDumpReport)(
      fileName,
      testData,
      this.mode === "merged"
    );
    reportPath && printReportMsg(reportPath);
  }
  async onBegin(config, suite) {
  }
  onTestBegin(_test, _result) {
  }
  onTestEnd(test, result) {
    const dumpAnnotation = test.annotations.find((annotation) => {
      return annotation.type === "MIDSCENE_DUMP_ANNOTATION";
    });
    if (!dumpAnnotation?.description)
      return;
    const retry = result.retry ? `(retry #${result.retry})` : "";
    const testId = `${test.id}${retry}`;
    const testData = {
      dumpString: dumpAnnotation.description,
      attributes: {
        playwright_test_id: testId,
        playwright_test_title: `${test.title}${retry}`,
        playwright_test_status: result.status,
        playwright_test_duration: result.duration
      }
    };
    this.updateReport(testData);
    test.annotations = test.annotations.filter(
      (annotation) => annotation.type !== "MIDSCENE_DUMP_ANNOTATION"
    );
  }
  onEnd(result) {
    logger(`Finished the run: ${result.status}`);
  }
};
var reporter_default = MidsceneReporter;
