// @midscene/web/dist/lib/ui-utils.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/common/ui-utils.ts
var ui_utils_exports = {};
__export(ui_utils_exports, {
  getKeyCommands: () => getKeyCommands,
  limitOpenNewTabScript: () => limitOpenNewTabScript,
  locateParamStr: () => locateParamStr,
  paramStr: () => paramStr,
  pullParamStr: () => pullParamStr,
  scrollParamStr: () => scrollParamStr,
  taskTitleStr: () => taskTitleStr,
  typeStr: () => typeStr
});
module.exports = __toCommonJS(ui_utils_exports);
function typeStr(task) {
  return task.subType && task.subType !== "Plan" ? `${task.type} / ${task.subType || ""}` : task.type;
}
function getKeyCommands(value) {
  const keys = Array.isArray(value) ? value : [value];
  return keys.reduce((acc, k) => {
    const includeMeta = keys.includes("Meta") || keys.includes("Control");
    if (includeMeta && (k === "a" || k === "A")) {
      return acc.concat([{ key: k, command: "SelectAll" }]);
    }
    if (includeMeta && (k === "c" || k === "C")) {
      return acc.concat([{ key: k, command: "Copy" }]);
    }
    if (includeMeta && (k === "v" || k === "V")) {
      return acc.concat([{ key: k, command: "Paste" }]);
    }
    return acc.concat([{ key: k }]);
  }, []);
}
function locateParamStr(locate) {
  if (!locate) {
    return "";
  }
  if (typeof locate === "string") {
    return locate;
  }
  return typeof locate.prompt === "string" ? locate.prompt : locate.prompt.prompt;
}
function scrollParamStr(scrollParam) {
  if (!scrollParam) {
    return "";
  }
  return `${scrollParam.direction || "down"}, ${scrollParam.scrollType || "once"}, ${scrollParam.distance || "distance-not-set"}`;
}
function pullParamStr(pullParam) {
  if (!pullParam) {
    return "";
  }
  const parts = [];
  parts.push(`direction: ${pullParam.direction || "down"}`);
  if (pullParam.startPoint) {
    parts.push(`start: (${pullParam.startPoint.x}, ${pullParam.startPoint.y})`);
  }
  if (pullParam.distance) {
    parts.push(`distance: ${pullParam.distance}`);
  }
  if (pullParam.duration) {
    parts.push(`duration: ${pullParam.duration}ms`);
  }
  return parts.join(", ");
}
function taskTitleStr(type, prompt) {
  if (prompt) {
    return `${type} - ${prompt}`;
  }
  return type;
}
function paramStr(task) {
  let value;
  if (task.type === "Planning") {
    value = task?.param?.userInstruction;
  }
  if (task.type === "Insight") {
    value = task?.param?.prompt || task?.param?.id || task?.param?.dataDemand || task?.param?.assertion;
  }
  if (task.type === "Action") {
    const locate = task?.locate;
    const locateStr = locate ? locateParamStr(locate) : "";
    value = task.thought || "";
    if (typeof task?.param?.timeMs === "number") {
      value = `${task?.param?.timeMs}ms`;
    } else if (typeof task?.param?.scrollType === "string") {
      value = scrollParamStr(task?.param);
    } else if (typeof task?.param?.direction === "string" && task?.subType === "AndroidPull") {
      value = pullParamStr(task?.param);
    } else if (typeof task?.param?.value !== "undefined") {
      value = task?.param?.value;
    }
    if (locateStr) {
      if (value) {
        value = `${locateStr} - ${value}`;
      } else {
        value = locateStr;
      }
    }
  }
  if (typeof value === "undefined")
    return "";
  return typeof value === "string" ? value : JSON.stringify(value, void 0, 2);
}
var limitOpenNewTabScript = `
if (!window.__MIDSCENE_NEW_TAB_INTERCEPTOR_INITIALIZED__) {
  window.__MIDSCENE_NEW_TAB_INTERCEPTOR_INITIALIZED__ = true;

  // Intercept the window.open method (only once)
  window.open = function(url) {
    console.log('Blocked window.open:', url);
    window.location.href = url;
    return null;
  };

  // Block all a tag clicks with target="_blank" (only once)
  document.addEventListener('click', function(e) {
    const target = e.target.closest('a');
    if (target && target.target === '_blank') {
      e.preventDefault();
      console.log('Blocked new tab:', target.href);
      window.location.href = target.href;
      target.removeAttribute('target');
    }
  }, true);
}
`;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getKeyCommands,
  limitOpenNewTabScript,
  locateParamStr,
  paramStr,
  pullParamStr,
  scrollParamStr,
  taskTitleStr,
  typeStr
});
