// @midscene/web/dist/lib/utils.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/common/utils.ts
var utils_exports = {};
__export(utils_exports, {
  ERROR_CODE_NOT_IMPLEMENTED_AS_DESIGNED: () => ERROR_CODE_NOT_IMPLEMENTED_AS_DESIGNED,
  forceClosePopup: () => forceClosePopup,
  generateCacheId: () => generateCacheId,
  getCurrentExecutionFile: () => getCurrentExecutionFile,
  getReportFileName: () => getReportFileName,
  matchElementFromCache: () => matchElementFromCache,
  matchElementFromPlan: () => matchElementFromPlan,
  parseContextFromWebPage: () => parseContextFromWebPage,
  parsePrompt: () => parsePrompt,
  printReportMsg: () => printReportMsg,
  replaceIllegalPathCharsAndSpace: () => replaceIllegalPathCharsAndSpace,
  trimContextByViewport: () => trimContextByViewport
});
module.exports = __toCommonJS(utils_exports);
var import_ai_model = require("@midscene/core/ai-model");
var import_utils4 = require("@midscene/core/utils");
var import_env2 = require("@midscene/shared/env");
var import_extractor = require("@midscene/shared/extractor");
var import_img = require("@midscene/shared/img");
var import_logger2 = require("@midscene/shared/logger");
var import_utils5 = require("@midscene/shared/utils");
var import_dayjs = __toESM(require("dayjs"));

// src/web-element.ts
var WebElementInfo = class {
  constructor({
    content,
    rect,
    id,
    attributes,
    indexId,
    xpaths,
    isVisible
  }) {
    this.content = content;
    this.rect = rect;
    this.center = [
      Math.floor(rect.left + rect.width / 2),
      Math.floor(rect.top + rect.height / 2)
    ];
    this.id = id;
    this.attributes = attributes;
    this.indexId = indexId;
    this.xpaths = xpaths;
    this.isVisible = isVisible;
  }
};

// src/common/task-cache.ts
var import_node_assert = __toESM(require("assert"));
var import_node_fs = require("fs");
var import_node_path = require("path");
var import_node_util = require("util");
var import_common = require("@midscene/shared/common");
var import_env = require("@midscene/shared/env");
var import_logger = require("@midscene/shared/logger");
var import_utils = require("@midscene/shared/utils");
var import_utils2 = require("@midscene/shared/utils");
var import_js_yaml = __toESM(require("js-yaml"));
var import_semver = __toESM(require("semver"));
var debug = (0, import_logger.getDebug)("cache");

// src/common/utils.ts
var debug2 = (0, import_logger2.getDebug)("tool:profile");
async function parseContextFromWebPage(page, _opt) {
  (0, import_utils5.assert)(page, "page is required");
  if (page._forceUsePageContext) {
    return await page._forceUsePageContext();
  }
  debug2("Getting page URL");
  const url = await page.url();
  debug2("URL end");
  debug2("Uploading test info to server");
  (0, import_utils4.uploadTestInfoToServer)({ testUrl: url });
  debug2("UploadTestInfoToServer end");
  let screenshotBase64;
  let tree;
  debug2("Starting parallel operations: screenshot and element tree");
  await Promise.all([
    page.screenshotBase64().then((base64) => {
      screenshotBase64 = base64;
      debug2("ScreenshotBase64 end");
    }),
    page.getElementsNodeTree().then(async (treeRoot) => {
      tree = treeRoot;
      debug2("GetElementsNodeTree end");
    })
  ]);
  debug2("ParseContextFromWebPage end");
  debug2("Traversing element tree");
  const webTree = (0, import_extractor.traverseTree)(tree, (elementInfo) => {
    const { rect, id, content, attributes, indexId, isVisible } = elementInfo;
    return new WebElementInfo({
      rect,
      id,
      content,
      attributes,
      indexId,
      isVisible
    });
  });
  debug2("TraverseTree end");
  (0, import_utils5.assert)(screenshotBase64, "screenshotBase64 is required");
  const size = await page.size();
  debug2(`size: ${size.width}x${size.height} dpr: ${size.dpr}`);
  if (size.dpr && size.dpr > 1) {
    debug2("Resizing screenshot for high DPR display");
    screenshotBase64 = await (0, import_img.resizeImgBase64)(screenshotBase64, {
      width: size.width,
      height: size.height
    });
    debug2("ResizeImgBase64 end");
  }
  return {
    tree: webTree,
    size,
    screenshotBase64,
    url
  };
}
function getReportFileName(tag = "web") {
  const reportTagName = (0, import_env2.getAIConfig)(import_env2.MIDSCENE_REPORT_TAG_NAME);
  const dateTimeInFileName = (0, import_dayjs.default)().format("YYYY-MM-DD_HH-mm-ss");
  const uniqueId = (0, import_utils5.uuid)().substring(0, 8);
  return `${reportTagName || tag}-${dateTimeInFileName}-${uniqueId}`;
}
function printReportMsg(filepath) {
  (0, import_utils5.logMsg)(`Midscene - report file updated: ${filepath}`);
}
function getCurrentExecutionFile(trace) {
  const error = new Error();
  const stackTrace = trace || error.stack;
  const pkgDir = process.cwd() || "";
  if (stackTrace) {
    const stackLines = stackTrace.split("\n");
    for (const line of stackLines) {
      if (line.includes(".spec.") || line.includes(".test.") || line.includes(".ts") || line.includes(".js")) {
        const match = line.match(/(?:at\s+)?(.*?\.(?:spec|test)\.[jt]s)/);
        if (match?.[1]) {
          const targetFileName = match[1].replace(pkgDir, "").trim().replace("at ", "");
          return targetFileName;
        }
      }
    }
  }
  return false;
}
var testFileIndex = /* @__PURE__ */ new Map();
function generateCacheId(fileName) {
  let taskFile = fileName || getCurrentExecutionFile();
  if (!taskFile) {
    taskFile = (0, import_utils5.uuid)();
    console.warn(
      "Midscene - using random UUID for cache id. Cache may be invalid."
    );
  }
  if (testFileIndex.has(taskFile)) {
    const currentIndex = testFileIndex.get(taskFile);
    if (currentIndex !== void 0) {
      testFileIndex.set(taskFile, currentIndex + 1);
    }
  } else {
    testFileIndex.set(taskFile, 1);
  }
  return `${taskFile}-${testFileIndex.get(taskFile)}`;
}
var ERROR_CODE_NOT_IMPLEMENTED_AS_DESIGNED = "NOT_IMPLEMENTED_AS_DESIGNED";
function replaceIllegalPathCharsAndSpace(str) {
  return str.replace(/[:*?"<>| ]/g, "-");
}
function forceClosePopup(page, debug3) {
  page.on("popup", async (popup) => {
    if (!popup) {
      console.warn("got a popup event, but the popup is not ready yet, skip");
      return;
    }
    const url = await popup.url();
    console.log(`Popup opened: ${url}`);
    if (!popup.isClosed()) {
      try {
        await popup.close();
      } catch (error) {
        debug3(`failed to close popup ${url}, error: ${error}`);
      }
    } else {
      debug3(`popup is already closed, skip close ${url}`);
    }
    if (!page.isClosed()) {
      try {
        await page.goto(url);
      } catch (error) {
        debug3(`failed to goto ${url}, error: ${error}`);
      }
    } else {
      debug3(`page is already closed, skip goto ${url}`);
    }
  });
}
function matchElementFromPlan(planLocateParam, tree) {
  if (!planLocateParam) {
    return void 0;
  }
  if (planLocateParam.id) {
    return (0, import_extractor.getNodeFromCacheList)(planLocateParam.id);
  }
  if (planLocateParam.bbox) {
    const centerPosition = {
      x: Math.floor((planLocateParam.bbox[0] + planLocateParam.bbox[2]) / 2),
      y: Math.floor((planLocateParam.bbox[1] + planLocateParam.bbox[3]) / 2)
    };
    let element = (0, import_ai_model.elementByPositionWithElementInfo)(tree, centerPosition);
    if (!element) {
      element = (0, import_extractor.generateElementByPosition)(centerPosition);
    }
    return element;
  }
  return void 0;
}
async function matchElementFromCache(taskExecutor, xpaths, cachePrompt, cacheable) {
  try {
    if (xpaths?.length && taskExecutor.taskCache?.isCacheResultUsed && cacheable !== false) {
      for (let i = 0; i < xpaths.length; i++) {
        const element = await taskExecutor.page.getElementInfoByXpath(
          xpaths[i]
        );
        if (element?.id) {
          debug("cache hit, prompt: %s", cachePrompt);
          debug(
            "found a new element with same xpath, xpath: %s, id: %s",
            xpaths[i],
            element?.id
          );
          return element;
        }
      }
    }
  } catch (error) {
    debug("get element info by xpath error: ", error);
  }
}
function trimContextByViewport(execution) {
  function filterVisibleTree(node) {
    if (!node)
      return null;
    const filteredChildren = Array.isArray(node.children) ? node.children.map(filterVisibleTree).filter((child) => child !== null) : [];
    if (node.node && node.node.isVisible === true) {
      return {
        ...node,
        children: filteredChildren
      };
    }
    if (filteredChildren.length > 0) {
      return {
        node: null,
        children: filteredChildren
      };
    }
    return null;
  }
  return {
    ...execution,
    tasks: Array.isArray(execution.tasks) ? execution.tasks.map((task) => {
      const newTask = { ...task };
      if (task.pageContext?.tree) {
        newTask.pageContext = {
          ...task.pageContext,
          tree: filterVisibleTree(task.pageContext.tree) || {
            node: null,
            children: []
          }
        };
      }
      return newTask;
    }) : execution.tasks
  };
}
var parsePrompt = (prompt) => {
  if (typeof prompt === "string") {
    return {
      textPrompt: prompt,
      multimodalPrompt: void 0
    };
  }
  return {
    textPrompt: prompt.prompt,
    multimodalPrompt: prompt.images ? {
      images: prompt.images,
      convertHttpImage2Base64: !!prompt.convertHttpImage2Base64
    } : void 0
  };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ERROR_CODE_NOT_IMPLEMENTED_AS_DESIGNED,
  forceClosePopup,
  generateCacheId,
  getCurrentExecutionFile,
  getReportFileName,
  matchElementFromCache,
  matchElementFromPlan,
  parseContextFromWebPage,
  parsePrompt,
  printReportMsg,
  replaceIllegalPathCharsAndSpace,
  trimContextByViewport
});
