// @midscene/web/dist/es/midscene-server.js
// src/playground/server.ts
import { randomUUID } from "crypto";
import { existsSync as existsSync2, readFileSync as readFileSync2, writeFileSync as writeFileSync2 } from "fs";
import { join as join2 } from "path";

// src/common/utils.ts
import { elementByPositionWithElementInfo } from "@midscene/core/ai-model";
import { uploadTestInfoToServer } from "@midscene/core/utils";
import { MIDSCENE_REPORT_TAG_NAME, getAIConfig } from "@midscene/shared/env";
import {
  generateElementByPosition,
  getNodeFromCacheList,
  traverseTree
} from "@midscene/shared/extractor";
import { resizeImgBase64 } from "@midscene/shared/img";
import { getDebug as getDebug2 } from "@midscene/shared/logger";
import { assert as assert2, logMsg, uuid } from "@midscene/shared/utils";
import dayjs from "dayjs";

// src/common/task-cache.ts
import assert from "assert";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "fs";
import { dirname, join } from "path";
import { isDeepStrictEqual } from "util";
import { getMidsceneRunSubDir } from "@midscene/shared/common";
import {
  MIDSCENE_CACHE_MAX_FILENAME_LENGTH,
  getAIConfigInNumber
} from "@midscene/shared/env";
import { getDebug } from "@midscene/shared/logger";
import { ifInBrowser, ifInWorker } from "@midscene/shared/utils";
import { generateHashId } from "@midscene/shared/utils";
import yaml from "js-yaml";
import semver from "semver";
var debug = getDebug("cache");

// src/common/utils.ts
var debug2 = getDebug2("tool:profile");
var ERROR_CODE_NOT_IMPLEMENTED_AS_DESIGNED = "NOT_IMPLEMENTED_AS_DESIGNED";

// src/playground/server.ts
import { getTmpDir } from "@midscene/core/utils";
import { PLAYGROUND_SERVER_PORT } from "@midscene/shared/constants";
import { overrideAIConfig } from "@midscene/shared/env";
import { ifInBrowser as ifInBrowser2, ifInWorker as ifInWorker2 } from "@midscene/shared/utils";
import cors from "cors";
import dotenv from "dotenv";
import express from "express";
var defaultPort = PLAYGROUND_SERVER_PORT;
var errorHandler = (err, req, res, next) => {
  console.error(err);
  res.status(500).json({
    error: err.message
  });
};
var setup = async () => {
  if (!ifInBrowser2 && !ifInWorker2) {
    const { parsed } = dotenv.config();
    if (parsed) {
      overrideAIConfig(parsed);
    }
  }
};
var PlaygroundServer = class {
  constructor(pageClass, agentClass, staticPath) {
    this.app = express();
    this.tmpDir = getTmpDir();
    this.pageClass = pageClass;
    this.agentClass = agentClass;
    this.staticPath = staticPath;
    this.taskProgressTips = {};
    this.activeAgents = {};
    setup();
  }
  filePathForUuid(uuid2) {
    return join2(this.tmpDir, `${uuid2}.json`);
  }
  saveContextFile(uuid2, context) {
    const tmpFile = this.filePathForUuid(uuid2);
    console.log(`save context file: ${tmpFile}`);
    writeFileSync2(tmpFile, context);
    return tmpFile;
  }
  async launch(port) {
    this.port = port || defaultPort;
    this.app.use(errorHandler);
    this.app.use(
      cors({
        origin: "*",
        credentials: true
      })
    );
    this.app.get("/status", cors(), async (req, res) => {
      res.send({
        status: "ok"
      });
    });
    this.app.get("/context/:uuid", async (req, res) => {
      const { uuid: uuid2 } = req.params;
      const contextFile = this.filePathForUuid(uuid2);
      if (!existsSync2(contextFile)) {
        return res.status(404).json({
          error: "Context not found"
        });
      }
      const context = readFileSync2(contextFile, "utf8");
      res.json({
        context
      });
    });
    this.app.get("/task-progress/:requestId", cors(), async (req, res) => {
      const { requestId } = req.params;
      res.json({
        tip: this.taskProgressTips[requestId] || ""
      });
    });
    this.app.post(
      "/playground-with-context",
      express.json({ limit: "50mb" }),
      async (req, res) => {
        const context = req.body.context;
        if (!context) {
          return res.status(400).json({
            error: "context is required"
          });
        }
        const uuid2 = randomUUID();
        this.saveContextFile(uuid2, context);
        return res.json({
          location: `/playground/${uuid2}`,
          uuid: uuid2
        });
      }
    );
    this.app.post(
      "/execute",
      express.json({ limit: "30mb" }),
      async (req, res) => {
        const { context, type, prompt, requestId, deepThink } = req.body;
        if (!context) {
          return res.status(400).json({
            error: "context is required"
          });
        }
        if (!type) {
          return res.status(400).json({
            error: "type is required"
          });
        }
        if (!prompt) {
          return res.status(400).json({
            error: "prompt is required"
          });
        }
        const page = new this.pageClass(context);
        const agent = new this.agentClass(page);
        if (requestId) {
          this.taskProgressTips[requestId] = "";
          this.activeAgents[requestId] = agent;
          agent.onTaskStartTip = (tip) => {
            this.taskProgressTips[requestId] = tip;
          };
        }
        const response = {
          result: null,
          dump: null,
          error: null,
          reportHTML: null,
          requestId
        };
        const startTime = Date.now();
        try {
          if (type === "aiQuery") {
            response.result = await agent.aiQuery(prompt);
          } else if (type === "aiAction") {
            response.result = await agent.aiAction(prompt);
          } else if (type === "aiAssert") {
            response.result = await agent.aiAssert(prompt, void 0, {
              keepRawResponse: true
            });
          } else if (type === "aiTap") {
            response.result = await agent.aiTap(prompt, {
              deepThink
            });
          } else {
            response.error = `Unknown type: ${type}`;
          }
        } catch (error) {
          if (!error.message.includes(ERROR_CODE_NOT_IMPLEMENTED_AS_DESIGNED)) {
            response.error = error.message;
          }
        }
        try {
          response.dump = JSON.parse(agent.dumpDataString());
          response.reportHTML = agent.reportHTMLString() || null;
          agent.writeOutActionDumps();
          agent.destroy();
        } catch (error) {
          console.error(
            `write out dump failed: requestId: ${requestId}, ${error.message}`
          );
        }
        res.send(response);
        const timeCost = Date.now() - startTime;
        if (response.error) {
          console.error(
            `handle request failed after ${timeCost}ms: requestId: ${requestId}, ${response.error}`
          );
        } else {
          console.log(
            `handle request done after ${timeCost}ms: requestId: ${requestId}`
          );
        }
        if (requestId && this.activeAgents[requestId]) {
          delete this.activeAgents[requestId];
        }
      }
    );
    this.app.get("/cancel/:requestId", async (req, res) => {
      const { requestId } = req.params;
      if (!requestId) {
        return res.status(400).json({
          error: "requestId is required"
        });
      }
      const agent = this.activeAgents[requestId];
      if (!agent) {
        return res.status(404).json({
          error: "No active agent found for this requestId"
        });
      }
      try {
        await agent.destroy();
        delete this.activeAgents[requestId];
        res.json({ status: "cancelled" });
      } catch (error) {
        console.error(`Failed to cancel agent: ${error.message}`);
        res.status(500).json({
          error: `Failed to cancel: ${error.message}`
        });
      }
    });
    this.app.post(
      "/config",
      express.json({ limit: "1mb" }),
      async (req, res) => {
        const { aiConfig } = req.body;
        if (!aiConfig || typeof aiConfig !== "object") {
          return res.status(400).json({
            error: "aiConfig is required and must be an object"
          });
        }
        try {
          overrideAIConfig(aiConfig);
          return res.json({
            status: "ok",
            message: "AI config updated successfully"
          });
        } catch (error) {
          console.error(`Failed to update AI config: ${error.message}`);
          return res.status(500).json({
            error: `Failed to update AI config: ${error.message}`
          });
        }
      }
    );
    if (this.staticPath) {
      this.app.get("/", (req, res) => {
        res.redirect("/index.html");
      });
      this.app.get("*", (req, res) => {
        const requestedPath = join2(this.staticPath, req.path);
        if (existsSync2(requestedPath)) {
          res.sendFile(requestedPath);
        } else {
          res.sendFile(join2(this.staticPath, "index.html"));
        }
      });
    }
    return new Promise((resolve, reject) => {
      const port2 = this.port;
      this.server = this.app.listen(port2, () => {
        resolve(this);
      });
    });
  }
  close() {
    if (this.server) {
      return this.server.close();
    }
  }
};
export {
  PlaygroundServer as default
};
