// @midscene/web/dist/es/playwright-reporter.js
// src/common/utils.ts
import { elementByPositionWithElementInfo } from "@midscene/core/ai-model";
import { uploadTestInfoToServer } from "@midscene/core/utils";
import { MIDSCENE_REPORT_TAG_NAME, getAIConfig } from "@midscene/shared/env";
import {
  generateElementByPosition,
  getNodeFromCacheList,
  traverseTree
} from "@midscene/shared/extractor";
import { resizeImgBase64 } from "@midscene/shared/img";
import { getDebug as getDebug2 } from "@midscene/shared/logger";
import { assert as assert2, logMsg, uuid } from "@midscene/shared/utils";
import dayjs from "dayjs";

// src/common/task-cache.ts
import assert from "assert";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "fs";
import { dirname, join } from "path";
import { isDeepStrictEqual } from "util";
import { getMidsceneRunSubDir } from "@midscene/shared/common";
import {
  MIDSCENE_CACHE_MAX_FILENAME_LENGTH,
  getAIConfigInNumber
} from "@midscene/shared/env";
import { getDebug } from "@midscene/shared/logger";
import { ifInBrowser, ifInWorker } from "@midscene/shared/utils";
import { generateHashId } from "@midscene/shared/utils";
import yaml from "js-yaml";
import semver from "semver";
var debug = getDebug("cache");

// src/common/utils.ts
var debug2 = getDebug2("tool:profile");
function getReportFileName(tag = "web") {
  const reportTagName = getAIConfig(MIDSCENE_REPORT_TAG_NAME);
  const dateTimeInFileName = dayjs().format("YYYY-MM-DD_HH-mm-ss");
  const uniqueId = uuid().substring(0, 8);
  return `${reportTagName || tag}-${dateTimeInFileName}-${uniqueId}`;
}
function printReportMsg(filepath) {
  logMsg(`Midscene - report file updated: ${filepath}`);
}
function replaceIllegalPathCharsAndSpace(str) {
  return str.replace(/[:*?"<>| ]/g, "-");
}

// src/playwright/reporter/index.ts
import { writeDumpReport } from "@midscene/core/utils";
function logger(...message) {
  if (process.env.DEBUG === "true") {
    console.log("Midscene e2e report:", ...message);
  }
}
var MidsceneReporter = class _MidsceneReporter {
  constructor(options = {}) {
    this.testTitleToFilename = /* @__PURE__ */ new Map();
    this.mode = _MidsceneReporter.getMode(options.type ?? "merged");
  }
  static getMode(reporterType) {
    if (!reporterType) {
      return "merged";
    }
    if (reporterType !== "merged" && reporterType !== "separate") {
      throw new Error(
        `Unknown reporter type in playwright config: ${reporterType}, only support 'merged' or 'separate'`
      );
    }
    return reporterType;
  }
  getSeparatedFilename(testTitle) {
    if (!this.testTitleToFilename.has(testTitle)) {
      const baseTag = `playwright-${replaceIllegalPathCharsAndSpace(testTitle)}`;
      const generatedFilename = getReportFileName(baseTag);
      this.testTitleToFilename.set(testTitle, generatedFilename);
    }
    return this.testTitleToFilename.get(testTitle);
  }
  getReportFilename(testTitle) {
    if (this.mode === "merged") {
      if (!this.mergedFilename) {
        this.mergedFilename = getReportFileName("playwright-merged");
      }
      return this.mergedFilename;
    } else if (this.mode === "separate") {
      if (!testTitle)
        throw new Error("testTitle is required in separate mode");
      return this.getSeparatedFilename(testTitle);
    }
    throw new Error(`Unknown mode: ${this.mode}`);
  }
  updateReport(testData) {
    if (!testData || !this.mode)
      return;
    const fileName = this.getReportFilename(
      testData.attributes?.playwright_test_title
    );
    const reportPath = writeDumpReport(
      fileName,
      testData,
      this.mode === "merged"
    );
    reportPath && printReportMsg(reportPath);
  }
  async onBegin(config, suite) {
  }
  onTestBegin(_test, _result) {
  }
  onTestEnd(test, result) {
    const dumpAnnotation = test.annotations.find((annotation) => {
      return annotation.type === "MIDSCENE_DUMP_ANNOTATION";
    });
    if (!dumpAnnotation?.description)
      return;
    const retry = result.retry ? `(retry #${result.retry})` : "";
    const testId = `${test.id}${retry}`;
    const testData = {
      dumpString: dumpAnnotation.description,
      attributes: {
        playwright_test_id: testId,
        playwright_test_title: `${test.title}${retry}`,
        playwright_test_status: result.status,
        playwright_test_duration: result.duration
      }
    };
    this.updateReport(testData);
    test.annotations = test.annotations.filter(
      (annotation) => annotation.type !== "MIDSCENE_DUMP_ANNOTATION"
    );
  }
  onEnd(result) {
    logger(`Finished the run: ${result.status}`);
  }
};
var reporter_default = MidsceneReporter;
export {
  reporter_default as default
};
