// @midscene/web/dist/es/playwright.js
// src/common/agent.ts
import {
  Insight
} from "@midscene/core";
import yaml4 from "js-yaml";

// src/yaml/player.ts
import { existsSync, mkdirSync, writeFileSync } from "fs";
import { basename, dirname, join, resolve } from "path";
import { assert, ifInBrowser, ifInWorker } from "@midscene/shared/utils";
import { getMidsceneRunSubDir } from "@midscene/shared/common";
var ScriptPlayer = class {
  constructor(script, setupAgent, onTaskStatusChange, scriptPath) {
    this.script = script;
    this.setupAgent = setupAgent;
    this.onTaskStatusChange = onTaskStatusChange;
    this.taskStatusList = [];
    this.status = "init";
    this.unnamedResultIndex = 0;
    this.pageAgent = null;
    this.scriptPath = scriptPath;
    this.result = {};
    this.target = script.target || script.web || script.android;
    if (ifInBrowser || ifInWorker) {
      this.output = void 0;
    } else if (this.target?.output) {
      this.output = resolve(process.cwd(), this.target.output);
    } else {
      const scriptName = this.scriptPath ? basename(this.scriptPath, ".yaml").replace(/\.(ya?ml)$/i, "") : "script";
      this.output = join(
        getMidsceneRunSubDir("output"),
        `${scriptName}-${Date.now()}.json`
      );
    }
    if (ifInBrowser || ifInWorker) {
      this.unstableLogContent = void 0;
    } else if (typeof this.target?.unstableLogContent === "string") {
      this.unstableLogContent = resolve(
        process.cwd(),
        this.target.unstableLogContent
      );
    } else if (this.target?.unstableLogContent === true) {
      this.unstableLogContent = join(
        getMidsceneRunSubDir("output"),
        "unstableLogContent.json"
      );
    }
    this.taskStatusList = (script.tasks || []).map((task, taskIndex) => ({
      ...task,
      index: taskIndex,
      status: "init",
      totalSteps: task.flow?.length || 0
    }));
  }
  setResult(key, value) {
    const keyToUse = key || this.unnamedResultIndex++;
    if (this.result[keyToUse]) {
      console.warn(`result key ${keyToUse} already exists, will overwrite`);
    }
    this.result[keyToUse] = value;
    return this.flushResult();
  }
  setPlayerStatus(status, error) {
    this.status = status;
    this.errorInSetup = error;
  }
  notifyCurrentTaskStatusChange(taskIndex) {
    const taskIndexToNotify = typeof taskIndex === "number" ? taskIndex : this.currentTaskIndex;
    if (typeof taskIndexToNotify !== "number") {
      return;
    }
    const taskStatus = this.taskStatusList[taskIndexToNotify];
    if (this.onTaskStatusChange) {
      this.onTaskStatusChange(taskStatus);
    }
  }
  async setTaskStatus(index, statusValue, error) {
    this.taskStatusList[index].status = statusValue;
    if (error) {
      this.taskStatusList[index].error = error;
    }
    this.notifyCurrentTaskStatusChange(index);
  }
  setTaskIndex(taskIndex) {
    this.currentTaskIndex = taskIndex;
  }
  flushResult() {
    if (this.output) {
      const output = resolve(process.cwd(), this.output);
      const outputDir = dirname(output);
      if (!existsSync(outputDir)) {
        mkdirSync(outputDir, { recursive: true });
      }
      writeFileSync(output, JSON.stringify(this.result || {}, void 0, 2));
    }
  }
  flushUnstableLogContent() {
    if (this.unstableLogContent) {
      const content = this.pageAgent?._unstableLogContent();
      const filePath = resolve(process.cwd(), this.unstableLogContent);
      const outputDir = dirname(filePath);
      if (!existsSync(outputDir)) {
        mkdirSync(outputDir, { recursive: true });
      }
      writeFileSync(filePath, JSON.stringify(content, null, 2));
    }
  }
  async playTask(taskStatus, agent) {
    const { flow } = taskStatus;
    assert(flow, "missing flow in task");
    for (const flowItemIndex in flow) {
      const currentStep = Number.parseInt(flowItemIndex, 10);
      taskStatus.currentStep = currentStep;
      const flowItem = flow[flowItemIndex];
      if ("aiAction" in flowItem || "ai" in flowItem) {
        const actionTask = flowItem;
        const prompt = actionTask.aiAction || actionTask.ai;
        assert(prompt, "missing prompt for ai (aiAction)");
        assert(
          typeof prompt === "string",
          "prompt for aiAction must be a string"
        );
        await agent.aiAction(prompt, {
          cacheable: actionTask.cacheable
        });
      } else if ("aiAssert" in flowItem) {
        const assertTask = flowItem;
        const prompt = assertTask.aiAssert;
        const msg = assertTask.errorMessage;
        assert(prompt, "missing prompt for aiAssert");
        assert(
          typeof prompt === "string",
          "prompt for aiAssert must be a string"
        );
        await agent.aiAssert(prompt, msg);
      } else if ("aiQuery" in flowItem) {
        const queryTask = flowItem;
        const prompt = queryTask.aiQuery;
        const options = {
          domIncluded: queryTask.domIncluded,
          screenshotIncluded: queryTask.screenshotIncluded
        };
        assert(prompt, "missing prompt for aiQuery");
        assert(
          typeof prompt === "string",
          "prompt for aiQuery must be a string"
        );
        const queryResult = await agent.aiQuery(prompt, options);
        this.setResult(queryTask.name, queryResult);
      } else if ("aiNumber" in flowItem) {
        const numberTask = flowItem;
        const prompt = numberTask.aiNumber;
        const options = {
          domIncluded: numberTask.domIncluded,
          screenshotIncluded: numberTask.screenshotIncluded
        };
        assert(prompt, "missing prompt for aiNumber");
        assert(
          typeof prompt === "string",
          "prompt for number must be a string"
        );
        const numberResult = await agent.aiNumber(prompt, options);
        this.setResult(numberTask.name, numberResult);
      } else if ("aiString" in flowItem) {
        const stringTask = flowItem;
        const prompt = stringTask.aiString;
        const options = {
          domIncluded: stringTask.domIncluded,
          screenshotIncluded: stringTask.screenshotIncluded
        };
        assert(prompt, "missing prompt for aiString");
        assert(
          typeof prompt === "string",
          "prompt for string must be a string"
        );
        const stringResult = await agent.aiString(prompt, options);
        this.setResult(stringTask.name, stringResult);
      } else if ("aiBoolean" in flowItem) {
        const booleanTask = flowItem;
        const prompt = booleanTask.aiBoolean;
        const options = {
          domIncluded: booleanTask.domIncluded,
          screenshotIncluded: booleanTask.screenshotIncluded
        };
        assert(prompt, "missing prompt for aiBoolean");
        assert(
          typeof prompt === "string",
          "prompt for boolean must be a string"
        );
        const booleanResult = await agent.aiBoolean(prompt, options);
        this.setResult(booleanTask.name, booleanResult);
      } else if ("aiAsk" in flowItem) {
        const askTask = flowItem;
        const prompt = askTask.aiAsk;
        assert(prompt, "missing prompt for aiAsk");
        assert(typeof prompt === "string", "prompt for aiAsk must be a string");
        const askResult = await agent.aiAsk(prompt);
        this.setResult(askTask.name, askResult);
      } else if ("aiLocate" in flowItem) {
        const locateTask = flowItem;
        const prompt = locateTask.aiLocate;
        assert(prompt, "missing prompt for aiLocate");
        assert(
          typeof prompt === "string",
          "prompt for aiLocate must be a string"
        );
        const locateResult = await agent.aiLocate(prompt, locateTask);
        this.setResult(locateTask.name, locateResult);
      } else if ("aiWaitFor" in flowItem) {
        const waitForTask = flowItem;
        const prompt = waitForTask.aiWaitFor;
        assert(prompt, "missing prompt for aiWaitFor");
        assert(
          typeof prompt === "string",
          "prompt for aiWaitFor must be a string"
        );
        const timeout = waitForTask.timeout;
        await agent.aiWaitFor(prompt, { timeoutMs: timeout });
      } else if ("sleep" in flowItem) {
        const sleepTask = flowItem;
        const ms = sleepTask.sleep;
        let msNumber = ms;
        if (typeof ms === "string") {
          msNumber = Number.parseInt(ms, 10);
        }
        assert(
          msNumber && msNumber > 0,
          `ms for sleep must be greater than 0, but got ${ms}`
        );
        await new Promise((resolve2) => setTimeout(resolve2, msNumber));
      } else if ("aiTap" in flowItem) {
        const tapTask = flowItem;
        await agent.aiTap(tapTask.aiTap, tapTask);
      } else if ("aiRightClick" in flowItem) {
        const rightClickTask = flowItem;
        await agent.aiRightClick(rightClickTask.aiRightClick, rightClickTask);
      } else if ("aiHover" in flowItem) {
        const hoverTask = flowItem;
        await agent.aiHover(hoverTask.aiHover, hoverTask);
      } else if ("aiInput" in flowItem) {
        const inputTask = flowItem;
        await agent.aiInput(inputTask.aiInput, inputTask.locate, inputTask);
      } else if ("aiKeyboardPress" in flowItem) {
        const keyboardPressTask = flowItem;
        await agent.aiKeyboardPress(
          keyboardPressTask.aiKeyboardPress,
          keyboardPressTask.locate,
          keyboardPressTask
        );
      } else if ("aiScroll" in flowItem) {
        const scrollTask = flowItem;
        await agent.aiScroll(scrollTask, scrollTask.locate, scrollTask);
      } else if ("javascript" in flowItem) {
        const evaluateJavaScriptTask = flowItem;
        const result = await agent.evaluateJavaScript(
          evaluateJavaScriptTask.javascript
        );
        this.setResult(evaluateJavaScriptTask.name, result);
      } else if ("logScreenshot" in flowItem) {
        const logScreenshotTask = flowItem;
        await agent.logScreenshot(logScreenshotTask.logScreenshot, {
          content: logScreenshotTask.content || ""
        });
      } else {
        throw new Error(`unknown flowItem: ${JSON.stringify(flowItem)}`);
      }
    }
    this.reportFile = agent.reportFile;
    await this.flushUnstableLogContent();
  }
  async run() {
    const { target, web, android, tasks } = this.script;
    const webEnv = web || target;
    const androidEnv = android;
    const platform = webEnv || androidEnv;
    this.setPlayerStatus("running");
    let agent = null;
    let freeFn = [];
    try {
      const { agent: newAgent, freeFn: newFreeFn } = await this.setupAgent(
        platform
      );
      agent = newAgent;
      const originalOnTaskStartTip = agent.onTaskStartTip;
      agent.onTaskStartTip = (tip) => {
        if (this.status === "running") {
          this.agentStatusTip = tip;
        }
        originalOnTaskStartTip?.(tip);
      };
      freeFn = [
        ...newFreeFn || [],
        {
          name: "restore-agent-onTaskStartTip",
          fn: () => {
            if (agent) {
              agent.onTaskStartTip = originalOnTaskStartTip;
            }
          }
        }
      ];
    } catch (e) {
      this.setPlayerStatus("error", e);
      return;
    }
    this.pageAgent = agent;
    let taskIndex = 0;
    this.setPlayerStatus("running");
    let errorFlag = false;
    while (taskIndex < tasks.length) {
      const taskStatus = this.taskStatusList[taskIndex];
      this.setTaskStatus(taskIndex, "running");
      this.setTaskIndex(taskIndex);
      try {
        await this.playTask(taskStatus, this.pageAgent);
        this.setTaskStatus(taskIndex, "done");
      } catch (e) {
        this.setTaskStatus(taskIndex, "error", e);
        if (taskStatus.continueOnError) {
        } else {
          this.reportFile = agent.reportFile;
          errorFlag = true;
          break;
        }
      }
      this.reportFile = agent?.reportFile;
      taskIndex++;
    }
    if (errorFlag) {
      this.setPlayerStatus("error");
    } else {
      this.setPlayerStatus("done");
    }
    this.agentStatusTip = "";
    for (const fn of freeFn) {
      try {
        await fn.fn();
      } catch (e) {
      }
    }
  }
};

// src/yaml/builder.ts
import yaml from "js-yaml";

// src/yaml/utils.ts
import { assert as assert2 } from "@midscene/shared/utils";
import yaml2 from "js-yaml";
function interpolateEnvVars(content) {
  return content.replace(/\$\{([^}]+)\}/g, (_, envVar) => {
    const value = process.env[envVar.trim()];
    if (value === void 0) {
      throw new Error(`Environment variable "${envVar.trim()}" is not defined`);
    }
    return value;
  });
}
function parseYamlScript(content, filePath, ignoreCheckingTarget) {
  let processedContent = content;
  if (content.indexOf("android") !== -1 && content.match(/deviceId:\s*(\d+)/)) {
    let matchedDeviceId;
    processedContent = content.replace(
      /deviceId:\s*(\d+)/g,
      (match, deviceId) => {
        matchedDeviceId = deviceId;
        return `deviceId: '${deviceId}'`;
      }
    );
    console.warn(
      `please use string-style deviceId in yaml script, for example: deviceId: "${matchedDeviceId}"`
    );
  }
  const interpolatedContent = interpolateEnvVars(processedContent);
  const obj = yaml2.load(interpolatedContent, {
    schema: yaml2.JSON_SCHEMA
  });
  const pathTip = filePath ? `, failed to load ${filePath}` : "";
  const android = typeof obj.android !== "undefined" ? Object.assign({}, obj.android || {}) : void 0;
  const webConfig = obj.web || obj.target;
  const web = typeof webConfig !== "undefined" ? Object.assign({}, webConfig || {}) : void 0;
  if (!ignoreCheckingTarget) {
    assert2(
      web || android,
      `at least one of "target", "web", or "android" properties is required in yaml script${pathTip}`
    );
    assert2(
      web && !android || !web && android,
      `only one of "target", "web", or "android" properties is allowed in yaml script${pathTip}`
    );
    if (web || android) {
      assert2(
        typeof web === "object" || typeof android === "object",
        `property "target/web/android" must be an object${pathTip}`
      );
    }
  }
  assert2(obj.tasks, `property "tasks" is required in yaml script ${pathTip}`);
  assert2(
    Array.isArray(obj.tasks),
    `property "tasks" must be an array in yaml script, but got ${obj.tasks}`
  );
  return obj;
}

// src/common/agent.ts
import {
  groupedActionDumpFileExt,
  reportHTMLContent,
  stringifyDumpData,
  writeLogFile
} from "@midscene/core/utils";
import {
  DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT,
  DEFAULT_WAIT_FOR_NETWORK_IDLE_TIMEOUT
} from "@midscene/shared/constants";
import { getAIConfigInBoolean, vlLocateMode } from "@midscene/shared/env";
import { getDebug as getDebug5 } from "@midscene/shared/logger";
import { assert as assert7 } from "@midscene/shared/utils";

// src/common/tasks.ts
import {
  Executor,
  plan
} from "@midscene/core";
import {
  elementByPositionWithElementInfo as elementByPositionWithElementInfo2,
  resizeImageForUiTars,
  vlmPlanning
} from "@midscene/core/ai-model";
import { sleep } from "@midscene/core/utils";
import { NodeType } from "@midscene/shared/constants";
import {
  MIDSCENE_REPLANNING_CYCLE_LIMIT,
  getAIConfigInNumber as getAIConfigInNumber2
} from "@midscene/shared/env";
import { getDebug as getDebug3 } from "@midscene/shared/logger";
import { assert as assert5 } from "@midscene/shared/utils";

// src/common/ui-utils.ts
function typeStr(task) {
  return task.subType && task.subType !== "Plan" ? `${task.type} / ${task.subType || ""}` : task.type;
}
function getKeyCommands(value) {
  const keys = Array.isArray(value) ? value : [value];
  return keys.reduce((acc, k) => {
    const includeMeta = keys.includes("Meta") || keys.includes("Control");
    if (includeMeta && (k === "a" || k === "A")) {
      return acc.concat([{ key: k, command: "SelectAll" }]);
    }
    if (includeMeta && (k === "c" || k === "C")) {
      return acc.concat([{ key: k, command: "Copy" }]);
    }
    if (includeMeta && (k === "v" || k === "V")) {
      return acc.concat([{ key: k, command: "Paste" }]);
    }
    return acc.concat([{ key: k }]);
  }, []);
}
function locateParamStr(locate) {
  if (!locate) {
    return "";
  }
  if (typeof locate === "string") {
    return locate;
  }
  return typeof locate.prompt === "string" ? locate.prompt : locate.prompt.prompt;
}
function scrollParamStr(scrollParam) {
  if (!scrollParam) {
    return "";
  }
  return `${scrollParam.direction || "down"}, ${scrollParam.scrollType || "once"}, ${scrollParam.distance || "distance-not-set"}`;
}
function pullParamStr(pullParam) {
  if (!pullParam) {
    return "";
  }
  const parts = [];
  parts.push(`direction: ${pullParam.direction || "down"}`);
  if (pullParam.startPoint) {
    parts.push(`start: (${pullParam.startPoint.x}, ${pullParam.startPoint.y})`);
  }
  if (pullParam.distance) {
    parts.push(`distance: ${pullParam.distance}`);
  }
  if (pullParam.duration) {
    parts.push(`duration: ${pullParam.duration}ms`);
  }
  return parts.join(", ");
}
function taskTitleStr(type, prompt) {
  if (prompt) {
    return `${type} - ${prompt}`;
  }
  return type;
}
function paramStr(task) {
  let value;
  if (task.type === "Planning") {
    value = task?.param?.userInstruction;
  }
  if (task.type === "Insight") {
    value = task?.param?.prompt || task?.param?.id || task?.param?.dataDemand || task?.param?.assertion;
  }
  if (task.type === "Action") {
    const locate = task?.locate;
    const locateStr = locate ? locateParamStr(locate) : "";
    value = task.thought || "";
    if (typeof task?.param?.timeMs === "number") {
      value = `${task?.param?.timeMs}ms`;
    } else if (typeof task?.param?.scrollType === "string") {
      value = scrollParamStr(task?.param);
    } else if (typeof task?.param?.direction === "string" && task?.subType === "AndroidPull") {
      value = pullParamStr(task?.param);
    } else if (typeof task?.param?.value !== "undefined") {
      value = task?.param?.value;
    }
    if (locateStr) {
      if (value) {
        value = `${locateStr} - ${value}`;
      } else {
        value = locateStr;
      }
    }
  }
  if (typeof value === "undefined")
    return "";
  return typeof value === "string" ? value : JSON.stringify(value, void 0, 2);
}

// src/common/utils.ts
import { elementByPositionWithElementInfo } from "@midscene/core/ai-model";
import { uploadTestInfoToServer } from "@midscene/core/utils";
import { MIDSCENE_REPORT_TAG_NAME, getAIConfig } from "@midscene/shared/env";
import {
  generateElementByPosition,
  getNodeFromCacheList,
  traverseTree
} from "@midscene/shared/extractor";
import { resizeImgBase64 } from "@midscene/shared/img";
import { getDebug as getDebug2 } from "@midscene/shared/logger";
import { assert as assert4, logMsg, uuid } from "@midscene/shared/utils";
import dayjs from "dayjs";

// src/web-element.ts
var WebElementInfo = class {
  constructor({
    content,
    rect,
    id,
    attributes,
    indexId,
    xpaths,
    isVisible
  }) {
    this.content = content;
    this.rect = rect;
    this.center = [
      Math.floor(rect.left + rect.width / 2),
      Math.floor(rect.top + rect.height / 2)
    ];
    this.id = id;
    this.attributes = attributes;
    this.indexId = indexId;
    this.xpaths = xpaths;
    this.isVisible = isVisible;
  }
};

// src/common/task-cache.ts
import assert3 from "assert";
import { existsSync as existsSync2, mkdirSync as mkdirSync2, readFileSync, writeFileSync as writeFileSync2 } from "fs";
import { dirname as dirname2, join as join2 } from "path";
import { isDeepStrictEqual } from "util";
import { getMidsceneRunSubDir as getMidsceneRunSubDir2 } from "@midscene/shared/common";
import {
  MIDSCENE_CACHE_MAX_FILENAME_LENGTH,
  getAIConfigInNumber
} from "@midscene/shared/env";
import { getDebug } from "@midscene/shared/logger";
import { ifInBrowser as ifInBrowser2, ifInWorker as ifInWorker2 } from "@midscene/shared/utils";
import { generateHashId } from "@midscene/shared/utils";
import yaml3 from "js-yaml";
import semver from "semver";

// package.json
var version = "0.25.3";

// src/common/task-cache.ts
var DEFAULT_CACHE_MAX_FILENAME_LENGTH = 200;
var debug = getDebug("cache");
var lowestSupportedMidsceneVersion = "0.16.10";
var cacheFileExt = ".cache.yaml";
var TaskCache = class {
  // Track matched records
  constructor(cacheId, isCacheResultUsed, cacheFilePath) {
    this.matchedCacheIndices = /* @__PURE__ */ new Set();
    assert3(cacheId, "cacheId is required");
    let safeCacheId = replaceIllegalPathCharsAndSpace(cacheId);
    const cacheMaxFilenameLength = getAIConfigInNumber(MIDSCENE_CACHE_MAX_FILENAME_LENGTH) || DEFAULT_CACHE_MAX_FILENAME_LENGTH;
    if (Buffer.byteLength(safeCacheId, "utf8") > cacheMaxFilenameLength) {
      const prefix = safeCacheId.slice(0, 32);
      const hash = generateHashId(void 0, safeCacheId);
      safeCacheId = `${prefix}-${hash}`;
    }
    this.cacheId = safeCacheId;
    this.cacheFilePath = ifInBrowser2 || ifInWorker2 ? void 0 : cacheFilePath || join2(getMidsceneRunSubDir2("cache"), `${this.cacheId}${cacheFileExt}`);
    this.isCacheResultUsed = isCacheResultUsed;
    let cacheContent;
    if (this.cacheFilePath) {
      cacheContent = this.loadCacheFromFile();
    }
    if (!cacheContent) {
      cacheContent = {
        midsceneVersion: version,
        cacheId: this.cacheId,
        caches: []
      };
    }
    this.cache = cacheContent;
    this.cacheOriginalLength = this.cache.caches.length;
  }
  matchCache(prompt, type) {
    for (let i = 0; i < this.cacheOriginalLength; i++) {
      const item = this.cache.caches[i];
      const promptStr = typeof prompt === "string" ? prompt : JSON.stringify(prompt);
      const key = `${type}:${promptStr}:${i}`;
      if (item.type === type && isDeepStrictEqual(item.prompt, prompt) && !this.matchedCacheIndices.has(key)) {
        this.matchedCacheIndices.add(key);
        debug(
          "cache found and marked as used, type: %s, prompt: %s, index: %d",
          type,
          prompt,
          i
        );
        return {
          cacheContent: item,
          updateFn: (cb) => {
            debug(
              "will call updateFn to update cache, type: %s, prompt: %s, index: %d",
              type,
              prompt,
              i
            );
            cb(item);
            debug(
              "cache updated, will flush to file, type: %s, prompt: %s, index: %d",
              type,
              prompt,
              i
            );
            this.flushCacheToFile();
          }
        };
      }
    }
    debug("no unused cache found, type: %s, prompt: %s", type, prompt);
    return void 0;
  }
  matchPlanCache(prompt) {
    return this.matchCache(prompt, "plan");
  }
  matchLocateCache(prompt) {
    return this.matchCache(prompt, "locate");
  }
  appendCache(cache) {
    debug("will append cache", cache);
    this.cache.caches.push(cache);
    this.flushCacheToFile();
  }
  loadCacheFromFile() {
    const cacheFile = this.cacheFilePath;
    assert3(cacheFile, "cache file path is required");
    if (!existsSync2(cacheFile)) {
      debug("no cache file found, path: %s", cacheFile);
      return void 0;
    }
    const jsonTypeCacheFile = cacheFile.replace(cacheFileExt, ".json");
    if (existsSync2(jsonTypeCacheFile) && this.isCacheResultUsed) {
      console.warn(
        `An outdated cache file from an earlier version of Midscene has been detected. Since version 0.17, we have implemented an improved caching strategy. Please delete the old file located at: ${jsonTypeCacheFile}.`
      );
      return void 0;
    }
    try {
      const data = readFileSync(cacheFile, "utf8");
      const jsonData = yaml3.load(data);
      if (!version) {
        debug("no midscene version info, will not read cache from file");
        return void 0;
      }
      if (semver.lt(jsonData.midsceneVersion, lowestSupportedMidsceneVersion) && !jsonData.midsceneVersion.includes("beta")) {
        console.warn(
          `You are using an old version of Midscene cache file, and we cannot match any info from it. Starting from Midscene v0.17, we changed our strategy to use xpath for cache info, providing better performance.
Please delete the existing cache and rebuild it. Sorry for the inconvenience.
cache file: ${cacheFile}`
        );
        return void 0;
      }
      debug(
        "cache loaded from file, path: %s, cache version: %s, record length: %s",
        cacheFile,
        jsonData.midsceneVersion,
        jsonData.caches.length
      );
      jsonData.midsceneVersion = version;
      return jsonData;
    } catch (err) {
      debug(
        "cache file exists but load failed, path: %s, error: %s",
        cacheFile,
        err
      );
      return void 0;
    }
  }
  flushCacheToFile() {
    if (!version) {
      debug("no midscene version info, will not write cache to file");
      return;
    }
    if (!this.cacheFilePath) {
      debug("no cache file path, will not write cache to file");
      return;
    }
    try {
      const dir = dirname2(this.cacheFilePath);
      if (!existsSync2(dir)) {
        mkdirSync2(dir, { recursive: true });
        debug("created cache directory: %s", dir);
      }
      const sortedCaches = [...this.cache.caches].sort((a, b) => {
        if (a.type === "plan" && b.type === "locate")
          return -1;
        if (a.type === "locate" && b.type === "plan")
          return 1;
        return 0;
      });
      const cacheToWrite = {
        ...this.cache,
        caches: sortedCaches
      };
      const yamlData = yaml3.dump(cacheToWrite);
      writeFileSync2(this.cacheFilePath, yamlData);
      debug("cache flushed to file: %s", this.cacheFilePath);
    } catch (err) {
      debug(
        "write cache to file failed, path: %s, error: %s",
        this.cacheFilePath,
        err
      );
    }
  }
  updateOrAppendCacheRecord(newRecord, cachedRecord) {
    if (cachedRecord) {
      if (newRecord.type === "plan") {
        cachedRecord.updateFn((cache) => {
          cache.yamlWorkflow = newRecord.yamlWorkflow;
        });
      } else {
        cachedRecord.updateFn((cache) => {
          cache.xpaths = newRecord.xpaths;
        });
      }
    } else {
      this.appendCache(newRecord);
    }
  }
};

// src/common/utils.ts
var debug2 = getDebug2("tool:profile");
async function parseContextFromWebPage(page, _opt) {
  assert4(page, "page is required");
  if (page._forceUsePageContext) {
    return await page._forceUsePageContext();
  }
  debug2("Getting page URL");
  const url = await page.url();
  debug2("URL end");
  debug2("Uploading test info to server");
  uploadTestInfoToServer({ testUrl: url });
  debug2("UploadTestInfoToServer end");
  let screenshotBase64;
  let tree;
  debug2("Starting parallel operations: screenshot and element tree");
  await Promise.all([
    page.screenshotBase64().then((base64) => {
      screenshotBase64 = base64;
      debug2("ScreenshotBase64 end");
    }),
    page.getElementsNodeTree().then(async (treeRoot) => {
      tree = treeRoot;
      debug2("GetElementsNodeTree end");
    })
  ]);
  debug2("ParseContextFromWebPage end");
  debug2("Traversing element tree");
  const webTree = traverseTree(tree, (elementInfo) => {
    const { rect, id, content, attributes, indexId, isVisible } = elementInfo;
    return new WebElementInfo({
      rect,
      id,
      content,
      attributes,
      indexId,
      isVisible
    });
  });
  debug2("TraverseTree end");
  assert4(screenshotBase64, "screenshotBase64 is required");
  const size = await page.size();
  debug2(`size: ${size.width}x${size.height} dpr: ${size.dpr}`);
  if (size.dpr && size.dpr > 1) {
    debug2("Resizing screenshot for high DPR display");
    screenshotBase64 = await resizeImgBase64(screenshotBase64, {
      width: size.width,
      height: size.height
    });
    debug2("ResizeImgBase64 end");
  }
  return {
    tree: webTree,
    size,
    screenshotBase64,
    url
  };
}
function getReportFileName(tag = "web") {
  const reportTagName = getAIConfig(MIDSCENE_REPORT_TAG_NAME);
  const dateTimeInFileName = dayjs().format("YYYY-MM-DD_HH-mm-ss");
  const uniqueId = uuid().substring(0, 8);
  return `${reportTagName || tag}-${dateTimeInFileName}-${uniqueId}`;
}
function printReportMsg(filepath) {
  logMsg(`Midscene - report file updated: ${filepath}`);
}
function replaceIllegalPathCharsAndSpace(str) {
  return str.replace(/[:*?"<>| ]/g, "-");
}
function forceClosePopup(page, debug7) {
  page.on("popup", async (popup) => {
    if (!popup) {
      console.warn("got a popup event, but the popup is not ready yet, skip");
      return;
    }
    const url = await popup.url();
    console.log(`Popup opened: ${url}`);
    if (!popup.isClosed()) {
      try {
        await popup.close();
      } catch (error) {
        debug7(`failed to close popup ${url}, error: ${error}`);
      }
    } else {
      debug7(`popup is already closed, skip close ${url}`);
    }
    if (!page.isClosed()) {
      try {
        await page.goto(url);
      } catch (error) {
        debug7(`failed to goto ${url}, error: ${error}`);
      }
    } else {
      debug7(`page is already closed, skip goto ${url}`);
    }
  });
}
function matchElementFromPlan(planLocateParam, tree) {
  if (!planLocateParam) {
    return void 0;
  }
  if (planLocateParam.id) {
    return getNodeFromCacheList(planLocateParam.id);
  }
  if (planLocateParam.bbox) {
    const centerPosition = {
      x: Math.floor((planLocateParam.bbox[0] + planLocateParam.bbox[2]) / 2),
      y: Math.floor((planLocateParam.bbox[1] + planLocateParam.bbox[3]) / 2)
    };
    let element = elementByPositionWithElementInfo(tree, centerPosition);
    if (!element) {
      element = generateElementByPosition(centerPosition);
    }
    return element;
  }
  return void 0;
}
async function matchElementFromCache(taskExecutor, xpaths, cachePrompt, cacheable) {
  try {
    if (xpaths?.length && taskExecutor.taskCache?.isCacheResultUsed && cacheable !== false) {
      for (let i = 0; i < xpaths.length; i++) {
        const element = await taskExecutor.page.getElementInfoByXpath(
          xpaths[i]
        );
        if (element?.id) {
          debug("cache hit, prompt: %s", cachePrompt);
          debug(
            "found a new element with same xpath, xpath: %s, id: %s",
            xpaths[i],
            element?.id
          );
          return element;
        }
      }
    }
  } catch (error) {
    debug("get element info by xpath error: ", error);
  }
}
function trimContextByViewport(execution) {
  function filterVisibleTree(node) {
    if (!node)
      return null;
    const filteredChildren = Array.isArray(node.children) ? node.children.map(filterVisibleTree).filter((child) => child !== null) : [];
    if (node.node && node.node.isVisible === true) {
      return {
        ...node,
        children: filteredChildren
      };
    }
    if (filteredChildren.length > 0) {
      return {
        node: null,
        children: filteredChildren
      };
    }
    return null;
  }
  return {
    ...execution,
    tasks: Array.isArray(execution.tasks) ? execution.tasks.map((task) => {
      const newTask = { ...task };
      if (task.pageContext?.tree) {
        newTask.pageContext = {
          ...task.pageContext,
          tree: filterVisibleTree(task.pageContext.tree) || {
            node: null,
            children: []
          }
        };
      }
      return newTask;
    }) : execution.tasks
  };
}
var parsePrompt = (prompt) => {
  if (typeof prompt === "string") {
    return {
      textPrompt: prompt,
      multimodalPrompt: void 0
    };
  }
  return {
    textPrompt: prompt.prompt,
    multimodalPrompt: prompt.images ? {
      images: prompt.images,
      convertHttpImage2Base64: !!prompt.convertHttpImage2Base64
    } : void 0
  };
};

// src/common/tasks.ts
var debug3 = getDebug3("page-task-executor");
var defaultReplanningCycleLimit = 10;
var isAndroidPage = (page) => {
  return page.pageType === "android";
};
var PageTaskExecutor = class {
  constructor(page, insight, opts) {
    this.conversationHistory = [];
    this.page = page;
    this.insight = insight;
    this.taskCache = opts.taskCache;
    this.onTaskStartCallback = opts?.onTaskStart;
  }
  async recordScreenshot(timing) {
    const base64 = await this.page.screenshotBase64();
    const item = {
      type: "screenshot",
      ts: Date.now(),
      screenshot: base64,
      timing
    };
    return item;
  }
  async getElementXpath(pageContext, element) {
    let elementId = element?.id;
    if (element?.isOrderSensitive !== void 0) {
      const xpaths = await this.page.getXpathsByPoint(
        {
          left: element.center[0],
          top: element.center[1]
        },
        element?.isOrderSensitive
      );
      return xpaths;
    }
    if (element?.attributes?.nodeType === NodeType.POSITION) {
      await this.insight.contextRetrieverFn("locate");
      const info = elementByPositionWithElementInfo2(
        pageContext.tree,
        {
          x: element.center[0],
          y: element.center[1]
        },
        {
          requireStrictDistance: false,
          filterPositionElements: true
        }
      );
      if (info?.id) {
        elementId = info.id;
      } else {
        debug3(
          "no element id found for position node, will not update cache",
          element
        );
      }
    }
    if (!elementId) {
      return void 0;
    }
    try {
      const result = await this.page.getXpathsById(elementId);
      return result;
    } catch (error) {
      debug3("getXpathsById error: ", error);
    }
  }
  prependExecutorWithScreenshot(taskApply, appendAfterExecution = false) {
    const taskWithScreenshot = {
      ...taskApply,
      executor: async (param, context, ...args) => {
        const recorder = [];
        const { task } = context;
        task.recorder = recorder;
        const shot = await this.recordScreenshot(`before ${task.type}`);
        recorder.push(shot);
        const result = await taskApply.executor(param, context, ...args);
        if (taskApply.type === "Action") {
          await Promise.all([
            (async () => {
              await sleep(100);
              if (this.page.waitUntilNetworkIdle) {
                try {
                  await this.page.waitUntilNetworkIdle();
                } catch (error) {
                }
              }
            })(),
            sleep(200)
          ]);
        }
        if (appendAfterExecution) {
          const shot2 = await this.recordScreenshot("after Action");
          recorder.push(shot2);
        }
        return result;
      }
    };
    return taskWithScreenshot;
  }
  async convertPlanToExecutable(plans, opts) {
    const tasks = [];
    plans.forEach((plan2) => {
      if (plan2.type === "Locate") {
        if (plan2.locate === null || plan2.locate?.id === null || plan2.locate?.id === "null") {
          return;
        }
        const taskFind = {
          type: "Insight",
          subType: "Locate",
          param: plan2.locate ? {
            ...plan2.locate,
            cacheable: opts?.cacheable
          } : void 0,
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (param, taskContext) => {
            const { task } = taskContext;
            assert5(
              param?.prompt || param?.id || param?.bbox,
              "No prompt or id or position or bbox to locate"
            );
            let insightDump;
            let usage;
            const dumpCollector = (dump) => {
              insightDump = dump;
              usage = dump?.taskInfo?.usage;
              task.log = {
                dump: insightDump
              };
              task.usage = usage;
            };
            this.insight.onceDumpUpdatedFn = dumpCollector;
            const shotTime = Date.now();
            const pageContext = await this.insight.contextRetrieverFn("locate");
            task.pageContext = pageContext;
            const recordItem = {
              type: "screenshot",
              ts: shotTime,
              screenshot: pageContext.screenshotBase64,
              timing: "before Insight"
            };
            task.recorder = [recordItem];
            const elementFromXpath = param.xpath ? await this.page.getElementInfoByXpath(param.xpath) : void 0;
            const userExpectedPathHitFlag = !!elementFromXpath;
            const cachePrompt = param.prompt;
            const locateCacheRecord = this.taskCache?.matchLocateCache(cachePrompt);
            const xpaths = locateCacheRecord?.cacheContent?.xpaths;
            const elementFromCache = userExpectedPathHitFlag ? null : await matchElementFromCache(
              this,
              xpaths,
              cachePrompt,
              param.cacheable
            );
            const cacheHitFlag = !!elementFromCache;
            const elementFromPlan = !userExpectedPathHitFlag && !cacheHitFlag ? matchElementFromPlan(param, pageContext.tree) : void 0;
            const planHitFlag = !!elementFromPlan;
            const elementFromAiLocate = !userExpectedPathHitFlag && !cacheHitFlag && !planHitFlag ? (await this.insight.locate(param, {
              // fallback to ai locate
              context: pageContext
            })).element : void 0;
            const aiLocateHitFlag = !!elementFromAiLocate;
            const element = elementFromXpath || // highest priority
            elementFromCache || // second priority
            elementFromPlan || // third priority
            elementFromAiLocate;
            let currentXpaths;
            if (element && this.taskCache && !cacheHitFlag && param?.cacheable !== false) {
              const elementXpaths = await this.getElementXpath(
                pageContext,
                element
              );
              if (elementXpaths?.length) {
                currentXpaths = elementXpaths;
                this.taskCache.updateOrAppendCacheRecord(
                  {
                    type: "locate",
                    prompt: cachePrompt,
                    xpaths: elementXpaths
                  },
                  locateCacheRecord
                );
              } else {
                debug3(
                  "no xpaths found, will not update cache",
                  cachePrompt,
                  elementXpaths
                );
              }
            }
            if (!element) {
              throw new Error(`Element not found: ${param.prompt}`);
            }
            let hitBy;
            if (userExpectedPathHitFlag) {
              hitBy = {
                from: "User expected path",
                context: {
                  xpath: param.xpath
                }
              };
            } else if (cacheHitFlag) {
              hitBy = {
                from: "Cache",
                context: {
                  xpathsFromCache: xpaths,
                  xpathsToSave: currentXpaths
                }
              };
            } else if (planHitFlag) {
              hitBy = {
                from: "Planning",
                context: {
                  id: elementFromPlan?.id,
                  bbox: elementFromPlan?.bbox
                }
              };
            } else if (aiLocateHitFlag) {
              hitBy = {
                from: "AI model",
                context: {
                  prompt: param.prompt
                }
              };
            }
            return {
              output: {
                element
              },
              pageContext,
              hitBy
            };
          }
        };
        tasks.push(taskFind);
      } else if (plan2.type === "Assert" || plan2.type === "AssertWithoutThrow") {
        const assertPlan = plan2;
        const taskAssert = {
          type: "Insight",
          subType: "Assert",
          param: assertPlan.param,
          thought: assertPlan.thought,
          locate: assertPlan.locate,
          executor: async (param, taskContext) => {
            const { task } = taskContext;
            let insightDump;
            const dumpCollector = (dump) => {
              insightDump = dump;
            };
            this.insight.onceDumpUpdatedFn = dumpCollector;
            const shotTime = Date.now();
            const pageContext = await this.insight.contextRetrieverFn("assert");
            task.pageContext = pageContext;
            const recordItem = {
              type: "screenshot",
              ts: shotTime,
              screenshot: pageContext.screenshotBase64,
              timing: "before Insight"
            };
            task.recorder = [recordItem];
            const assertion = await this.insight.assert(
              assertPlan.param.assertion
            );
            if (!assertion.pass) {
              if (plan2.type === "Assert") {
                task.output = assertion;
                task.log = {
                  dump: insightDump
                };
                throw new Error(
                  assertion.thought || "Assertion failed without reason"
                );
              }
              task.error = new Error(assertion.thought);
            }
            return {
              output: assertion,
              pageContext,
              log: {
                dump: insightDump
              },
              usage: assertion.usage
            };
          }
        };
        tasks.push(taskAssert);
      } else if (plan2.type === "Input") {
        const taskActionInput = {
          type: "Action",
          subType: "Input",
          param: plan2.param,
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (taskParam, { element }) => {
            if (element) {
              await this.page.clearInput(element);
              if (!taskParam || !taskParam.value) {
                return;
              }
            }
            await this.page.keyboard.type(taskParam.value, {
              autoDismissKeyboard: taskParam.autoDismissKeyboard
            });
          }
        };
        tasks.push(taskActionInput);
      } else if (plan2.type === "KeyboardPress") {
        const taskActionKeyboardPress = {
          type: "Action",
          subType: "KeyboardPress",
          param: plan2.param,
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (taskParam) => {
            const keys = getKeyCommands(taskParam.value);
            await this.page.keyboard.press(keys);
          }
        };
        tasks.push(taskActionKeyboardPress);
      } else if (plan2.type === "Tap") {
        const taskActionTap = {
          type: "Action",
          subType: "Tap",
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (param, { element }) => {
            assert5(element, "Element not found, cannot tap");
            await this.page.mouse.click(element.center[0], element.center[1]);
          }
        };
        tasks.push(taskActionTap);
      } else if (plan2.type === "RightClick") {
        const taskActionRightClick = {
          type: "Action",
          subType: "RightClick",
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (param, { element }) => {
            assert5(element, "Element not found, cannot right click");
            await this.page.mouse.click(
              element.center[0],
              element.center[1],
              { button: "right" }
            );
          }
        };
        tasks.push(taskActionRightClick);
      } else if (plan2.type === "Drag") {
        const taskActionDrag = {
          type: "Action",
          subType: "Drag",
          param: plan2.param,
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (taskParam) => {
            assert5(
              taskParam?.start_box && taskParam?.end_box,
              "No start_box or end_box to drag"
            );
            await this.page.mouse.drag(taskParam.start_box, taskParam.end_box);
          }
        };
        tasks.push(taskActionDrag);
      } else if (plan2.type === "Hover") {
        const taskActionHover = {
          type: "Action",
          subType: "Hover",
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (param, { element }) => {
            assert5(element, "Element not found, cannot hover");
            await this.page.mouse.move(element.center[0], element.center[1]);
          }
        };
        tasks.push(taskActionHover);
      } else if (plan2.type === "Scroll") {
        const taskActionScroll = {
          type: "Action",
          subType: "Scroll",
          param: plan2.param,
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (taskParam, { element }) => {
            const startingPoint = element ? {
              left: element.center[0],
              top: element.center[1]
            } : void 0;
            const scrollToEventName = taskParam?.scrollType;
            if (scrollToEventName === "untilTop") {
              await this.page.scrollUntilTop(startingPoint);
            } else if (scrollToEventName === "untilBottom") {
              await this.page.scrollUntilBottom(startingPoint);
            } else if (scrollToEventName === "untilRight") {
              await this.page.scrollUntilRight(startingPoint);
            } else if (scrollToEventName === "untilLeft") {
              await this.page.scrollUntilLeft(startingPoint);
            } else if (scrollToEventName === "once" || !scrollToEventName) {
              if (taskParam?.direction === "down" || !taskParam || !taskParam.direction) {
                await this.page.scrollDown(
                  taskParam?.distance || void 0,
                  startingPoint
                );
              } else if (taskParam.direction === "up") {
                await this.page.scrollUp(
                  taskParam.distance || void 0,
                  startingPoint
                );
              } else if (taskParam.direction === "left") {
                await this.page.scrollLeft(
                  taskParam.distance || void 0,
                  startingPoint
                );
              } else if (taskParam.direction === "right") {
                await this.page.scrollRight(
                  taskParam.distance || void 0,
                  startingPoint
                );
              } else {
                throw new Error(
                  `Unknown scroll direction: ${taskParam.direction}`
                );
              }
              await sleep(500);
            } else {
              throw new Error(
                `Unknown scroll event type: ${scrollToEventName}, taskParam: ${JSON.stringify(
                  taskParam
                )}`
              );
            }
          }
        };
        tasks.push(taskActionScroll);
      } else if (plan2.type === "Sleep") {
        const taskActionSleep = {
          type: "Action",
          subType: "Sleep",
          param: plan2.param,
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (taskParam) => {
            await sleep(taskParam?.timeMs || 3e3);
          }
        };
        tasks.push(taskActionSleep);
      } else if (plan2.type === "Error") {
        const taskActionError = {
          type: "Action",
          subType: "Error",
          param: plan2.param,
          thought: plan2.thought || plan2.param?.thought,
          locate: plan2.locate,
          executor: async () => {
            throw new Error(
              plan2?.thought || plan2.param?.thought || "error without thought"
            );
          }
        };
        tasks.push(taskActionError);
      } else if (plan2.type === "ExpectedFalsyCondition") {
        const taskActionFalsyConditionStatement = {
          type: "Action",
          subType: "ExpectedFalsyCondition",
          param: null,
          thought: plan2.param?.reason,
          locate: plan2.locate,
          executor: async () => {
          }
        };
        tasks.push(taskActionFalsyConditionStatement);
      } else if (plan2.type === "Finished") {
        const taskActionFinished = {
          type: "Action",
          subType: "Finished",
          param: null,
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (param) => {
          }
        };
        tasks.push(taskActionFinished);
      } else if (plan2.type === "AndroidHomeButton") {
        const taskActionAndroidHomeButton = {
          type: "Action",
          subType: "AndroidHomeButton",
          param: null,
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (param) => {
            assert5(
              isAndroidPage(this.page),
              "Cannot use home button on non-Android devices"
            );
            await this.page.home();
          }
        };
        tasks.push(taskActionAndroidHomeButton);
      } else if (plan2.type === "AndroidBackButton") {
        const taskActionAndroidBackButton = {
          type: "Action",
          subType: "AndroidBackButton",
          param: null,
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (param) => {
            assert5(
              isAndroidPage(this.page),
              "Cannot use back button on non-Android devices"
            );
            await this.page.back();
          }
        };
        tasks.push(taskActionAndroidBackButton);
      } else if (plan2.type === "AndroidRecentAppsButton") {
        const taskActionAndroidRecentAppsButton = {
          type: "Action",
          subType: "AndroidRecentAppsButton",
          param: null,
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (param) => {
            assert5(
              isAndroidPage(this.page),
              "Cannot use recent apps button on non-Android devices"
            );
            await this.page.recentApps();
          }
        };
        tasks.push(taskActionAndroidRecentAppsButton);
      } else if (plan2.type === "AndroidLongPress") {
        const taskActionAndroidLongPress = {
          type: "Action",
          subType: "AndroidLongPress",
          param: plan2.param,
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (param) => {
            assert5(
              isAndroidPage(this.page),
              "Cannot use long press on non-Android devices"
            );
            const { x, y, duration } = param;
            await this.page.longPress(x, y, duration);
          }
        };
        tasks.push(taskActionAndroidLongPress);
      } else if (plan2.type === "AndroidPull") {
        const taskActionAndroidPull = {
          type: "Action",
          subType: "AndroidPull",
          param: plan2.param,
          thought: plan2.thought,
          locate: plan2.locate,
          executor: async (param) => {
            assert5(
              isAndroidPage(this.page),
              "Cannot use pull action on non-Android devices"
            );
            const { direction, startPoint, distance, duration } = param;
            const convertedStartPoint = startPoint ? { left: startPoint.x, top: startPoint.y } : void 0;
            if (direction === "down") {
              await this.page.pullDown(
                convertedStartPoint,
                distance,
                duration
              );
            } else if (direction === "up") {
              await this.page.pullUp(convertedStartPoint, distance, duration);
            } else {
              throw new Error(`Unknown pull direction: ${direction}`);
            }
          }
        };
        tasks.push(taskActionAndroidPull);
      } else {
        throw new Error(`Unknown or unsupported task type: ${plan2.type}`);
      }
    });
    const wrappedTasks = tasks.map(
      (task, index) => {
        if (task.type === "Action") {
          return this.prependExecutorWithScreenshot(
            task,
            index === tasks.length - 1
          );
        }
        return task;
      }
    );
    return {
      tasks: wrappedTasks
    };
  }
  async setupPlanningContext(executorContext) {
    const shotTime = Date.now();
    const pageContext = await this.insight.contextRetrieverFn("locate");
    const recordItem = {
      type: "screenshot",
      ts: shotTime,
      screenshot: pageContext.screenshotBase64,
      timing: "before Planning"
    };
    executorContext.task.recorder = [recordItem];
    executorContext.task.pageContext = pageContext;
    return {
      pageContext
    };
  }
  async loadYamlFlowAsPlanning(userInstruction, yamlString) {
    const taskExecutor = new Executor(taskTitleStr("Action", userInstruction), {
      onTaskStart: this.onTaskStartCallback
    });
    const task = {
      type: "Planning",
      subType: "LoadYaml",
      locate: null,
      param: {
        userInstruction
      },
      executor: async (param, executorContext) => {
        await this.setupPlanningContext(executorContext);
        return {
          output: {
            actions: [],
            more_actions_needed_by_instruction: false,
            log: "",
            yamlString
          },
          cache: {
            hit: true
          }
        };
      }
    };
    await taskExecutor.append(task);
    await taskExecutor.flush();
    return {
      executor: taskExecutor
    };
  }
  planningTaskFromPrompt(userInstruction, log, actionContext) {
    const task = {
      type: "Planning",
      subType: "Plan",
      locate: null,
      param: {
        userInstruction,
        log
      },
      executor: async (param, executorContext) => {
        const startTime = Date.now();
        const { pageContext } = await this.setupPlanningContext(executorContext);
        const planResult = await plan(param.userInstruction, {
          context: pageContext,
          log: param.log,
          actionContext,
          pageType: this.page.pageType
        });
        const {
          actions,
          log: log2,
          more_actions_needed_by_instruction,
          error,
          usage,
          rawResponse,
          sleep: sleep3
        } = planResult;
        executorContext.task.log = {
          ...executorContext.task.log || {},
          rawResponse
        };
        executorContext.task.usage = usage;
        let stopCollecting = false;
        let bboxCollected = false;
        let planParsingError = "";
        const finalActions = (actions || []).reduce(
          (acc, planningAction) => {
            if (stopCollecting) {
              return acc;
            }
            if (planningAction.locate) {
              if (bboxCollected && planningAction.locate.bbox) {
                delete planningAction.locate.bbox;
              }
              if (planningAction.locate.bbox) {
                bboxCollected = true;
              }
              acc.push({
                type: "Locate",
                locate: planningAction.locate,
                param: null,
                // thought is prompt created by ai, always a string
                thought: planningAction.locate.prompt
              });
            } else if (["Tap", "Hover", "Input"].includes(planningAction.type)) {
              planParsingError = `invalid planning response: ${JSON.stringify(planningAction)}`;
              stopCollecting = true;
              return acc;
            }
            acc.push(planningAction);
            return acc;
          },
          []
        );
        if (sleep3) {
          const timeNow = Date.now();
          const timeRemaining = sleep3 - (timeNow - startTime);
          if (timeRemaining > 0) {
            finalActions.push({
              type: "Sleep",
              param: {
                timeMs: timeRemaining
              },
              locate: null
            });
          }
        }
        if (finalActions.length === 0) {
          assert5(
            !more_actions_needed_by_instruction || sleep3,
            error ? `Failed to plan: ${error}` : planParsingError || "No plan found"
          );
        }
        return {
          output: {
            actions: finalActions,
            more_actions_needed_by_instruction,
            log: log2,
            yamlFlow: planResult.yamlFlow
          },
          cache: {
            hit: false
          },
          pageContext
        };
      }
    };
    return task;
  }
  planningTaskToGoal(userInstruction) {
    const task = {
      type: "Planning",
      subType: "Plan",
      locate: null,
      param: {
        userInstruction
      },
      executor: async (param, executorContext) => {
        const { pageContext } = await this.setupPlanningContext(executorContext);
        const imagePayload = await resizeImageForUiTars(
          pageContext.screenshotBase64,
          pageContext.size
        );
        this.appendConversationHistory({
          role: "user",
          content: [
            {
              type: "image_url",
              image_url: {
                url: imagePayload
              }
            }
          ]
        });
        const planResult = await vlmPlanning({
          userInstruction: param.userInstruction,
          conversationHistory: this.conversationHistory,
          size: pageContext.size
        });
        const { actions, action_summary, usage } = planResult;
        executorContext.task.log = {
          ...executorContext.task.log || {},
          rawResponse: planResult.rawResponse
        };
        executorContext.task.usage = usage;
        this.appendConversationHistory({
          role: "assistant",
          content: action_summary
        });
        return {
          output: {
            actions,
            thought: actions[0]?.thought,
            actionType: actions[0].type,
            more_actions_needed_by_instruction: true,
            log: "",
            yamlFlow: planResult.yamlFlow
          },
          cache: {
            hit: false
          }
        };
      }
    };
    return task;
  }
  async runPlans(title, plans, opts) {
    const taskExecutor = new Executor(title, {
      onTaskStart: this.onTaskStartCallback
    });
    const { tasks } = await this.convertPlanToExecutable(plans, opts);
    await taskExecutor.append(tasks);
    const result = await taskExecutor.flush();
    const { output } = result;
    return {
      output,
      executor: taskExecutor
    };
  }
  async action(userPrompt, actionContext, opts) {
    const taskExecutor = new Executor(taskTitleStr("Action", userPrompt), {
      onTaskStart: this.onTaskStartCallback
    });
    let planningTask = this.planningTaskFromPrompt(userPrompt, void 0, actionContext);
    let replanCount = 0;
    const logList = [];
    const yamlFlow = [];
    const replanningCycleLimit = getAIConfigInNumber2(MIDSCENE_REPLANNING_CYCLE_LIMIT) || defaultReplanningCycleLimit;
    while (planningTask) {
      if (replanCount > replanningCycleLimit) {
        const errorMsg = "Replanning too many times, please split the task into multiple steps";
        return this.appendErrorPlan(taskExecutor, errorMsg);
      }
      await taskExecutor.append(planningTask);
      const result = await taskExecutor.flush();
      const planResult = result?.output;
      if (taskExecutor.isInErrorState()) {
        return {
          output: planResult,
          executor: taskExecutor
        };
      }
      const plans = planResult.actions || [];
      yamlFlow.push(...planResult.yamlFlow || []);
      let executables;
      try {
        executables = await this.convertPlanToExecutable(plans, opts);
        taskExecutor.append(executables.tasks);
      } catch (error) {
        return this.appendErrorPlan(
          taskExecutor,
          `Error converting plans to executable tasks: ${error}, plans: ${JSON.stringify(
            plans
          )}`
        );
      }
      await taskExecutor.flush();
      if (taskExecutor.isInErrorState()) {
        return {
          output: void 0,
          executor: taskExecutor
        };
      }
      if (planResult?.log) {
        logList.push(planResult.log);
      }
      if (!planResult.more_actions_needed_by_instruction) {
        planningTask = null;
        break;
      }
      planningTask = this.planningTaskFromPrompt(
        userPrompt,
        logList.length > 0 ? `- ${logList.join("\n- ")}` : void 0,
        actionContext
      );
      replanCount++;
    }
    return {
      output: {
        yamlFlow
      },
      executor: taskExecutor
    };
  }
  async actionToGoal(userPrompt, opts) {
    const taskExecutor = new Executor(taskTitleStr("Action", userPrompt), {
      onTaskStart: this.onTaskStartCallback
    });
    this.conversationHistory = [];
    const isCompleted = false;
    let currentActionNumber = 0;
    const maxActionNumber = 40;
    const yamlFlow = [];
    while (!isCompleted && currentActionNumber < maxActionNumber) {
      currentActionNumber++;
      const planningTask = this.planningTaskToGoal(userPrompt);
      await taskExecutor.append(planningTask);
      const result = await taskExecutor.flush();
      if (taskExecutor.isInErrorState()) {
        return {
          output: void 0,
          executor: taskExecutor
        };
      }
      if (!result) {
        throw new Error(
          "result of taskExecutor.flush() is undefined in function actionToGoal"
        );
      }
      const { output } = result;
      const plans = output.actions;
      yamlFlow.push(...output.yamlFlow || []);
      let executables;
      try {
        executables = await this.convertPlanToExecutable(plans, opts);
        taskExecutor.append(executables.tasks);
      } catch (error) {
        return this.appendErrorPlan(
          taskExecutor,
          `Error converting plans to executable tasks: ${error}, plans: ${JSON.stringify(
            plans
          )}`
        );
      }
      await taskExecutor.flush();
      if (taskExecutor.isInErrorState()) {
        return {
          output: void 0,
          executor: taskExecutor
        };
      }
      if (plans[0].type === "Finished") {
        break;
      }
    }
    return {
      output: {
        yamlFlow
      },
      executor: taskExecutor
    };
  }
  async createTypeQueryTask(type, demand, opt, multimodalPrompt) {
    const taskExecutor = new Executor(
      taskTitleStr(
        type,
        typeof demand === "string" ? demand : JSON.stringify(demand)
      ),
      {
        onTaskStart: this.onTaskStartCallback
      }
    );
    const queryTask = {
      type: "Insight",
      subType: type,
      locate: null,
      param: {
        // TODO: display image thumbnail in report
        dataDemand: multimodalPrompt ? {
          demand,
          multimodalPrompt
        } : demand
        // for user param presentation in report right sidebar
      },
      executor: async (param) => {
        let insightDump;
        const dumpCollector = (dump) => {
          insightDump = dump;
        };
        this.insight.onceDumpUpdatedFn = dumpCollector;
        const ifTypeRestricted = type !== "Query";
        let demandInput = demand;
        if (ifTypeRestricted) {
          demandInput = {
            result: `${type}, ${demand}`
          };
        }
        const { data, usage, thought: thought2 } = await this.insight.extract(
          demandInput,
          opt,
          multimodalPrompt
        );
        let outputResult = data;
        if (ifTypeRestricted) {
          assert5(data?.result !== void 0, "No result in query data");
          outputResult = data.result;
        }
        return {
          output: outputResult,
          log: { dump: insightDump },
          usage,
          thought: thought2
        };
      }
    };
    await taskExecutor.append(this.prependExecutorWithScreenshot(queryTask));
    const result = await taskExecutor.flush();
    if (!result) {
      throw new Error(
        "result of taskExecutor.flush() is undefined in function createTypeQueryTask"
      );
    }
    const { output, thought } = result;
    return {
      output,
      thought,
      executor: taskExecutor
    };
  }
  async query(demand, opt) {
    return this.createTypeQueryTask("Query", demand, opt);
  }
  async boolean(prompt, opt) {
    const { textPrompt, multimodalPrompt } = parsePrompt(prompt);
    return this.createTypeQueryTask(
      "Boolean",
      textPrompt,
      opt,
      multimodalPrompt
    );
  }
  async number(prompt, opt) {
    const { textPrompt, multimodalPrompt } = parsePrompt(prompt);
    return this.createTypeQueryTask(
      "Number",
      textPrompt,
      opt,
      multimodalPrompt
    );
  }
  async string(prompt, opt) {
    const { textPrompt, multimodalPrompt } = parsePrompt(prompt);
    return this.createTypeQueryTask(
      "String",
      textPrompt,
      opt,
      multimodalPrompt
    );
  }
  async assert(assertion, opt) {
    const { textPrompt, multimodalPrompt } = parsePrompt(assertion);
    return await this.createTypeQueryTask(
      "Assert",
      textPrompt,
      opt,
      multimodalPrompt
    );
  }
  /**
   * Append a message to the conversation history
   * For user messages with images:
   * - Keep max 4 user image messages in history
   * - Remove oldest user image message when limit reached
   * For assistant messages:
   * - Simply append to history
   * @param conversationHistory Message to append
   */
  appendConversationHistory(conversationHistory) {
    if (conversationHistory.role === "user") {
      const userImgItems = this.conversationHistory.filter(
        (item) => item.role === "user"
      );
      if (userImgItems.length >= 4 && conversationHistory.role === "user") {
        const firstUserImgIndex = this.conversationHistory.findIndex(
          (item) => item.role === "user"
        );
        if (firstUserImgIndex >= 0) {
          this.conversationHistory.splice(firstUserImgIndex, 1);
        }
      }
    }
    this.conversationHistory.push(conversationHistory);
  }
  async appendErrorPlan(taskExecutor, errorMsg) {
    const errorPlan = {
      type: "Error",
      param: {
        thought: errorMsg
      },
      locate: null
    };
    const { tasks } = await this.convertPlanToExecutable([errorPlan]);
    await taskExecutor.append(this.prependExecutorWithScreenshot(tasks[0]));
    await taskExecutor.flush();
    return {
      output: void 0,
      executor: taskExecutor
    };
  }
  async waitFor(assertion, opt) {
    const description = `waitFor: ${assertion}`;
    const taskExecutor = new Executor(taskTitleStr("WaitFor", description), {
      onTaskStart: this.onTaskStartCallback
    });
    const { timeoutMs, checkIntervalMs } = opt;
    assert5(assertion, "No assertion for waitFor");
    assert5(timeoutMs, "No timeoutMs for waitFor");
    assert5(checkIntervalMs, "No checkIntervalMs for waitFor");
    const overallStartTime = Date.now();
    let startTime = Date.now();
    let errorThought = "";
    while (Date.now() - overallStartTime < timeoutMs) {
      startTime = Date.now();
      const assertPlan = {
        type: "AssertWithoutThrow",
        param: {
          assertion
        },
        locate: null
      };
      const { tasks: assertTasks } = await this.convertPlanToExecutable([
        assertPlan
      ]);
      await taskExecutor.append(
        this.prependExecutorWithScreenshot(assertTasks[0])
      );
      const result = await taskExecutor.flush();
      if (!result) {
        throw new Error(
          "result of taskExecutor.flush() is undefined in function waitFor"
        );
      }
      const { output } = result;
      if (output?.pass) {
        return {
          output: void 0,
          executor: taskExecutor
        };
      }
      errorThought = output?.thought || `unknown error when waiting for assertion: ${assertion}`;
      const now = Date.now();
      if (now - startTime < checkIntervalMs) {
        const timeRemaining = checkIntervalMs - (now - startTime);
        const sleepPlan = {
          type: "Sleep",
          param: {
            timeMs: timeRemaining
          },
          locate: null
        };
        const { tasks: sleepTasks } = await this.convertPlanToExecutable([
          sleepPlan
        ]);
        await taskExecutor.append(
          this.prependExecutorWithScreenshot(sleepTasks[0])
        );
        await taskExecutor.flush();
      }
    }
    return this.appendErrorPlan(
      taskExecutor,
      `waitFor timeout: ${errorThought}`
    );
  }
};

// src/common/plan-builder.ts
import { getDebug as getDebug4 } from "@midscene/shared/logger";
import { assert as assert6 } from "@midscene/shared/utils";
var debug4 = getDebug4("plan-builder");
function buildPlans(type, locateParam, param) {
  let returnPlans = [];
  const locatePlan = locateParam ? {
    type: "Locate",
    locate: locateParam,
    param: locateParam,
    thought: ""
  } : null;
  if (type === "Tap" || type === "Hover" || type === "RightClick") {
    assert6(locateParam, `missing locate info for action "${type}"`);
    assert6(locatePlan, `missing locate info for action "${type}"`);
    const tapPlan = {
      type,
      param: null,
      thought: "",
      locate: locateParam
    };
    returnPlans = [locatePlan, tapPlan];
  }
  if (type === "Input" || type === "KeyboardPress") {
    if (type === "Input") {
      assert6(locateParam, `missing locate info for action "${type}"`);
    }
    assert6(param, `missing param for action "${type}"`);
    const inputPlan = {
      type,
      param,
      thought: "",
      locate: locateParam
    };
    if (locatePlan) {
      returnPlans = [locatePlan, inputPlan];
    } else {
      returnPlans = [inputPlan];
    }
  }
  if (type === "Scroll") {
    assert6(param, `missing param for action "${type}"`);
    const scrollPlan = {
      type,
      param,
      thought: "",
      locate: locateParam
    };
    if (locatePlan) {
      returnPlans = [locatePlan, scrollPlan];
    } else {
      returnPlans = [scrollPlan];
    }
  }
  if (type === "Sleep") {
    assert6(param, `missing param for action "${type}"`);
    const sleepPlan = {
      type,
      param,
      thought: "",
      locate: null
    };
    returnPlans = [sleepPlan];
  }
  if (type === "Locate") {
    assert6(locateParam, `missing locate info for action "${type}"`);
    const locatePlan2 = {
      type,
      param: locateParam,
      locate: locateParam,
      thought: ""
    };
    returnPlans = [locatePlan2];
  }
  if (returnPlans) {
    debug4("buildPlans", returnPlans);
    return returnPlans;
  }
  throw new Error(`Not supported type: ${type}`);
}

// src/common/agent.ts
var debug5 = getDebug5("web-integration");
var distanceOfTwoPoints = (p1, p2) => {
  const [x1, y1] = p1;
  const [x2, y2] = p2;
  return Math.round(Math.sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2));
};
var includedInRect = (point, rect) => {
  const [x, y] = point;
  const { left, top, width, height } = rect;
  return x >= left && x <= left + width && y >= top && y <= top + height;
};
var defaultInsightExtractOption = {
  domIncluded: false,
  screenshotIncluded: true
};
var PageAgent = class {
  constructor(page, opts) {
    /**
     * If true, the agent will not perform any actions
     */
    this.dryMode = false;
    this.destroyed = false;
    this.page = page;
    this.opts = Object.assign(
      {
        generateReport: true,
        autoPrintReportMsg: true,
        groupName: "Midscene Report",
        groupDescription: ""
      },
      opts || {}
    );
    if (this.page.pageType === "puppeteer" || this.page.pageType === "playwright") {
      this.page.waitForNavigationTimeout = this.opts.waitForNavigationTimeout ?? DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT;
      this.page.waitForNetworkIdleTimeout = this.opts.waitForNetworkIdleTimeout ?? DEFAULT_WAIT_FOR_NETWORK_IDLE_TIMEOUT;
    }
    this.onTaskStartTip = this.opts.onTaskStartTip;
    this.insight = new Insight(
      async (action) => {
        return this.getUIContext(action);
      }
    );
    if (opts?.cacheId && this.page.pageType !== "android") {
      this.taskCache = new TaskCache(
        opts.cacheId,
        getAIConfigInBoolean("MIDSCENE_CACHE")
        // if we should use cache to match the element
      );
    }
    this.taskExecutor = new PageTaskExecutor(this.page, this.insight, {
      taskCache: this.taskCache,
      onTaskStart: this.callbackOnTaskStartTip.bind(this)
    });
    this.dump = this.resetDump();
    this.reportFileName = opts?.reportFileName || getReportFileName(opts?.testId || this.page.pageType || "web");
  }
  async getUIContext(action) {
    if (action && (action === "extract" || action === "assert")) {
      return await parseContextFromWebPage(this.page, {
        ignoreMarker: true
      });
    }
    return await parseContextFromWebPage(this.page, {
      ignoreMarker: !!vlLocateMode()
    });
  }
  async setAIActionContext(prompt) {
    this.opts.aiActionContext = prompt;
  }
  resetDump() {
    this.dump = {
      groupName: this.opts.groupName,
      groupDescription: this.opts.groupDescription,
      executions: []
    };
    return this.dump;
  }
  appendExecutionDump(execution) {
    const trimmedExecution = trimContextByViewport(execution);
    const currentDump = this.dump;
    currentDump.executions.push(trimmedExecution);
  }
  dumpDataString() {
    this.dump.groupName = this.opts.groupName;
    this.dump.groupDescription = this.opts.groupDescription;
    return stringifyDumpData(this.dump);
  }
  reportHTMLString() {
    return reportHTMLContent(this.dumpDataString());
  }
  writeOutActionDumps() {
    if (this.destroyed) {
      throw new Error(
        "PageAgent has been destroyed. Cannot update report file."
      );
    }
    const { generateReport, autoPrintReportMsg } = this.opts;
    this.reportFile = writeLogFile({
      fileName: this.reportFileName,
      fileExt: groupedActionDumpFileExt,
      fileContent: this.dumpDataString(),
      type: "dump",
      generateReport
    });
    debug5("writeOutActionDumps", this.reportFile);
    if (generateReport && autoPrintReportMsg && this.reportFile) {
      printReportMsg(this.reportFile);
    }
  }
  async callbackOnTaskStartTip(task) {
    const param = paramStr(task);
    const tip = param ? `${typeStr(task)} - ${param}` : typeStr(task);
    if (this.onTaskStartTip) {
      await this.onTaskStartTip(tip);
    }
  }
  async afterTaskRunning(executor, doNotThrowError = false) {
    this.appendExecutionDump(executor.dump());
    try {
      await this.onDumpUpdate?.(this.dumpDataString());
    } catch (error) {
      console.error("Error in onDumpUpdate", error);
    }
    this.writeOutActionDumps();
    if (executor.isInErrorState() && !doNotThrowError) {
      const errorTask = executor.latestErrorTask();
      throw new Error(`${errorTask?.errorMessage}
${errorTask?.errorStack}`, {
        cause: errorTask?.error
      });
    }
  }
  buildDetailedLocateParam(locatePrompt, opt) {
    assert7(locatePrompt, "missing locate prompt");
    if (typeof opt === "object") {
      const prompt = locatePrompt;
      const deepThink = opt.deepThink ?? false;
      const cacheable = opt.cacheable ?? true;
      const xpath = opt.xpath;
      return {
        prompt,
        deepThink,
        cacheable,
        xpath
      };
    }
    return {
      prompt: locatePrompt
    };
  }
  async aiTap(locatePrompt, opt) {
    const detailedLocateParam = this.buildDetailedLocateParam(
      locatePrompt,
      opt
    );
    const plans = buildPlans("Tap", detailedLocateParam);
    const { executor, output } = await this.taskExecutor.runPlans(
      taskTitleStr("Tap", locateParamStr(detailedLocateParam)),
      plans,
      { cacheable: opt?.cacheable }
    );
    await this.afterTaskRunning(executor);
    return output;
  }
  async aiRightClick(locatePrompt, opt) {
    const detailedLocateParam = this.buildDetailedLocateParam(
      locatePrompt,
      opt
    );
    const plans = buildPlans("RightClick", detailedLocateParam);
    const { executor, output } = await this.taskExecutor.runPlans(
      taskTitleStr("RightClick", locateParamStr(detailedLocateParam)),
      plans,
      { cacheable: opt?.cacheable }
    );
    await this.afterTaskRunning(executor);
    return output;
  }
  async aiHover(locatePrompt, opt) {
    const detailedLocateParam = this.buildDetailedLocateParam(
      locatePrompt,
      opt
    );
    const plans = buildPlans("Hover", detailedLocateParam);
    const { executor, output } = await this.taskExecutor.runPlans(
      taskTitleStr("Hover", locateParamStr(detailedLocateParam)),
      plans,
      { cacheable: opt?.cacheable }
    );
    await this.afterTaskRunning(executor);
    return output;
  }
  async aiInput(value, locatePrompt, opt) {
    assert7(
      typeof value === "string",
      "input value must be a string, use empty string if you want to clear the input"
    );
    assert7(locatePrompt, "missing locate prompt for input");
    const detailedLocateParam = this.buildDetailedLocateParam(
      locatePrompt,
      opt
    );
    const plans = buildPlans("Input", detailedLocateParam, {
      value,
      autoDismissKeyboard: opt?.autoDismissKeyboard
    });
    const { executor, output } = await this.taskExecutor.runPlans(
      taskTitleStr("Input", locateParamStr(detailedLocateParam)),
      plans,
      {
        cacheable: opt?.cacheable
      }
    );
    await this.afterTaskRunning(executor);
    return output;
  }
  async aiKeyboardPress(keyName, locatePrompt, opt) {
    assert7(keyName, "missing keyName for keyboard press");
    const detailedLocateParam = locatePrompt ? this.buildDetailedLocateParam(locatePrompt, opt) : void 0;
    const plans = buildPlans("KeyboardPress", detailedLocateParam, {
      value: keyName
    });
    const { executor, output } = await this.taskExecutor.runPlans(
      taskTitleStr("KeyboardPress", locateParamStr(detailedLocateParam)),
      plans,
      { cacheable: opt?.cacheable }
    );
    await this.afterTaskRunning(executor);
    return output;
  }
  async aiScroll(scrollParam, locatePrompt, opt) {
    const detailedLocateParam = locatePrompt ? this.buildDetailedLocateParam(locatePrompt, opt) : void 0;
    const plans = buildPlans("Scroll", detailedLocateParam, scrollParam);
    const paramInTitle = locatePrompt ? `${locateParamStr(detailedLocateParam)} - ${scrollParamStr(scrollParam)}` : scrollParamStr(scrollParam);
    const { executor, output } = await this.taskExecutor.runPlans(
      taskTitleStr("Scroll", paramInTitle),
      plans,
      { cacheable: opt?.cacheable }
    );
    await this.afterTaskRunning(executor);
    return output;
  }
  async aiAction(taskPrompt, opt) {
    const cacheable = opt?.cacheable;
    const isVlmUiTars = vlLocateMode() === "vlm-ui-tars";
    const matchedCache = isVlmUiTars || cacheable === false ? void 0 : this.taskCache?.matchPlanCache(taskPrompt);
    if (matchedCache && this.taskCache?.isCacheResultUsed) {
      const { executor: executor2 } = await this.taskExecutor.loadYamlFlowAsPlanning(
        taskPrompt,
        matchedCache.cacheContent?.yamlWorkflow
      );
      await await this.afterTaskRunning(executor2);
      debug5("matched cache, will call .runYaml to run the action");
      const yaml5 = matchedCache.cacheContent?.yamlWorkflow;
      return this.runYaml(yaml5);
    }
    const { output, executor } = await (isVlmUiTars ? this.taskExecutor.actionToGoal(taskPrompt, { cacheable }) : this.taskExecutor.action(taskPrompt, this.opts.aiActionContext, {
      cacheable
    }));
    if (this.taskCache && output?.yamlFlow && cacheable !== false) {
      const yamlContent = {
        tasks: [
          {
            name: taskPrompt,
            flow: output.yamlFlow
          }
        ]
      };
      const yamlFlowStr = yaml4.dump(yamlContent);
      this.taskCache.updateOrAppendCacheRecord(
        {
          type: "plan",
          prompt: taskPrompt,
          yamlWorkflow: yamlFlowStr
        },
        matchedCache
      );
    }
    await this.afterTaskRunning(executor);
    return output;
  }
  async aiQuery(demand, opt = defaultInsightExtractOption) {
    const { output, executor } = await this.taskExecutor.query(demand, opt);
    await this.afterTaskRunning(executor);
    return output;
  }
  async aiBoolean(prompt, opt = defaultInsightExtractOption) {
    const { output, executor } = await this.taskExecutor.boolean(prompt, opt);
    await this.afterTaskRunning(executor);
    return output;
  }
  async aiNumber(prompt, opt = defaultInsightExtractOption) {
    const { output, executor } = await this.taskExecutor.number(prompt, opt);
    await this.afterTaskRunning(executor);
    return output;
  }
  async aiString(prompt, opt = defaultInsightExtractOption) {
    const { output, executor } = await this.taskExecutor.string(prompt, opt);
    await this.afterTaskRunning(executor);
    return output;
  }
  async aiAsk(prompt, opt = defaultInsightExtractOption) {
    return this.aiString(prompt, opt);
  }
  async describeElementAtPoint(center, opt) {
    const { verifyPrompt = true, retryLimit = 3 } = opt || {};
    let success = false;
    let retryCount = 0;
    let resultPrompt = "";
    let deepThink = opt?.deepThink || false;
    let verifyResult;
    while (!success && retryCount < retryLimit) {
      if (retryCount >= 2) {
        deepThink = true;
      }
      debug5(
        "aiDescribe",
        center,
        "verifyPrompt",
        verifyPrompt,
        "retryCount",
        retryCount,
        "deepThink",
        deepThink
      );
      const text = await this.insight.describe(center, { deepThink });
      debug5("aiDescribe text", text);
      assert7(text.description, `failed to describe element at [${center}]`);
      resultPrompt = text.description;
      verifyResult = await this.verifyLocator(
        resultPrompt,
        deepThink ? { deepThink: true } : void 0,
        center,
        opt
      );
      if (verifyResult.pass) {
        success = true;
      } else {
        retryCount++;
      }
    }
    return {
      prompt: resultPrompt,
      deepThink,
      verifyResult
    };
  }
  async verifyLocator(prompt, locateOpt, expectCenter, verifyLocateOption) {
    debug5("verifyLocator", prompt, locateOpt, expectCenter, verifyLocateOption);
    const { center: verifyCenter, rect: verifyRect } = await this.aiLocate(
      prompt,
      locateOpt
    );
    const distance = distanceOfTwoPoints(expectCenter, verifyCenter);
    const included = includedInRect(expectCenter, verifyRect);
    const pass = distance <= (verifyLocateOption?.centerDistanceThreshold || 20) || included;
    const verifyResult = {
      pass,
      rect: verifyRect,
      center: verifyCenter,
      centerDistance: distance
    };
    debug5("aiDescribe verifyResult", verifyResult);
    return verifyResult;
  }
  async aiLocate(prompt, opt) {
    const detailedLocateParam = this.buildDetailedLocateParam(prompt, opt);
    const plans = buildPlans("Locate", detailedLocateParam);
    const { executor, output } = await this.taskExecutor.runPlans(
      taskTitleStr("Locate", locateParamStr(detailedLocateParam)),
      plans,
      { cacheable: opt?.cacheable }
    );
    await this.afterTaskRunning(executor);
    const { element } = output;
    return {
      rect: element?.rect,
      center: element?.center,
      scale: (await this.page.size()).dpr
    };
  }
  async aiAssert(assertion, msg, opt) {
    const { output, executor, thought } = await this.taskExecutor.assert(
      assertion,
      {
        returnThought: true
      }
    );
    await this.afterTaskRunning(executor, true);
    if (opt?.keepRawResponse) {
      return {
        pass: output,
        thought
      };
    }
    if (!output) {
      const errMsg = msg || `Assertion failed: ${assertion}`;
      const reasonMsg = `Reason: ${thought || executor.latestErrorTask()?.error || "(no_reason)"}`;
      throw new Error(`${errMsg}
${reasonMsg}`);
    }
  }
  async aiWaitFor(assertion, opt) {
    const { executor } = await this.taskExecutor.waitFor(assertion, {
      timeoutMs: opt?.timeoutMs || 15 * 1e3,
      checkIntervalMs: opt?.checkIntervalMs || 3 * 1e3,
      assertion
    });
    await this.afterTaskRunning(executor, true);
    if (executor.isInErrorState()) {
      const errorTask = executor.latestErrorTask();
      throw new Error(`${errorTask?.error}
${errorTask?.errorStack}`);
    }
  }
  async ai(taskPrompt, type = "action") {
    if (type === "action") {
      return this.aiAction(taskPrompt);
    }
    if (type === "query") {
      return this.aiQuery(taskPrompt);
    }
    if (type === "assert") {
      return this.aiAssert(taskPrompt);
    }
    if (type === "tap") {
      return this.aiTap(taskPrompt);
    }
    if (type === "rightClick") {
      return this.aiRightClick(taskPrompt);
    }
    throw new Error(
      `Unknown type: ${type}, only support 'action', 'query', 'assert', 'tap', 'rightClick'`
    );
  }
  async runYaml(yamlScriptContent) {
    const script = parseYamlScript(yamlScriptContent, "yaml", true);
    const player = new ScriptPlayer(script, async (target) => {
      return { agent: this, freeFn: [] };
    });
    await player.run();
    if (player.status === "error") {
      const errors = player.taskStatusList.filter((task) => task.status === "error").map((task) => {
        return `task - ${task.name}: ${task.error?.message}`;
      }).join("\n");
      throw new Error(`Error(s) occurred in running yaml script:
${errors}`);
    }
    return {
      result: player.result
    };
  }
  async evaluateJavaScript(script) {
    assert7(
      this.page.evaluateJavaScript,
      "evaluateJavaScript is not supported in current agent"
    );
    return this.page.evaluateJavaScript(script);
  }
  async destroy() {
    await this.page.destroy();
    this.resetDump();
    this.destroyed = true;
  }
  async logScreenshot(title, opt) {
    const base64 = await this.page.screenshotBase64();
    const now = Date.now();
    const recorder = [
      {
        type: "screenshot",
        ts: now,
        screenshot: base64
      }
    ];
    const task = {
      type: "Log",
      subType: "Screenshot",
      status: "finished",
      recorder,
      timing: {
        start: now,
        end: now,
        cost: 0
      },
      param: {
        content: opt?.content || ""
      },
      executor: async () => {
      }
    };
    const executionDump = {
      sdkVersion: "",
      logTime: now,
      model_name: "",
      model_description: "",
      name: `Log - ${title || "untitled"}`,
      description: opt?.content || "",
      tasks: [task]
    };
    this.appendExecutionDump(executionDump);
    try {
      this.onDumpUpdate?.(this.dumpDataString());
    } catch (error) {
      console.error("Failed to update dump", error);
    }
    this.writeOutActionDumps();
  }
  _unstableLogContent() {
    const { groupName, groupDescription, executions } = this.dump;
    const newExecutions = Array.isArray(executions) ? executions.map((execution) => {
      const { tasks, ...restExecution } = execution;
      let newTasks = tasks;
      if (Array.isArray(tasks)) {
        newTasks = tasks.map((task) => {
          const { pageContext, log, ...restTask } = task;
          return restTask;
        });
      }
      return { ...restExecution, ...newTasks ? { tasks: newTasks } : {} };
    }) : [];
    return {
      groupName,
      groupDescription,
      executions: newExecutions
    };
  }
};

// src/playwright/page.ts
import {
  DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT as DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT3,
  DEFAULT_WAIT_FOR_NETWORK_IDLE_TIMEOUT as DEFAULT_WAIT_FOR_NETWORK_IDLE_TIMEOUT2
} from "@midscene/shared/constants";

// src/puppeteer/base-page.ts
import { sleep as sleep2 } from "@midscene/core/utils";
import { DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT as DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT2 } from "@midscene/shared/constants";
import { treeToList } from "@midscene/shared/extractor";
import {
  getElementInfosScriptContent,
  getExtraReturnLogic
} from "@midscene/shared/fs";
import { getDebug as getDebug6 } from "@midscene/shared/logger";
import { assert as assert8 } from "@midscene/shared/utils";
var debugPage = getDebug6("web:page");
var Page = class {
  constructor(underlyingPage, pageType, opts) {
    this.everMoved = false;
    this.underlyingPage = underlyingPage;
    this.pageType = pageType;
    this.waitForNavigationTimeout = opts?.waitForNavigationTimeout ?? DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT2;
  }
  async evaluate(pageFunction, arg) {
    let result;
    debugPage("evaluate function begin");
    if (this.pageType === "puppeteer") {
      result = await this.underlyingPage.evaluate(
        pageFunction,
        arg
      );
    } else {
      result = await this.underlyingPage.evaluate(
        pageFunction,
        arg
      );
    }
    debugPage("evaluate function end");
    return result;
  }
  async evaluateJavaScript(script) {
    return this.evaluate(script);
  }
  async waitForNavigation() {
    if (this.waitForNavigationTimeout === 0) {
      debugPage("waitForNavigation timeout is 0, skip waiting");
      return;
    }
    if (this.pageType === "puppeteer" || this.pageType === "playwright") {
      debugPage("waitForNavigation begin");
      debugPage(`waitForNavigation timeout: ${this.waitForNavigationTimeout}`);
      try {
        await this.underlyingPage.waitForSelector("html", {
          timeout: this.waitForNavigationTimeout
        });
      } catch (error) {
        console.warn(
          "[midscene:warning] Waiting for the navigation has timed out, but Midscene will continue execution. Please check https://midscenejs.com/faq.html#customize-the-network-timeout for more information on customizing the network timeout"
        );
      }
      debugPage("waitForNavigation end");
    }
  }
  // @deprecated
  async getElementsInfo() {
    await this.waitForNavigation();
    debugPage("getElementsInfo begin");
    const tree = await this.getElementsNodeTree();
    debugPage("getElementsInfo end");
    return treeToList(tree);
  }
  async getXpathsById(id) {
    const elementInfosScriptContent = getElementInfosScriptContent();
    return this.evaluateJavaScript(
      `${elementInfosScriptContent}midscene_element_inspector.getXpathsById('${id}')`
    );
  }
  async getXpathsByPoint(point, isOrderSensitive) {
    const elementInfosScriptContent = getElementInfosScriptContent();
    return this.evaluateJavaScript(
      `${elementInfosScriptContent}midscene_element_inspector.getXpathsByPoint({left: ${point.left}, top: ${point.top}}, ${isOrderSensitive})`
    );
  }
  async getElementInfoByXpath(xpath) {
    const elementInfosScriptContent = getElementInfosScriptContent();
    return this.evaluateJavaScript(
      `${elementInfosScriptContent}midscene_element_inspector.getElementInfoByXpath('${xpath}')`
    );
  }
  async getElementsNodeTree() {
    await this.waitForNavigation();
    const scripts = await getExtraReturnLogic(true);
    assert8(scripts, "scripts should be set before writing report in browser");
    const startTime = Date.now();
    const captureElementSnapshot = await this.evaluate(scripts);
    const endTime = Date.now();
    debugPage(`getElementsNodeTree end, cost: ${endTime - startTime}ms`);
    return captureElementSnapshot;
  }
  async size() {
    if (this.viewportSize)
      return this.viewportSize;
    const sizeInfo = await this.evaluate(() => {
      return {
        width: document.documentElement.clientWidth,
        height: document.documentElement.clientHeight,
        dpr: window.devicePixelRatio
      };
    });
    this.viewportSize = sizeInfo;
    return sizeInfo;
  }
  async screenshotBase64() {
    const imgType = "jpeg";
    const quality = 90;
    await this.waitForNavigation();
    const startTime = Date.now();
    debugPage("screenshotBase64 begin");
    let base64;
    if (this.pageType === "puppeteer") {
      const result = await this.underlyingPage.screenshot({
        type: imgType,
        quality,
        encoding: "base64"
      });
      base64 = `data:image/jpeg;base64,${result}`;
    } else if (this.pageType === "playwright") {
      const buffer = await this.underlyingPage.screenshot({
        type: imgType,
        quality,
        timeout: 10 * 1e3
      });
      base64 = `data:image/jpeg;base64,${buffer.toString("base64")}`;
    } else {
      throw new Error("Unsupported page type for screenshot");
    }
    const endTime = Date.now();
    debugPage(`screenshotBase64 end, cost: ${endTime - startTime}ms`);
    return base64;
  }
  async url() {
    return this.underlyingPage.url();
  }
  get mouse() {
    return {
      click: async (x, y, options) => {
        await this.mouse.move(x, y);
        this.underlyingPage.mouse.click(x, y, {
          button: options?.button || "left",
          count: options?.count || 1
        });
      },
      wheel: async (deltaX, deltaY) => {
        if (this.pageType === "puppeteer") {
          await this.underlyingPage.mouse.wheel({
            deltaX,
            deltaY
          });
        } else if (this.pageType === "playwright") {
          await this.underlyingPage.mouse.wheel(
            deltaX,
            deltaY
          );
        }
      },
      move: async (x, y) => {
        this.everMoved = true;
        return this.underlyingPage.mouse.move(x, y);
      },
      drag: async (from, to) => {
        if (this.pageType === "puppeteer") {
          await this.underlyingPage.mouse.drag(
            {
              x: from.x,
              y: from.y
            },
            {
              x: to.x,
              y: to.y
            }
          );
        } else if (this.pageType === "playwright") {
          await this.underlyingPage.mouse.move(
            from.x,
            from.y
          );
          await this.underlyingPage.mouse.down();
          await this.underlyingPage.mouse.move(to.x, to.y);
          await this.underlyingPage.mouse.up();
        }
      }
    };
  }
  get keyboard() {
    return {
      type: async (text) => this.underlyingPage.keyboard.type(text, { delay: 80 }),
      press: async (action) => {
        const keys = Array.isArray(action) ? action : [action];
        for (const k of keys) {
          const commands = k.command ? [k.command] : [];
          await this.underlyingPage.keyboard.down(k.key, { commands });
        }
        for (const k of [...keys].reverse()) {
          await this.underlyingPage.keyboard.up(k.key);
        }
      },
      down: async (key) => {
        this.underlyingPage.keyboard.down(key);
      },
      up: async (key) => {
        this.underlyingPage.keyboard.up(key);
      }
    };
  }
  async clearInput(element) {
    if (!element) {
      console.warn("No element to clear input");
      return;
    }
    const backspace = async () => {
      await sleep2(100);
      await this.keyboard.press([{ key: "Backspace" }]);
    };
    const isMac = process.platform === "darwin";
    if (isMac) {
      if (this.pageType === "puppeteer") {
        await this.mouse.click(element.center[0], element.center[1], {
          count: 3
        });
        await backspace();
      }
      await this.mouse.click(element.center[0], element.center[1]);
      await this.underlyingPage.keyboard.down("Meta");
      await this.underlyingPage.keyboard.press("a");
      await this.underlyingPage.keyboard.up("Meta");
      await backspace();
    } else {
      await this.mouse.click(element.center[0], element.center[1]);
      await this.underlyingPage.keyboard.down("Control");
      await this.underlyingPage.keyboard.press("a");
      await this.underlyingPage.keyboard.up("Control");
      await backspace();
    }
  }
  async moveToPointBeforeScroll(point) {
    if (point) {
      await this.mouse.move(point.left, point.top);
    } else if (!this.everMoved) {
      const size = await this.size();
      const targetX = Math.floor(size.width / 2);
      const targetY = Math.floor(size.height / 2);
      await this.mouse.move(targetX, targetY);
    }
  }
  async scrollUntilTop(startingPoint) {
    await this.moveToPointBeforeScroll(startingPoint);
    return this.mouse.wheel(0, -9999999);
  }
  async scrollUntilBottom(startingPoint) {
    await this.moveToPointBeforeScroll(startingPoint);
    return this.mouse.wheel(0, 9999999);
  }
  async scrollUntilLeft(startingPoint) {
    await this.moveToPointBeforeScroll(startingPoint);
    return this.mouse.wheel(-9999999, 0);
  }
  async scrollUntilRight(startingPoint) {
    await this.moveToPointBeforeScroll(startingPoint);
    return this.mouse.wheel(9999999, 0);
  }
  async scrollUp(distance, startingPoint) {
    const innerHeight = await this.evaluate(() => window.innerHeight);
    const scrollDistance = distance || innerHeight * 0.7;
    await this.moveToPointBeforeScroll(startingPoint);
    return this.mouse.wheel(0, -scrollDistance);
  }
  async scrollDown(distance, startingPoint) {
    const innerHeight = await this.evaluate(() => window.innerHeight);
    const scrollDistance = distance || innerHeight * 0.7;
    await this.moveToPointBeforeScroll(startingPoint);
    return this.mouse.wheel(0, scrollDistance);
  }
  async scrollLeft(distance, startingPoint) {
    const innerWidth = await this.evaluate(() => window.innerWidth);
    const scrollDistance = distance || innerWidth * 0.7;
    await this.moveToPointBeforeScroll(startingPoint);
    return this.mouse.wheel(-scrollDistance, 0);
  }
  async scrollRight(distance, startingPoint) {
    const innerWidth = await this.evaluate(() => window.innerWidth);
    const scrollDistance = distance || innerWidth * 0.7;
    await this.moveToPointBeforeScroll(startingPoint);
    return this.mouse.wheel(scrollDistance, 0);
  }
  async navigate(url) {
    if (this.pageType === "puppeteer") {
      await this.underlyingPage.goto(url);
    } else if (this.pageType === "playwright") {
      await this.underlyingPage.goto(url);
    } else {
      throw new Error("Unsupported page type for navigate");
    }
  }
  async destroy() {
  }
};

// src/playwright/page.ts
var WebPage = class extends Page {
  constructor(page, opts) {
    super(page, "playwright", opts);
    const {
      waitForNavigationTimeout = DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT3,
      waitForNetworkIdleTimeout = DEFAULT_WAIT_FOR_NETWORK_IDLE_TIMEOUT2
    } = opts ?? {};
    this.waitForNavigationTimeout = waitForNavigationTimeout;
    this.waitForNetworkIdleTimeout = waitForNetworkIdleTimeout;
  }
};

// src/playwright/ai-fixture.ts
import { randomUUID } from "crypto";
import {
  DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT as DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT4,
  DEFAULT_WAIT_FOR_NETWORK_IDLE_TIMEOUT as DEFAULT_WAIT_FOR_NETWORK_IDLE_TIMEOUT3
} from "@midscene/shared/constants";
import { getDebug as getDebug7 } from "@midscene/shared/logger";
import { test } from "@playwright/test";
var debugPage2 = getDebug7("web:playwright:ai-fixture");
var groupAndCaseForTest = (testInfo) => {
  let taskFile;
  let taskTitle;
  const titlePath = [...testInfo.titlePath];
  if (titlePath.length > 1) {
    taskFile = titlePath.shift() || "unnamed";
    taskTitle = titlePath.join("__");
  } else if (titlePath.length === 1) {
    taskTitle = titlePath[0];
    taskFile = `${taskTitle}`;
  } else {
    taskTitle = "unnamed";
    taskFile = "unnamed";
  }
  const taskTitleWithRetry = `${taskTitle}${testInfo.retry ? `(retry #${testInfo.retry})` : ""}`;
  return {
    file: taskFile,
    id: replaceIllegalPathCharsAndSpace(`${taskFile}(${taskTitle})`),
    title: replaceIllegalPathCharsAndSpace(taskTitleWithRetry)
  };
};
var midsceneAgentKeyId = "_midsceneAgentId";
var midsceneDumpAnnotationId = "MIDSCENE_DUMP_ANNOTATION";
var PlaywrightAiFixture = (options) => {
  const {
    forceSameTabNavigation = true,
    waitForNetworkIdleTimeout = DEFAULT_WAIT_FOR_NETWORK_IDLE_TIMEOUT3,
    waitForNavigationTimeout = DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT4
  } = options ?? {};
  const pageAgentMap = {};
  const createOrReuseAgentForPage = (page, testInfo, opts) => {
    let idForPage = page[midsceneAgentKeyId];
    if (!idForPage) {
      idForPage = randomUUID();
      page[midsceneAgentKeyId] = idForPage;
      const { testId } = testInfo;
      const { file, id, title } = groupAndCaseForTest(testInfo);
      pageAgentMap[idForPage] = new PlaywrightAgent(page, {
        testId: `playwright-${testId}-${idForPage}`,
        forceSameTabNavigation,
        cacheId: id,
        groupName: title,
        groupDescription: file,
        generateReport: false,
        // we will generate it in the reporter
        ...opts
      });
      pageAgentMap[idForPage].onDumpUpdate = (dump) => {
        updateDumpAnnotation(testInfo, dump);
      };
      page.on("close", () => {
        debugPage2("page closed");
        pageAgentMap[idForPage].destroy();
        delete pageAgentMap[idForPage];
      });
    }
    return pageAgentMap[idForPage];
  };
  async function generateAiFunction(options2) {
    const { page, testInfo, use, aiActionType } = options2;
    const agent = createOrReuseAgentForPage(page, testInfo, {
      waitForNavigationTimeout,
      waitForNetworkIdleTimeout
    });
    await use(async (taskPrompt, ...args) => {
      return new Promise((resolve2, reject) => {
        test.step(`ai-${aiActionType} - ${JSON.stringify(taskPrompt)}`, async () => {
          try {
            debugPage2(
              `waitForNetworkIdle timeout: ${waitForNetworkIdleTimeout}`
            );
            await agent.waitForNetworkIdle(waitForNetworkIdleTimeout);
          } catch (error) {
            console.warn(
              "[midscene:warning] Waiting for network idle has timed out, but Midscene will continue execution. Please check https://midscenejs.com/faq.html#customize-the-network-timeout for more information on customizing the network timeout"
            );
          }
          try {
            const result = await agent[aiActionType](
              taskPrompt,
              ...args || []
            );
            resolve2(result);
          } catch (error) {
            reject(error);
          }
        });
      });
    });
  }
  const updateDumpAnnotation = (test2, dump) => {
    const currentAnnotation = test2.annotations.find((item) => {
      return item.type === midsceneDumpAnnotationId;
    });
    if (currentAnnotation) {
      currentAnnotation.description = dump;
    } else {
      test2.annotations.push({
        type: midsceneDumpAnnotationId,
        description: dump
      });
    }
  };
  return {
    agentForPage: async ({ page }, use, testInfo) => {
      await use(
        async (propsPage, opts) => {
          const agent = createOrReuseAgentForPage(propsPage || page, testInfo, {
            waitForNavigationTimeout,
            waitForNetworkIdleTimeout,
            ...opts
          });
          return agent;
        }
      );
    },
    ai: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "ai"
      });
    },
    aiAction: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiAction"
      });
    },
    aiTap: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiTap"
      });
    },
    aiRightClick: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiRightClick"
      });
    },
    aiHover: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiHover"
      });
    },
    aiInput: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiInput"
      });
    },
    aiKeyboardPress: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiKeyboardPress"
      });
    },
    aiScroll: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiScroll"
      });
    },
    aiQuery: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiQuery"
      });
    },
    aiAssert: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiAssert"
      });
    },
    aiWaitFor: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiWaitFor"
      });
    },
    aiLocate: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiLocate"
      });
    },
    aiNumber: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiNumber"
      });
    },
    aiString: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiString"
      });
    },
    aiBoolean: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiBoolean"
      });
    },
    aiAsk: async ({ page }, use, testInfo) => {
      await generateAiFunction({
        page,
        testInfo,
        use,
        aiActionType: "aiAsk"
      });
    }
  };
};

// src/playwright/index.ts
import { overrideAIConfig } from "@midscene/shared/env";
import { getDebug as getDebug8 } from "@midscene/shared/logger";
var debug6 = getDebug8("playwright:agent");
var PlaywrightAgent = class extends PageAgent {
  constructor(page, opts) {
    const webPage = new WebPage(page, opts);
    super(webPage, opts);
    const { forceSameTabNavigation = true } = opts ?? {};
    if (forceSameTabNavigation) {
      forceClosePopup(page, debug6);
    }
  }
  async waitForNetworkIdle(timeout = 1e3) {
    await this.page.underlyingPage.waitForLoadState("networkidle", { timeout });
  }
};
export {
  PlaywrightAgent,
  PlaywrightAiFixture,
  WebPage as PlaywrightWebPage,
  overrideAIConfig
};
