// @midscene/web/dist/es/utils.js
// src/common/utils.ts
import { elementByPositionWithElementInfo } from "@midscene/core/ai-model";
import { uploadTestInfoToServer } from "@midscene/core/utils";
import { MIDSCENE_REPORT_TAG_NAME, getAIConfig } from "@midscene/shared/env";
import {
  generateElementByPosition,
  getNodeFromCacheList,
  traverseTree
} from "@midscene/shared/extractor";
import { resizeImgBase64 } from "@midscene/shared/img";
import { getDebug as getDebug2 } from "@midscene/shared/logger";
import { assert as assert2, logMsg, uuid } from "@midscene/shared/utils";
import dayjs from "dayjs";

// src/web-element.ts
var WebElementInfo = class {
  constructor({
    content,
    rect,
    id,
    attributes,
    indexId,
    xpaths,
    isVisible
  }) {
    this.content = content;
    this.rect = rect;
    this.center = [
      Math.floor(rect.left + rect.width / 2),
      Math.floor(rect.top + rect.height / 2)
    ];
    this.id = id;
    this.attributes = attributes;
    this.indexId = indexId;
    this.xpaths = xpaths;
    this.isVisible = isVisible;
  }
};

// src/common/task-cache.ts
import assert from "assert";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "fs";
import { dirname, join } from "path";
import { isDeepStrictEqual } from "util";
import { getMidsceneRunSubDir } from "@midscene/shared/common";
import {
  MIDSCENE_CACHE_MAX_FILENAME_LENGTH,
  getAIConfigInNumber
} from "@midscene/shared/env";
import { getDebug } from "@midscene/shared/logger";
import { ifInBrowser, ifInWorker } from "@midscene/shared/utils";
import { generateHashId } from "@midscene/shared/utils";
import yaml from "js-yaml";
import semver from "semver";
var debug = getDebug("cache");

// src/common/utils.ts
var debug2 = getDebug2("tool:profile");
async function parseContextFromWebPage(page, _opt) {
  assert2(page, "page is required");
  if (page._forceUsePageContext) {
    return await page._forceUsePageContext();
  }
  debug2("Getting page URL");
  const url = await page.url();
  debug2("URL end");
  debug2("Uploading test info to server");
  uploadTestInfoToServer({ testUrl: url });
  debug2("UploadTestInfoToServer end");
  let screenshotBase64;
  let tree;
  debug2("Starting parallel operations: screenshot and element tree");
  await Promise.all([
    page.screenshotBase64().then((base64) => {
      screenshotBase64 = base64;
      debug2("ScreenshotBase64 end");
    }),
    page.getElementsNodeTree().then(async (treeRoot) => {
      tree = treeRoot;
      debug2("GetElementsNodeTree end");
    })
  ]);
  debug2("ParseContextFromWebPage end");
  debug2("Traversing element tree");
  const webTree = traverseTree(tree, (elementInfo) => {
    const { rect, id, content, attributes, indexId, isVisible } = elementInfo;
    return new WebElementInfo({
      rect,
      id,
      content,
      attributes,
      indexId,
      isVisible
    });
  });
  debug2("TraverseTree end");
  assert2(screenshotBase64, "screenshotBase64 is required");
  const size = await page.size();
  debug2(`size: ${size.width}x${size.height} dpr: ${size.dpr}`);
  if (size.dpr && size.dpr > 1) {
    debug2("Resizing screenshot for high DPR display");
    screenshotBase64 = await resizeImgBase64(screenshotBase64, {
      width: size.width,
      height: size.height
    });
    debug2("ResizeImgBase64 end");
  }
  return {
    tree: webTree,
    size,
    screenshotBase64,
    url
  };
}
function getReportFileName(tag = "web") {
  const reportTagName = getAIConfig(MIDSCENE_REPORT_TAG_NAME);
  const dateTimeInFileName = dayjs().format("YYYY-MM-DD_HH-mm-ss");
  const uniqueId = uuid().substring(0, 8);
  return `${reportTagName || tag}-${dateTimeInFileName}-${uniqueId}`;
}
function printReportMsg(filepath) {
  logMsg(`Midscene - report file updated: ${filepath}`);
}
function getCurrentExecutionFile(trace) {
  const error = new Error();
  const stackTrace = trace || error.stack;
  const pkgDir = process.cwd() || "";
  if (stackTrace) {
    const stackLines = stackTrace.split("\n");
    for (const line of stackLines) {
      if (line.includes(".spec.") || line.includes(".test.") || line.includes(".ts") || line.includes(".js")) {
        const match = line.match(/(?:at\s+)?(.*?\.(?:spec|test)\.[jt]s)/);
        if (match?.[1]) {
          const targetFileName = match[1].replace(pkgDir, "").trim().replace("at ", "");
          return targetFileName;
        }
      }
    }
  }
  return false;
}
var testFileIndex = /* @__PURE__ */ new Map();
function generateCacheId(fileName) {
  let taskFile = fileName || getCurrentExecutionFile();
  if (!taskFile) {
    taskFile = uuid();
    console.warn(
      "Midscene - using random UUID for cache id. Cache may be invalid."
    );
  }
  if (testFileIndex.has(taskFile)) {
    const currentIndex = testFileIndex.get(taskFile);
    if (currentIndex !== void 0) {
      testFileIndex.set(taskFile, currentIndex + 1);
    }
  } else {
    testFileIndex.set(taskFile, 1);
  }
  return `${taskFile}-${testFileIndex.get(taskFile)}`;
}
var ERROR_CODE_NOT_IMPLEMENTED_AS_DESIGNED = "NOT_IMPLEMENTED_AS_DESIGNED";
function replaceIllegalPathCharsAndSpace(str) {
  return str.replace(/[:*?"<>| ]/g, "-");
}
function forceClosePopup(page, debug3) {
  page.on("popup", async (popup) => {
    if (!popup) {
      console.warn("got a popup event, but the popup is not ready yet, skip");
      return;
    }
    const url = await popup.url();
    console.log(`Popup opened: ${url}`);
    if (!popup.isClosed()) {
      try {
        await popup.close();
      } catch (error) {
        debug3(`failed to close popup ${url}, error: ${error}`);
      }
    } else {
      debug3(`popup is already closed, skip close ${url}`);
    }
    if (!page.isClosed()) {
      try {
        await page.goto(url);
      } catch (error) {
        debug3(`failed to goto ${url}, error: ${error}`);
      }
    } else {
      debug3(`page is already closed, skip goto ${url}`);
    }
  });
}
function matchElementFromPlan(planLocateParam, tree) {
  if (!planLocateParam) {
    return void 0;
  }
  if (planLocateParam.id) {
    return getNodeFromCacheList(planLocateParam.id);
  }
  if (planLocateParam.bbox) {
    const centerPosition = {
      x: Math.floor((planLocateParam.bbox[0] + planLocateParam.bbox[2]) / 2),
      y: Math.floor((planLocateParam.bbox[1] + planLocateParam.bbox[3]) / 2)
    };
    let element = elementByPositionWithElementInfo(tree, centerPosition);
    if (!element) {
      element = generateElementByPosition(centerPosition);
    }
    return element;
  }
  return void 0;
}
async function matchElementFromCache(taskExecutor, xpaths, cachePrompt, cacheable) {
  try {
    if (xpaths?.length && taskExecutor.taskCache?.isCacheResultUsed && cacheable !== false) {
      for (let i = 0; i < xpaths.length; i++) {
        const element = await taskExecutor.page.getElementInfoByXpath(
          xpaths[i]
        );
        if (element?.id) {
          debug("cache hit, prompt: %s", cachePrompt);
          debug(
            "found a new element with same xpath, xpath: %s, id: %s",
            xpaths[i],
            element?.id
          );
          return element;
        }
      }
    }
  } catch (error) {
    debug("get element info by xpath error: ", error);
  }
}
function trimContextByViewport(execution) {
  function filterVisibleTree(node) {
    if (!node)
      return null;
    const filteredChildren = Array.isArray(node.children) ? node.children.map(filterVisibleTree).filter((child) => child !== null) : [];
    if (node.node && node.node.isVisible === true) {
      return {
        ...node,
        children: filteredChildren
      };
    }
    if (filteredChildren.length > 0) {
      return {
        node: null,
        children: filteredChildren
      };
    }
    return null;
  }
  return {
    ...execution,
    tasks: Array.isArray(execution.tasks) ? execution.tasks.map((task) => {
      const newTask = { ...task };
      if (task.pageContext?.tree) {
        newTask.pageContext = {
          ...task.pageContext,
          tree: filterVisibleTree(task.pageContext.tree) || {
            node: null,
            children: []
          }
        };
      }
      return newTask;
    }) : execution.tasks
  };
}
var parsePrompt = (prompt) => {
  if (typeof prompt === "string") {
    return {
      textPrompt: prompt,
      multimodalPrompt: void 0
    };
  }
  return {
    textPrompt: prompt.prompt,
    multimodalPrompt: prompt.images ? {
      images: prompt.images,
      convertHttpImage2Base64: !!prompt.convertHttpImage2Base64
    } : void 0
  };
};
export {
  ERROR_CODE_NOT_IMPLEMENTED_AS_DESIGNED,
  forceClosePopup,
  generateCacheId,
  getCurrentExecutionFile,
  getReportFileName,
  matchElementFromCache,
  matchElementFromPlan,
  parseContextFromWebPage,
  parsePrompt,
  printReportMsg,
  replaceIllegalPathCharsAndSpace,
  trimContextByViewport
};
