// @midscene/web/dist/es/yaml.js
// src/yaml/player.ts
import { existsSync, mkdirSync, writeFileSync } from "fs";
import { basename, dirname, join, resolve } from "path";
import { assert, ifInBrowser, ifInWorker } from "@midscene/shared/utils";
import { getMidsceneRunSubDir } from "@midscene/shared/common";
var ScriptPlayer = class {
  constructor(script, setupAgent, onTaskStatusChange, scriptPath) {
    this.script = script;
    this.setupAgent = setupAgent;
    this.onTaskStatusChange = onTaskStatusChange;
    this.taskStatusList = [];
    this.status = "init";
    this.unnamedResultIndex = 0;
    this.pageAgent = null;
    this.scriptPath = scriptPath;
    this.result = {};
    this.target = script.target || script.web || script.android;
    if (ifInBrowser || ifInWorker) {
      this.output = void 0;
    } else if (this.target?.output) {
      this.output = resolve(process.cwd(), this.target.output);
    } else {
      const scriptName = this.scriptPath ? basename(this.scriptPath, ".yaml").replace(/\.(ya?ml)$/i, "") : "script";
      this.output = join(
        getMidsceneRunSubDir("output"),
        `${scriptName}-${Date.now()}.json`
      );
    }
    if (ifInBrowser || ifInWorker) {
      this.unstableLogContent = void 0;
    } else if (typeof this.target?.unstableLogContent === "string") {
      this.unstableLogContent = resolve(
        process.cwd(),
        this.target.unstableLogContent
      );
    } else if (this.target?.unstableLogContent === true) {
      this.unstableLogContent = join(
        getMidsceneRunSubDir("output"),
        "unstableLogContent.json"
      );
    }
    this.taskStatusList = (script.tasks || []).map((task, taskIndex) => ({
      ...task,
      index: taskIndex,
      status: "init",
      totalSteps: task.flow?.length || 0
    }));
  }
  setResult(key, value) {
    const keyToUse = key || this.unnamedResultIndex++;
    if (this.result[keyToUse]) {
      console.warn(`result key ${keyToUse} already exists, will overwrite`);
    }
    this.result[keyToUse] = value;
    return this.flushResult();
  }
  setPlayerStatus(status, error) {
    this.status = status;
    this.errorInSetup = error;
  }
  notifyCurrentTaskStatusChange(taskIndex) {
    const taskIndexToNotify = typeof taskIndex === "number" ? taskIndex : this.currentTaskIndex;
    if (typeof taskIndexToNotify !== "number") {
      return;
    }
    const taskStatus = this.taskStatusList[taskIndexToNotify];
    if (this.onTaskStatusChange) {
      this.onTaskStatusChange(taskStatus);
    }
  }
  async setTaskStatus(index, statusValue, error) {
    this.taskStatusList[index].status = statusValue;
    if (error) {
      this.taskStatusList[index].error = error;
    }
    this.notifyCurrentTaskStatusChange(index);
  }
  setTaskIndex(taskIndex) {
    this.currentTaskIndex = taskIndex;
  }
  flushResult() {
    if (this.output) {
      const output = resolve(process.cwd(), this.output);
      const outputDir = dirname(output);
      if (!existsSync(outputDir)) {
        mkdirSync(outputDir, { recursive: true });
      }
      writeFileSync(output, JSON.stringify(this.result || {}, void 0, 2));
    }
  }
  flushUnstableLogContent() {
    if (this.unstableLogContent) {
      const content = this.pageAgent?._unstableLogContent();
      const filePath = resolve(process.cwd(), this.unstableLogContent);
      const outputDir = dirname(filePath);
      if (!existsSync(outputDir)) {
        mkdirSync(outputDir, { recursive: true });
      }
      writeFileSync(filePath, JSON.stringify(content, null, 2));
    }
  }
  async playTask(taskStatus, agent) {
    const { flow } = taskStatus;
    assert(flow, "missing flow in task");
    for (const flowItemIndex in flow) {
      const currentStep = Number.parseInt(flowItemIndex, 10);
      taskStatus.currentStep = currentStep;
      const flowItem = flow[flowItemIndex];
      if ("aiAction" in flowItem || "ai" in flowItem) {
        const actionTask = flowItem;
        const prompt = actionTask.aiAction || actionTask.ai;
        assert(prompt, "missing prompt for ai (aiAction)");
        assert(
          typeof prompt === "string",
          "prompt for aiAction must be a string"
        );
        await agent.aiAction(prompt, {
          cacheable: actionTask.cacheable
        });
      } else if ("aiAssert" in flowItem) {
        const assertTask = flowItem;
        const prompt = assertTask.aiAssert;
        const msg = assertTask.errorMessage;
        assert(prompt, "missing prompt for aiAssert");
        assert(
          typeof prompt === "string",
          "prompt for aiAssert must be a string"
        );
        await agent.aiAssert(prompt, msg);
      } else if ("aiQuery" in flowItem) {
        const queryTask = flowItem;
        const prompt = queryTask.aiQuery;
        const options = {
          domIncluded: queryTask.domIncluded,
          screenshotIncluded: queryTask.screenshotIncluded
        };
        assert(prompt, "missing prompt for aiQuery");
        assert(
          typeof prompt === "string",
          "prompt for aiQuery must be a string"
        );
        const queryResult = await agent.aiQuery(prompt, options);
        this.setResult(queryTask.name, queryResult);
      } else if ("aiNumber" in flowItem) {
        const numberTask = flowItem;
        const prompt = numberTask.aiNumber;
        const options = {
          domIncluded: numberTask.domIncluded,
          screenshotIncluded: numberTask.screenshotIncluded
        };
        assert(prompt, "missing prompt for aiNumber");
        assert(
          typeof prompt === "string",
          "prompt for number must be a string"
        );
        const numberResult = await agent.aiNumber(prompt, options);
        this.setResult(numberTask.name, numberResult);
      } else if ("aiString" in flowItem) {
        const stringTask = flowItem;
        const prompt = stringTask.aiString;
        const options = {
          domIncluded: stringTask.domIncluded,
          screenshotIncluded: stringTask.screenshotIncluded
        };
        assert(prompt, "missing prompt for aiString");
        assert(
          typeof prompt === "string",
          "prompt for string must be a string"
        );
        const stringResult = await agent.aiString(prompt, options);
        this.setResult(stringTask.name, stringResult);
      } else if ("aiBoolean" in flowItem) {
        const booleanTask = flowItem;
        const prompt = booleanTask.aiBoolean;
        const options = {
          domIncluded: booleanTask.domIncluded,
          screenshotIncluded: booleanTask.screenshotIncluded
        };
        assert(prompt, "missing prompt for aiBoolean");
        assert(
          typeof prompt === "string",
          "prompt for boolean must be a string"
        );
        const booleanResult = await agent.aiBoolean(prompt, options);
        this.setResult(booleanTask.name, booleanResult);
      } else if ("aiAsk" in flowItem) {
        const askTask = flowItem;
        const prompt = askTask.aiAsk;
        assert(prompt, "missing prompt for aiAsk");
        assert(typeof prompt === "string", "prompt for aiAsk must be a string");
        const askResult = await agent.aiAsk(prompt);
        this.setResult(askTask.name, askResult);
      } else if ("aiLocate" in flowItem) {
        const locateTask = flowItem;
        const prompt = locateTask.aiLocate;
        assert(prompt, "missing prompt for aiLocate");
        assert(
          typeof prompt === "string",
          "prompt for aiLocate must be a string"
        );
        const locateResult = await agent.aiLocate(prompt, locateTask);
        this.setResult(locateTask.name, locateResult);
      } else if ("aiWaitFor" in flowItem) {
        const waitForTask = flowItem;
        const prompt = waitForTask.aiWaitFor;
        assert(prompt, "missing prompt for aiWaitFor");
        assert(
          typeof prompt === "string",
          "prompt for aiWaitFor must be a string"
        );
        const timeout = waitForTask.timeout;
        await agent.aiWaitFor(prompt, { timeoutMs: timeout });
      } else if ("sleep" in flowItem) {
        const sleepTask = flowItem;
        const ms = sleepTask.sleep;
        let msNumber = ms;
        if (typeof ms === "string") {
          msNumber = Number.parseInt(ms, 10);
        }
        assert(
          msNumber && msNumber > 0,
          `ms for sleep must be greater than 0, but got ${ms}`
        );
        await new Promise((resolve2) => setTimeout(resolve2, msNumber));
      } else if ("aiTap" in flowItem) {
        const tapTask = flowItem;
        await agent.aiTap(tapTask.aiTap, tapTask);
      } else if ("aiRightClick" in flowItem) {
        const rightClickTask = flowItem;
        await agent.aiRightClick(rightClickTask.aiRightClick, rightClickTask);
      } else if ("aiHover" in flowItem) {
        const hoverTask = flowItem;
        await agent.aiHover(hoverTask.aiHover, hoverTask);
      } else if ("aiInput" in flowItem) {
        const inputTask = flowItem;
        await agent.aiInput(inputTask.aiInput, inputTask.locate, inputTask);
      } else if ("aiKeyboardPress" in flowItem) {
        const keyboardPressTask = flowItem;
        await agent.aiKeyboardPress(
          keyboardPressTask.aiKeyboardPress,
          keyboardPressTask.locate,
          keyboardPressTask
        );
      } else if ("aiScroll" in flowItem) {
        const scrollTask = flowItem;
        await agent.aiScroll(scrollTask, scrollTask.locate, scrollTask);
      } else if ("javascript" in flowItem) {
        const evaluateJavaScriptTask = flowItem;
        const result = await agent.evaluateJavaScript(
          evaluateJavaScriptTask.javascript
        );
        this.setResult(evaluateJavaScriptTask.name, result);
      } else if ("logScreenshot" in flowItem) {
        const logScreenshotTask = flowItem;
        await agent.logScreenshot(logScreenshotTask.logScreenshot, {
          content: logScreenshotTask.content || ""
        });
      } else {
        throw new Error(`unknown flowItem: ${JSON.stringify(flowItem)}`);
      }
    }
    this.reportFile = agent.reportFile;
    await this.flushUnstableLogContent();
  }
  async run() {
    const { target, web, android, tasks } = this.script;
    const webEnv = web || target;
    const androidEnv = android;
    const platform = webEnv || androidEnv;
    this.setPlayerStatus("running");
    let agent = null;
    let freeFn = [];
    try {
      const { agent: newAgent, freeFn: newFreeFn } = await this.setupAgent(
        platform
      );
      agent = newAgent;
      const originalOnTaskStartTip = agent.onTaskStartTip;
      agent.onTaskStartTip = (tip) => {
        if (this.status === "running") {
          this.agentStatusTip = tip;
        }
        originalOnTaskStartTip?.(tip);
      };
      freeFn = [
        ...newFreeFn || [],
        {
          name: "restore-agent-onTaskStartTip",
          fn: () => {
            if (agent) {
              agent.onTaskStartTip = originalOnTaskStartTip;
            }
          }
        }
      ];
    } catch (e) {
      this.setPlayerStatus("error", e);
      return;
    }
    this.pageAgent = agent;
    let taskIndex = 0;
    this.setPlayerStatus("running");
    let errorFlag = false;
    while (taskIndex < tasks.length) {
      const taskStatus = this.taskStatusList[taskIndex];
      this.setTaskStatus(taskIndex, "running");
      this.setTaskIndex(taskIndex);
      try {
        await this.playTask(taskStatus, this.pageAgent);
        this.setTaskStatus(taskIndex, "done");
      } catch (e) {
        this.setTaskStatus(taskIndex, "error", e);
        if (taskStatus.continueOnError) {
        } else {
          this.reportFile = agent.reportFile;
          errorFlag = true;
          break;
        }
      }
      this.reportFile = agent?.reportFile;
      taskIndex++;
    }
    if (errorFlag) {
      this.setPlayerStatus("error");
    } else {
      this.setPlayerStatus("done");
    }
    this.agentStatusTip = "";
    for (const fn of freeFn) {
      try {
        await fn.fn();
      } catch (e) {
      }
    }
  }
};

// src/yaml/builder.ts
import yaml from "js-yaml";
function buildYaml(env, tasks) {
  const result = {
    target: env,
    tasks
  };
  return yaml.dump(result, {
    indent: 2
  });
}

// src/yaml/utils.ts
import { assert as assert2 } from "@midscene/shared/utils";
import yaml2 from "js-yaml";
function interpolateEnvVars(content) {
  return content.replace(/\$\{([^}]+)\}/g, (_, envVar) => {
    const value = process.env[envVar.trim()];
    if (value === void 0) {
      throw new Error(`Environment variable "${envVar.trim()}" is not defined`);
    }
    return value;
  });
}
function parseYamlScript(content, filePath, ignoreCheckingTarget) {
  let processedContent = content;
  if (content.indexOf("android") !== -1 && content.match(/deviceId:\s*(\d+)/)) {
    let matchedDeviceId;
    processedContent = content.replace(
      /deviceId:\s*(\d+)/g,
      (match, deviceId) => {
        matchedDeviceId = deviceId;
        return `deviceId: '${deviceId}'`;
      }
    );
    console.warn(
      `please use string-style deviceId in yaml script, for example: deviceId: "${matchedDeviceId}"`
    );
  }
  const interpolatedContent = interpolateEnvVars(processedContent);
  const obj = yaml2.load(interpolatedContent, {
    schema: yaml2.JSON_SCHEMA
  });
  const pathTip = filePath ? `, failed to load ${filePath}` : "";
  const android = typeof obj.android !== "undefined" ? Object.assign({}, obj.android || {}) : void 0;
  const webConfig = obj.web || obj.target;
  const web = typeof webConfig !== "undefined" ? Object.assign({}, webConfig || {}) : void 0;
  if (!ignoreCheckingTarget) {
    assert2(
      web || android,
      `at least one of "target", "web", or "android" properties is required in yaml script${pathTip}`
    );
    assert2(
      web && !android || !web && android,
      `only one of "target", "web", or "android" properties is allowed in yaml script${pathTip}`
    );
    if (web || android) {
      assert2(
        typeof web === "object" || typeof android === "object",
        `property "target/web/android" must be an object${pathTip}`
      );
    }
  }
  assert2(obj.tasks, `property "tasks" is required in yaml script ${pathTip}`);
  assert2(
    Array.isArray(obj.tasks),
    `property "tasks" must be an array in yaml script, but got ${obj.tasks}`
  );
  return obj;
}
export {
  ScriptPlayer,
  buildYaml,
  interpolateEnvVars,
  parseYamlScript
};
